// @aztec/circuit-types/dest/prover_coordination/epoch_proof_quote_payload.js
import { EthAddress } from '@aztec/circuits.js';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
export class EpochProofQuotePayload {
    constructor(epochToProve, validUntilSlot, bondAmount, prover, basisPointFee) {
        this.epochToProve = epochToProve;
        this.validUntilSlot = validUntilSlot;
        this.bondAmount = bondAmount;
        this.prover = prover;
        this.basisPointFee = basisPointFee;
    }
    static getFields(fields) {
        return [
            fields.epochToProve,
            fields.validUntilSlot,
            fields.bondAmount,
            fields.prover,
            fields.basisPointFee,
        ];
    }
    toBuffer() {
        // We cache the buffer to avoid recalculating it
        if (this.asBuffer) {
            return this.asBuffer;
        }
        this.asBuffer = serializeToBuffer(...EpochProofQuotePayload.getFields(this));
        this.size = this.asBuffer.length;
        return this.asBuffer;
    }
    static fromBuffer(buf) {
        const reader = BufferReader.asReader(buf);
        return new EpochProofQuotePayload(reader.readUInt256(), reader.readUInt256(), reader.readUInt256(), reader.readObject(EthAddress), reader.readNumber());
    }
    static from(fields) {
        return new EpochProofQuotePayload(fields.epochToProve, fields.validUntilSlot, fields.bondAmount, fields.prover, fields.basisPointFee);
    }
    toJSON() {
        return {
            epochToProve: this.epochToProve.toString(),
            validUntilSlot: this.validUntilSlot.toString(),
            bondAmount: this.bondAmount.toString(),
            prover: this.prover.toString(),
            basisPointFee: this.basisPointFee,
        };
    }
    static fromJSON(obj) {
        return new EpochProofQuotePayload(BigInt(obj.epochToProve), BigInt(obj.validUntilSlot), BigInt(obj.bondAmount), EthAddress.fromString(obj.prover), obj.basisPointFee);
    }
    toViemArgs() {
        return {
            epochToProve: this.epochToProve,
            validUntilSlot: this.validUntilSlot,
            bondAmount: this.bondAmount,
            prover: this.prover.toString(),
            basisPointFee: this.basisPointFee,
        };
    }
    getSize() {
        // We cache size to avoid recalculating it
        if (this.size) {
            return this.size;
        }
        // Size is cached when calling toBuffer
        this.toBuffer();
        return this.size;
    }
    [inspect.custom]() {
        return `EpochProofQuotePayload { epochToProve: ${this.epochToProve}, validUntilSlot: ${this.validUntilSlot}, bondAmount: ${this.bondAmount}, prover: ${this.prover}, basisPointFee: ${this.basisPointFee} }`;
    }
}
