// @aztec/circuit-types/dest/prover_coordination/epoch_proof_quote.js
import { Buffer32 } from '@aztec/foundation/buffer';
import { keccak256 } from '@aztec/foundation/crypto';
import { Signature } from '@aztec/foundation/eth-signature';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { Gossipable } from '../p2p/gossipable.js';
import { TopicType, createTopicString } from '../p2p/topic_type.js';
import { EpochProofQuotePayload } from './epoch_proof_quote_payload.js';
export class EpochProofQuote extends Gossipable {
    constructor(payload, signature) {
        super();
        this.payload = payload;
        this.signature = signature;
    }
    static getFields(fields) {
        return [fields.payload, fields.signature];
    }
    p2pMessageIdentifier() {
        // TODO: https://github.com/AztecProtocol/aztec-packages/issues/8911
        return new Buffer32(keccak256(this.signature.toBuffer()));
    }
    toBuffer() {
        return serializeToBuffer(...EpochProofQuote.getFields(this));
    }
    static fromBuffer(buf) {
        const reader = BufferReader.asReader(buf);
        return new EpochProofQuote(reader.readObject(EpochProofQuotePayload), reader.readObject(Signature));
    }
    toJSON() {
        return {
            payload: this.payload.toJSON(),
            signature: this.signature.to0xString(),
        };
    }
    static fromJSON(obj) {
        return new EpochProofQuote(EpochProofQuotePayload.fromJSON(obj.payload), Signature.from0xString(obj.signature));
    }
    // TODO: https://github.com/AztecProtocol/aztec-packages/issues/8911
    /**
     * Creates a new quote with a signature.
     * The digest provided must match what the rollup contract will produce i.e. `_hashTypedDataV4(EpochProofQuoteLib.hash(quote))`
     *
     * @param digest the digest of the payload that should be signed
     * @param payload the actual quote
     * @param signer the signer
     * @returns a quote with an accompanying signature
     */
    static new(digest, payload, signer) {
        if (!payload.prover.equals(signer.address)) {
            throw new Error(`Quote prover does not match signer. Prover [${payload.prover}], Signer [${signer.address}]`);
        }
        const signature = signer.sign(digest);
        const quote = new EpochProofQuote(payload, signature);
        return quote;
    }
    toViemArgs() {
        return {
            quote: this.payload.toViemArgs(),
            signature: this.signature.toViemSignature(),
        };
    }
    /**
     * Get the size of the epoch proof quote in bytes.
     * @returns The size of the epoch proof quote in bytes.
     */
    getSize() {
        return this.payload.getSize() + this.signature.getSize();
    }
}
EpochProofQuote.p2pTopic = createTopicString(TopicType.epoch_proof_quote);
