// @aztec/circuits.js/dest/structs/gas.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { MAX_L2_GAS_PER_ENQUEUED_CALL } from '../constants.gen.js';
export const GasDimensions = ['da', 'l2'];
/** Gas amounts in each dimension. */
export class Gas {
    constructor(daGas, l2Gas) {
        this.daGas = daGas;
        this.l2Gas = l2Gas;
    }
    clone() {
        return new Gas(this.daGas, this.l2Gas);
    }
    get(dimension) {
        return this[`${dimension}Gas`];
    }
    equals(other) {
        return this.daGas === other.daGas && this.l2Gas === other.l2Gas;
    }
    static from(fields) {
        return new Gas(fields.daGas ?? 0, fields.l2Gas ?? 0);
    }
    static empty() {
        return new Gas(0, 0);
    }
    /** Returns large enough gas amounts for testing purposes. */
    static test() {
        return new Gas(1e9, MAX_L2_GAS_PER_ENQUEUED_CALL);
    }
    isEmpty() {
        return this.daGas === 0 && this.l2Gas === 0;
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new Gas(reader.readNumber(), reader.readNumber());
    }
    toBuffer() {
        return serializeToBuffer(this.daGas, this.l2Gas);
    }
    [inspect.custom]() {
        return `Gas { daGas=${this.daGas} l2Gas=${this.l2Gas} }`;
    }
    add(other) {
        return new Gas(this.daGas + other.daGas, this.l2Gas + other.l2Gas);
    }
    sub(other) {
        return new Gas(this.daGas - other.daGas, this.l2Gas - other.l2Gas);
    }
    mul(scalar) {
        return new Gas(Math.ceil(this.daGas * scalar), Math.ceil(this.l2Gas * scalar));
    }
    computeFee(gasFees) {
        return GasDimensions.reduce((acc, dimension) => acc.add(gasFees.get(dimension).mul(new Fr(this.get(dimension)))), Fr.ZERO);
    }
    toFields() {
        return serializeToFields(this.daGas, this.l2Gas);
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new Gas(reader.readU32(), reader.readU32());
    }
    toJSON() {
        return { daGas: this.daGas, l2Gas: this.l2Gas };
    }
    static fromJSON(json) {
        return new Gas(json.daGas, json.l2Gas);
    }
}
