// @aztec/circuits.js/dest/structs/gas_fees.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { inspect } from 'util';
/** Gas prices for each dimension. */
export class GasFees {
    constructor(feePerDaGas, feePerL2Gas) {
        this.feePerDaGas = new Fr(feePerDaGas);
        this.feePerL2Gas = new Fr(feePerL2Gas);
    }
    clone() {
        return new GasFees(this.feePerDaGas, this.feePerL2Gas);
    }
    equals(other) {
        return this.feePerDaGas.equals(other.feePerDaGas) && this.feePerL2Gas.equals(other.feePerL2Gas);
    }
    get(dimension) {
        switch (dimension) {
            case 'da':
                return this.feePerDaGas;
            case 'l2':
                return this.feePerL2Gas;
        }
    }
    static from(fields) {
        return new GasFees(fields.feePerDaGas, fields.feePerL2Gas);
    }
    static random() {
        return new GasFees(Fr.random(), Fr.random());
    }
    static empty() {
        return new GasFees(Fr.ZERO, Fr.ZERO);
    }
    /** Fixed gas fee values used until we define how gas fees in the protocol are computed. */
    static default() {
        return new GasFees(Fr.ONE, Fr.ONE);
    }
    isEmpty() {
        return this.feePerDaGas.isZero() && this.feePerL2Gas.isZero();
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new GasFees(reader.readObject(Fr), reader.readObject(Fr));
    }
    toBuffer() {
        return serializeToBuffer(this.feePerDaGas, this.feePerL2Gas);
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new GasFees(reader.readField(), reader.readField());
    }
    toFields() {
        return serializeToFields(this.feePerDaGas, this.feePerL2Gas);
    }
    static fromJSON(obj) {
        return new GasFees(Fr.fromString(obj.feePerDaGas), Fr.fromString(obj.feePerL2Gas));
    }
    toJSON() {
        return {
            feePerDaGas: this.feePerDaGas.toString(),
            feePerL2Gas: this.feePerL2Gas.toString(),
        };
    }
    [inspect.custom]() {
        return `GasFees { feePerDaGas=${this.feePerDaGas} feePerL2Gas=${this.feePerL2Gas} }`;
    }
}
