// @aztec/circuits.js/dest/structs/header.js
import { poseidon2HashWithSeparator } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { GeneratorIndex, HEADER_LENGTH } from '../constants.gen.js';
import { ContentCommitment } from './content_commitment.js';
import { GlobalVariables } from './global_variables.js';
import { AppendOnlyTreeSnapshot } from './rollup/append_only_tree_snapshot.js';
import { StateReference } from './state_reference.js';
/** A header of an L2 block. */
export class Header {
    constructor(
    /** Snapshot of archive before the block is applied. */
    lastArchive, 
    /** Hash of the body of an L2 block. */
    contentCommitment, 
    /** State reference. */
    state, 
    /** Global variables of an L2 block. */
    globalVariables, 
    /** Total fees in the block, computed by the root rollup circuit */
    totalFees) {
        this.lastArchive = lastArchive;
        this.contentCommitment = contentCommitment;
        this.state = state;
        this.globalVariables = globalVariables;
        this.totalFees = totalFees;
    }
    static getFields(fields) {
        // Note: The order here must match the order in the HeaderLib solidity library.
        return [
            fields.lastArchive,
            fields.contentCommitment,
            fields.state,
            fields.globalVariables,
            fields.totalFees,
        ];
    }
    static from(fields) {
        return new Header(...Header.getFields(fields));
    }
    getSize() {
        return (this.lastArchive.getSize() +
            this.contentCommitment.getSize() +
            this.state.getSize() +
            this.globalVariables.getSize() +
            this.totalFees.size);
    }
    toBuffer() {
        return serializeToBuffer(...Header.getFields(this));
    }
    toFields() {
        const fields = serializeToFields(...Header.getFields(this));
        if (fields.length !== HEADER_LENGTH) {
            throw new Error(`Invalid number of fields for Header. Expected ${HEADER_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    clone() {
        return Header.fromBuffer(this.toBuffer());
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new Header(reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(ContentCommitment), reader.readObject(StateReference), reader.readObject(GlobalVariables), reader.readObject(Fr));
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new Header(AppendOnlyTreeSnapshot.fromFields(reader), ContentCommitment.fromFields(reader), StateReference.fromFields(reader), GlobalVariables.fromFields(reader), reader.readField());
    }
    static empty(fields = {}) {
        return Header.from({
            lastArchive: AppendOnlyTreeSnapshot.zero(),
            contentCommitment: ContentCommitment.empty(),
            state: StateReference.empty(),
            globalVariables: GlobalVariables.empty(),
            totalFees: Fr.ZERO,
            ...fields,
        });
    }
    isEmpty() {
        return (this.lastArchive.isZero() &&
            this.contentCommitment.isEmpty() &&
            this.state.isEmpty() &&
            this.globalVariables.isEmpty());
    }
    /**
     * Serializes this instance into a string.
     * @returns Encoded string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromString(str) {
        const buffer = Buffer.from(str.replace(/^0x/i, ''), 'hex');
        return Header.fromBuffer(buffer);
    }
    hash() {
        return poseidon2HashWithSeparator(this.toFields(), GeneratorIndex.BLOCK_HASH);
    }
    [inspect.custom]() {
        return `Header {
  lastArchive: ${inspect(this.lastArchive)},
  contentCommitment.numTx: ${this.contentCommitment.numTxs.toNumber()},
  contentCommitment.txsEffectsHash: ${this.contentCommitment.txsEffectsHash.toString('hex')},
  contentCommitment.inHash: ${this.contentCommitment.inHash.toString('hex')},
  contentCommitment.outHash: ${this.contentCommitment.outHash.toString('hex')},
  state.l1ToL2MessageTree: ${inspect(this.state.l1ToL2MessageTree)},
  state.noteHashTree: ${inspect(this.state.partial.noteHashTree)},
  state.nullifierTree: ${inspect(this.state.partial.nullifierTree)},
  state.publicDataTree: ${inspect(this.state.partial.publicDataTree)},
  globalVariables: ${inspect(this.globalVariables)},
  totalFees: ${this.totalFees},
}`;
    }
    equals(other) {
        return (this.contentCommitment.equals(other.contentCommitment) &&
            this.state.equals(other.state) &&
            this.globalVariables.equals(other.globalVariables) &&
            this.totalFees.equals(other.totalFees) &&
            this.lastArchive.equals(other.lastArchive));
    }
}
