// @aztec/circuits.js/dest/structs/non_existent_read_request_hints.js
import { makeTuple } from '@aztec/foundation/array';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { MAX_NULLIFIERS_PER_TX, MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_TX, NULLIFIER_TREE_HEIGHT, } from '../constants.gen.js';
import { MembershipWitness } from './membership_witness.js';
import { Nullifier } from './nullifier.js';
import { NullifierLeafPreimage } from './trees/index.js';
export class NonMembershipHint {
    constructor(membershipWitness, leafPreimage) {
        this.membershipWitness = membershipWitness;
        this.leafPreimage = leafPreimage;
    }
    static empty(treeHeight, makeEmptyLeafPreimage) {
        return new NonMembershipHint(MembershipWitness.empty(treeHeight), makeEmptyLeafPreimage());
    }
    static fromBuffer(buffer, treeHeight, leafPreimageFromBuffer) {
        const reader = BufferReader.asReader(buffer);
        return new NonMembershipHint(MembershipWitness.fromBuffer(reader, treeHeight), reader.readObject(leafPreimageFromBuffer));
    }
    toBuffer() {
        return serializeToBuffer(this.membershipWitness, this.leafPreimage);
    }
}
export class NonExistentReadRequestHints {
    constructor(
    /**
     * The hints for the low leaves of the read requests.
     */
    nonMembershipHints, 
    /**
     * Indices of the smallest pending values greater than the read requests.
     */
    nextPendingValueIndices, sortedPendingValues, sortedPendingValueHints) {
        this.nonMembershipHints = nonMembershipHints;
        this.nextPendingValueIndices = nextPendingValueIndices;
        this.sortedPendingValues = sortedPendingValues;
        this.sortedPendingValueHints = sortedPendingValueHints;
    }
    static fromBuffer(buffer, readRequestLen, treeHeight, leafPreimageFromBuffer, pendingValueLen, orderedValueFromBuffer) {
        const reader = BufferReader.asReader(buffer);
        return new NonExistentReadRequestHints(reader.readArray(readRequestLen, {
            fromBuffer: buf => NonMembershipHint.fromBuffer(buf, treeHeight, leafPreimageFromBuffer),
        }), reader.readNumbers(readRequestLen), reader.readArray(pendingValueLen, orderedValueFromBuffer), reader.readNumbers(pendingValueLen));
    }
    toBuffer() {
        return serializeToBuffer(this.nonMembershipHints, this.nextPendingValueIndices, this.sortedPendingValues, this.sortedPendingValueHints);
    }
}
export function nullifierNonExistentReadRequestHintsFromBuffer(buffer) {
    return NonExistentReadRequestHints.fromBuffer(buffer, MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_TX, NULLIFIER_TREE_HEIGHT, NullifierLeafPreimage, MAX_NULLIFIERS_PER_TX, Nullifier);
}
export class NullifierNonExistentReadRequestHintsBuilder {
    constructor(sortedPendingNullifiers, sortedPendingNullifierIndexHints) {
        this.readRequestIndex = 0;
        this.hints = new NonExistentReadRequestHints(makeTuple(MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_TX, () => NonMembershipHint.empty(NULLIFIER_TREE_HEIGHT, NullifierLeafPreimage.empty)), makeTuple(MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_TX, () => 0), sortedPendingNullifiers, sortedPendingNullifierIndexHints);
    }
    static empty() {
        const emptySortedPendingNullifiers = makeTuple(MAX_NULLIFIERS_PER_TX, Nullifier.empty);
        const emptySortedPendingNullifierIndexHints = makeTuple(MAX_NULLIFIERS_PER_TX, () => 0);
        return new NullifierNonExistentReadRequestHintsBuilder(emptySortedPendingNullifiers, emptySortedPendingNullifierIndexHints).toHints();
    }
    addHint(membershipWitness, lowLeafPreimage, nextPendingValueIndex) {
        this.hints.nonMembershipHints[this.readRequestIndex] = new NonMembershipHint(membershipWitness, lowLeafPreimage);
        this.hints.nextPendingValueIndices[this.readRequestIndex] = nextPendingValueIndex;
        this.readRequestIndex++;
    }
    toHints() {
        return this.hints;
    }
}
