// @aztec/circuits.js/dest/structs/proof.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AGGREGATION_OBJECT_LENGTH } from '../constants.gen.js';
const EMPTY_PROOF_SIZE = 42;
/**
 * The Proof class is a wrapper around the circuits proof.
 * Underlying it is a buffer of proof data in a form a barretenberg prover understands.
 * It provides methods to easily create, serialize, and deserialize the proof data for efficient
 * communication and storage.
 */
export class Proof {
    constructor(
    /**
     * Holds the serialized proof data in a binary buffer format.
     */
    buffer, numPublicInputs) {
        this.buffer = buffer;
        this.numPublicInputs = numPublicInputs;
        // Honk proofs start with a 4 byte length prefix
        // the proof metadata starts immediately after
        this.metadataOffset = 4;
        // the metadata is 3 Frs long
        // the public inputs are after it
        this.publicInputsOffset = 100;
    }
    /**
     * Create a Proof from a Buffer or BufferReader.
     * Expects a length-encoding.
     *
     * @param buffer - A Buffer or BufferReader containing the length-encoded proof data.
     * @returns A Proof instance containing the decoded proof data.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const size = reader.readNumber();
        const buf = reader.readBytes(size);
        const numPublicInputs = reader.readNumber();
        return new Proof(buf, numPublicInputs);
    }
    /**
     * Convert the Proof instance to a custom Buffer format.
     * This function serializes the Proof's buffer length and data sequentially into a new Buffer.
     *
     * @returns A Buffer containing the serialized proof data in custom format.
     */
    toBuffer() {
        return serializeToBuffer(this.buffer.length, this.buffer, this.numPublicInputs);
    }
    /**
     * Serialize the Proof instance to a hex string.
     * @returns The hex string representation of the proof data.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    withoutPublicInputs() {
        return Buffer.concat([
            this.buffer.subarray(this.metadataOffset, this.publicInputsOffset),
            this.buffer.subarray(this.publicInputsOffset + Fr.SIZE_IN_BYTES * this.numPublicInputs),
        ]);
    }
    extractPublicInputs() {
        const reader = BufferReader.asReader(this.buffer.subarray(this.publicInputsOffset, this.publicInputsOffset + Fr.SIZE_IN_BYTES * this.numPublicInputs));
        return reader.readArray(this.numPublicInputs, Fr);
    }
    extractAggregationObject() {
        const publicInputs = this.extractPublicInputs();
        return publicInputs.slice(-1 * AGGREGATION_OBJECT_LENGTH);
    }
    /**
     * Deserialize a Proof instance from a hex string.
     * @param str - A hex string to deserialize from.
     * @returns - A new Proof instance.
     */
    static fromString(str) {
        return Proof.fromBuffer(Buffer.from(str, 'hex'));
    }
    /** Returns whether this proof is actually empty. */
    isEmpty() {
        return (this.buffer.length === EMPTY_PROOF_SIZE && this.buffer.every(byte => byte === 0) && this.numPublicInputs === 0);
    }
}
/**
 * Makes an empty proof.
 * Note: Used for local devnet milestone where we are not proving anything yet.
 * @returns The empty "proof".
 */
export function makeEmptyProof() {
    return new Proof(Buffer.alloc(EMPTY_PROOF_SIZE, 0), 0);
}
