// @aztec/circuits.js/dest/structs/revert_code.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader } from '@aztec/foundation/serialize';
import { inspect } from 'util';
var RevertCodeEnum;
(function (RevertCodeEnum) {
    RevertCodeEnum[RevertCodeEnum["OK"] = 0] = "OK";
    RevertCodeEnum[RevertCodeEnum["APP_LOGIC_REVERTED"] = 1] = "APP_LOGIC_REVERTED";
    RevertCodeEnum[RevertCodeEnum["TEARDOWN_REVERTED"] = 2] = "TEARDOWN_REVERTED";
    RevertCodeEnum[RevertCodeEnum["BOTH_REVERTED"] = 3] = "BOTH_REVERTED";
})(RevertCodeEnum || (RevertCodeEnum = {}));
function isRevertCodeEnum(value) {
    return (value === RevertCodeEnum.OK ||
        value === RevertCodeEnum.APP_LOGIC_REVERTED ||
        value === RevertCodeEnum.TEARDOWN_REVERTED ||
        value === RevertCodeEnum.BOTH_REVERTED);
}
/**
 * Wrapper class over a field to safely represent a revert code.
 */
export class RevertCode {
    constructor(e) {
        this.code = e.valueOf();
    }
    equals(other) {
        return this.code === other.code;
    }
    isOK() {
        return this.equals(RevertCode.OK);
    }
    getDescription() {
        switch (this.code) {
            case RevertCodeEnum.OK:
                return 'OK';
            case RevertCodeEnum.APP_LOGIC_REVERTED:
                return 'Application logic reverted';
            case RevertCodeEnum.TEARDOWN_REVERTED:
                return 'Teardown reverted';
            case RevertCodeEnum.BOTH_REVERTED:
                return 'Both reverted';
            default:
                return `Unknown RevertCode: ${this.code}`;
        }
    }
    toHashPreimage() {
        const padding = Buffer.alloc(RevertCode.PREIMAGE_SIZE_IN_BYTES - RevertCode.PACKED_SIZE_IN_BYTES);
        return Buffer.concat([padding, this.toBuffer()]);
    }
    toBuffer() {
        const b = Buffer.alloc(RevertCode.PACKED_SIZE_IN_BYTES);
        b.writeUInt8(this.code, 0);
        return b;
    }
    toField() {
        return new Fr(this.toBuffer());
    }
    getSerializedLength() {
        return this.toBuffer().length;
    }
    static fromField(fr) {
        if (!isRevertCodeEnum(fr.toNumber())) {
            throw new Error(`Invalid RevertCode: ${fr.toNumber()}`);
        }
        return new RevertCode(fr.toNumber());
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return RevertCode.fromField(reader.readField());
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const code = reader.readBytes(RevertCode.PACKED_SIZE_IN_BYTES).readUInt8(0);
        if (!isRevertCodeEnum(code)) {
            throw new Error(`Invalid RevertCode: ${code}`);
        }
        return new RevertCode(code);
    }
    static random() {
        return new RevertCode(Math.floor(Math.random() * RevertCode.NUM_OPTIONS));
    }
    [inspect.custom]() {
        return `RevertCode<${this.code.toString()}>`;
    }
}
RevertCode.OK = new RevertCode(RevertCodeEnum.OK);
RevertCode.APP_LOGIC_REVERTED = new RevertCode(RevertCodeEnum.APP_LOGIC_REVERTED);
RevertCode.TEARDOWN_REVERTED = new RevertCode(RevertCodeEnum.TEARDOWN_REVERTED);
RevertCode.BOTH_REVERTED = new RevertCode(RevertCodeEnum.BOTH_REVERTED);
/**
 * Having different serialization methods allows for
 * decoupling the serialization for producing the content commitment hash
 * (where we use fields)
 * from serialization for transmitting the data.
 */
RevertCode.PREIMAGE_SIZE_IN_BYTES = 32;
RevertCode.PACKED_SIZE_IN_BYTES = 1;
RevertCode.NUM_OPTIONS = 4;
