// @aztec/circuits.js/dest/structs/scoped_key_validation_request_and_generator.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { SCOPED_KEY_VALIDATION_REQUEST_AND_GENERATOR_LENGTH } from '../constants.gen.js';
import { KeyValidationRequestAndGenerator } from './key_validation_request_and_generator.js';
/**
 * Request for validating keys used in the app.
 */
export class ScopedKeyValidationRequestAndGenerator {
    constructor(request, contractAddress) {
        this.request = request;
        this.contractAddress = contractAddress;
    }
    toBuffer() {
        return serializeToBuffer(this.request, this.contractAddress);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ScopedKeyValidationRequestAndGenerator(KeyValidationRequestAndGenerator.fromBuffer(reader), AztecAddress.fromBuffer(reader));
    }
    toFields() {
        const fields = [...this.request.toFields(), this.contractAddress];
        if (fields.length !== SCOPED_KEY_VALIDATION_REQUEST_AND_GENERATOR_LENGTH) {
            throw new Error(`Invalid number of fields for ScopedKeyValidationRequestAndGenerator. Expected ${SCOPED_KEY_VALIDATION_REQUEST_AND_GENERATOR_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new ScopedKeyValidationRequestAndGenerator(KeyValidationRequestAndGenerator.fromFields(reader), AztecAddress.fromFields(reader));
    }
    isEmpty() {
        return this.request.isEmpty() && this.contractAddress.isZero();
    }
    static empty() {
        return new ScopedKeyValidationRequestAndGenerator(KeyValidationRequestAndGenerator.empty(), AztecAddress.ZERO);
    }
}
