// @aztec/circuits.js/dest/structs/side_effects.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
/**
 * Side-effect object consisting of a value and a counter.
 * cpp/src/aztec3/circuits/abis/side_effects.hpp.
 */
export class SideEffect {
    constructor(
    /**
     * The value of the side-effect object.
     */
    value, 
    /**
     * The side-effect counter.
     */
    counter) {
        this.value = value;
        this.counter = counter;
    }
    toString() {
        return `value=${this.value.toString()} counter=${this.counter.toString()}`;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.value, this.counter);
    }
    /**
     * Convert to an array of fields.
     * @returns The array of fields.
     */
    toFields() {
        return [this.value, this.counter];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new SideEffect(reader.readField(), reader.readField());
    }
    /**
     * Returns whether this instance of side-effect is empty.
     * @returns True if the value and counter both are zero.
     */
    isEmpty() {
        return SideEffect.isEmpty(this);
    }
    /**
     * Checks whether this instance of side-effect is empty.
     * @returns True if the value and counter both are zero.
     */
    static isEmpty(sideEffect) {
        return sideEffect.value.isZero() && sideEffect.counter.isZero();
    }
    /**
     * Returns an empty instance of side-effect.
     * @returns Side-effect with both value and counter being zero.
     */
    static empty() {
        return new SideEffect(Fr.zero(), Fr.zero());
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of SideEffect.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new SideEffect(Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
}
