// @aztec/circuits.js/dest/structs/trees/nullifier_leaf.js
import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
/**
 * Class containing the data of a preimage of a single leaf in the nullifier tree.
 * Note: It's called preimage because this data gets hashed before being inserted as a node into the `IndexedTree`.
 */
export class NullifierLeafPreimage {
    constructor(
    /**
     * Leaf value inside the indexed tree's linked list.
     */
    nullifier, 
    /**
     * Next value inside the indexed tree's linked list.
     */
    nextNullifier, 
    /**
     * Index of the next leaf in the indexed tree's linked list.
     */
    nextIndex) {
        this.nullifier = nullifier;
        this.nextNullifier = nextNullifier;
        this.nextIndex = nextIndex;
    }
    getKey() {
        return this.nullifier.toBigInt();
    }
    getNextKey() {
        return this.nextNullifier.toBigInt();
    }
    getNextIndex() {
        return this.nextIndex;
    }
    asLeaf() {
        return new NullifierLeaf(this.nullifier);
    }
    toBuffer() {
        return Buffer.concat(this.toHashInputs());
    }
    toHashInputs() {
        return [
            Buffer.from(this.nullifier.toBuffer()),
            Buffer.from(this.nextNullifier.toBuffer()),
            Buffer.from(toBufferBE(this.nextIndex, 32)),
        ];
    }
    toFields() {
        return this.toHashInputs().map(buf => Fr.fromBuffer(buf));
    }
    clone() {
        return new NullifierLeafPreimage(this.nullifier, this.nextNullifier, this.nextIndex);
    }
    toJSON() {
        return {
            nullifier: this.nullifier.toString(),
            nextNullifier: this.nextNullifier.toString(),
            nextIndex: '0x' + this.nextIndex.toString(16),
        };
    }
    static empty() {
        return new NullifierLeafPreimage(Fr.ZERO, Fr.ZERO, 0n);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new NullifierLeafPreimage(reader.readObject(Fr), reader.readObject(Fr), toBigIntBE(reader.readBytes(32)));
    }
    static fromLeaf(leaf, nextKey, nextIndex) {
        return new NullifierLeafPreimage(leaf.nullifier, new Fr(nextKey), nextIndex);
    }
    static clone(preimage) {
        return new NullifierLeafPreimage(preimage.nullifier, preimage.nextNullifier, preimage.nextIndex);
    }
    static fromJSON(json) {
        return new NullifierLeafPreimage(Fr.fromString(json.nullifier), Fr.fromString(json.nextNullifier), BigInt(json.nextIndex));
    }
}
/**
 * A nullifier to be inserted in the nullifier tree.
 */
export class NullifierLeaf {
    constructor(
    /**
     * Nullifier value.
     */
    nullifier) {
        this.nullifier = nullifier;
    }
    getKey() {
        return this.nullifier.toBigInt();
    }
    toBuffer() {
        return this.nullifier.toBuffer();
    }
    isEmpty() {
        return this.nullifier.isZero();
    }
    updateTo(_another) {
        throw new Error('Nullifiers are create only');
    }
    static buildDummy(key) {
        return new NullifierLeaf(new Fr(key));
    }
    static fromBuffer(buf) {
        return new NullifierLeaf(Fr.fromBuffer(buf));
    }
}
