// @aztec/circuits.js/dest/structs/trees/public_data_leaf.js
import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
/**
 * Class containing the data of a preimage of a single leaf in the public data tree.
 * Note: It's called preimage because this data gets hashed before being inserted as a node into the `IndexedTree`.
 */
export class PublicDataTreeLeafPreimage {
    constructor(
    /**
     * The slot of the leaf
     */
    slot, 
    /**
     * The value of the leaf
     */
    value, 
    /**
     * Next value inside the indexed tree's linked list.
     */
    nextSlot, 
    /**
     * Index of the next leaf in the indexed tree's linked list.
     */
    nextIndex) {
        this.slot = slot;
        this.value = value;
        this.nextSlot = nextSlot;
        this.nextIndex = nextIndex;
    }
    getKey() {
        return this.slot.toBigInt();
    }
    getNextKey() {
        return this.nextSlot.toBigInt();
    }
    getNextIndex() {
        return this.nextIndex;
    }
    asLeaf() {
        return new PublicDataTreeLeaf(this.slot, this.value);
    }
    toBuffer() {
        return Buffer.concat(this.toHashInputs());
    }
    toHashInputs() {
        return [
            Buffer.from(this.slot.toBuffer()),
            Buffer.from(this.value.toBuffer()),
            Buffer.from(toBufferBE(this.nextIndex, 32)),
            Buffer.from(this.nextSlot.toBuffer()),
        ];
    }
    clone() {
        return new PublicDataTreeLeafPreimage(this.slot, this.value, this.nextSlot, this.nextIndex);
    }
    static empty() {
        return new PublicDataTreeLeafPreimage(Fr.ZERO, Fr.ZERO, Fr.ZERO, 0n);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const slot = Fr.fromBuffer(reader);
        const value = Fr.fromBuffer(reader);
        const nextIndex = toBigIntBE(reader.readBytes(32));
        const nextSlot = Fr.fromBuffer(reader);
        return new PublicDataTreeLeafPreimage(slot, value, nextSlot, nextIndex);
    }
    static fromLeaf(leaf, nextKey, nextIndex) {
        return new PublicDataTreeLeafPreimage(leaf.slot, leaf.value, new Fr(nextKey), nextIndex);
    }
    static clone(preimage) {
        return new PublicDataTreeLeafPreimage(preimage.slot, preimage.value, preimage.nextSlot, preimage.nextIndex);
    }
}
/**
 * A leaf in the public data indexed tree.
 */
export class PublicDataTreeLeaf {
    constructor(
    /**
     * The slot the value is stored in
     */
    slot, 
    /**
     * The value stored in the slot
     */
    value) {
        this.slot = slot;
        this.value = value;
    }
    getKey() {
        return this.slot.toBigInt();
    }
    toBuffer() {
        return serializeToBuffer([this.slot, this.value]);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicDataTreeLeaf(Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    equals(another) {
        return this.slot.equals(another.slot) && this.value.equals(another.value);
    }
    toString() {
        return `PublicDataTreeLeaf(${this.slot.toString()}, ${this.value.toString()})`;
    }
    isEmpty() {
        return this.slot.isZero() && this.value.isZero();
    }
    updateTo(another) {
        if (!this.slot.equals(another.slot)) {
            throw new Error('Invalid update: slots do not match');
        }
        return new PublicDataTreeLeaf(this.slot, another.value);
    }
    static buildDummy(key) {
        return new PublicDataTreeLeaf(new Fr(key), new Fr(0));
    }
    static empty() {
        return new PublicDataTreeLeaf(Fr.ZERO, Fr.ZERO);
    }
}
