// @aztec/circuit-types/dest/logs/l1_payload/l1_event_payload.js
var _a, _L1EventPayload_fromIncomingBodyPlaintextAndContractAddress;
import { __classPrivateFieldGet } from "tslib";
import { AztecAddress } from '@aztec/circuits.js';
import { EventSelector } from '@aztec/foundation/abi';
import { poseidon2HashWithSeparator } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { EncryptedLogPayload } from './encrypted_log_payload.js';
import { Event } from './payload.js';
/**
 * A class which wraps event data which is pushed on L1.
 */
export class L1EventPayload {
    constructor(
    /**
     * A event as emitted from Noir contract. Can be used along with private key to compute nullifier.
     */
    event, 
    /**
     * Address of the contract this tx is interacting with.
     */
    contractAddress, 
    /**
     * Randomness used to mask the contract address.
     */
    randomness, 
    /**
     * Type identifier for the underlying event, required to determine how to compute its hash and nullifier.
     */
    eventTypeId) {
        this.event = event;
        this.contractAddress = contractAddress;
        this.randomness = randomness;
        this.eventTypeId = eventTypeId;
    }
    static decryptAsIncoming(log, sk) {
        const decryptedLog = EncryptedLogPayload.decryptAsIncoming(log.data, sk);
        if (!decryptedLog) {
            return undefined;
        }
        return __classPrivateFieldGet(this, _a, "m", _L1EventPayload_fromIncomingBodyPlaintextAndContractAddress).call(this, decryptedLog.incomingBodyPlaintext, decryptedLog.contractAddress, log.maskedContractAddress);
    }
    static decryptAsOutgoing(log, sk) {
        const decryptedLog = EncryptedLogPayload.decryptAsOutgoing(log.data, sk);
        if (!decryptedLog) {
            return undefined;
        }
        return __classPrivateFieldGet(this, _a, "m", _L1EventPayload_fromIncomingBodyPlaintextAndContractAddress).call(this, decryptedLog.incomingBodyPlaintext, decryptedLog.contractAddress, log.maskedContractAddress);
    }
    /**
     * Serializes the L1EventPayload object into a Buffer.
     * @returns Buffer representation of the L1EventPayload object.
     */
    toIncomingBodyPlaintext() {
        const fields = [this.randomness, this.eventTypeId.toField(), ...this.event.items];
        return serializeToBuffer(fields);
    }
    /**
     * Create a random L1EventPayload object (useful for testing purposes).
     * @param contract - The address of a contract the event was emitted from.
     * @returns A random L1EventPayload object.
     */
    static random(contract = AztecAddress.random()) {
        return new _a(Event.random(), contract, Fr.random(), EventSelector.random());
    }
    equals(other) {
        return (this.event.equals(other.event) &&
            this.contractAddress.equals(other.contractAddress) &&
            this.randomness.equals(other.randomness) &&
            this.eventTypeId.equals(other.eventTypeId));
    }
}
_a = L1EventPayload, _L1EventPayload_fromIncomingBodyPlaintextAndContractAddress = function _L1EventPayload_fromIncomingBodyPlaintextAndContractAddress(plaintext, contractAddress, maskedContractAddress) {
    let payload;
    try {
        const reader = BufferReader.asReader(plaintext);
        const fields = reader.readArray(plaintext.length / Fr.SIZE_IN_BYTES, Fr);
        const randomness = fields[0];
        const eventTypeId = EventSelector.fromField(fields[1]);
        const event = new Event(fields.slice(2));
        payload = new _a(event, contractAddress, randomness, eventTypeId);
    }
    catch (e) {
        return undefined;
    }
    ensureMatchedMaskedContractAddress(contractAddress, payload.randomness, maskedContractAddress);
    return payload;
};
function ensureMatchedMaskedContractAddress(contractAddress, randomness, maskedContractAddress) {
    if (!poseidon2HashWithSeparator([contractAddress, randomness], 0).equals(maskedContractAddress)) {
        throw new Error('The provided masked contract address does not match with the incoming address from header and randomness from body');
    }
}
