// @aztec/accounts/dest/single_key/account_contract.js
import { AuthWitness } from '@aztec/circuit-types';
import { Schnorr } from '@aztec/circuits.js/barretenberg';
import { DefaultAccountContract } from '../defaults/account_contract.js';
import { SchnorrSingleKeyAccountContractArtifact } from './artifact.js';
/**
 * Account contract that authenticates transactions using Schnorr signatures verified against
 * the note encryption key, relying on a single private key for both encryption and authentication.
 */
export class SingleKeyAccountContract extends DefaultAccountContract {
    constructor(encryptionPrivateKey) {
        super(SchnorrSingleKeyAccountContractArtifact);
        this.encryptionPrivateKey = encryptionPrivateKey;
    }
    getDeploymentArgs() {
        return undefined;
    }
    getAuthWitnessProvider(account) {
        return new SingleKeyAuthWitnessProvider(this.encryptionPrivateKey, account);
    }
}
/**
 * Creates auth witnesses using Schnorr signatures and including the partial address and public key
 * in the witness, so verifiers do not need to store the public key and can instead validate it
 * by reconstructing the current address.
 */
class SingleKeyAuthWitnessProvider {
    constructor(privateKey, account) {
        this.privateKey = privateKey;
        this.account = account;
    }
    createAuthWit(messageHash) {
        const schnorr = new Schnorr();
        const signature = schnorr.constructSignature(messageHash.toBuffer(), this.privateKey);
        const witness = [...this.account.publicKeys.toFields(), ...signature.toBuffer(), this.account.partialAddress];
        return Promise.resolve(new AuthWitness(messageHash, witness));
    }
}
