// @aztec/accounts/dest/testing/create_account.js
import { Fr, deriveSigningKey } from '@aztec/circuits.js';
import { getSchnorrAccount } from '../schnorr/index.js';
/**
 * Deploys and registers a new account using random private keys and returns the associated Schnorr account wallet. Useful for testing.
 * @param pxe - PXE.
 * @returns - A wallet for a fresh account.
 */
export function createAccount(pxe) {
    const secretKey = Fr.random();
    const signingKey = deriveSigningKey(secretKey);
    return getSchnorrAccount(pxe, secretKey, signingKey).waitSetup();
}
/**
 * Creates a given number of random accounts using the Schnorr account wallet.
 * @param pxe - PXE.
 * @param numberOfAccounts - How many accounts to create.
 * @param secrets - Optional array of secrets to use for the accounts. If empty, random secrets will be generated.
 * @throws If the secrets array is not empty and does not have the same length as the number of accounts.
 * @returns The created account wallets.
 */
export async function createAccounts(pxe, numberOfAccounts = 1, secrets = [], waitOpts = { interval: 0.1 }) {
    if (secrets.length == 0) {
        secrets = Array.from({ length: numberOfAccounts }, () => Fr.random());
    }
    else if (secrets.length > 0 && secrets.length !== numberOfAccounts) {
        throw new Error('Secrets array must be empty or have the same length as the number of accounts');
    }
    // Prepare deployments
    const accountsAndDeployments = await Promise.all(secrets.map(async (secret) => {
        const signingKey = deriveSigningKey(secret);
        const account = getSchnorrAccount(pxe, secret, signingKey);
        const deployMethod = await account.getDeployMethod();
        const provenTx = await deployMethod.prove({
            contractAddressSalt: account.salt,
            skipClassRegistration: true,
            skipPublicDeployment: true,
            universalDeploy: true,
        });
        return { account, provenTx };
    }));
    // Send them and await them to be mined
    await Promise.all(accountsAndDeployments.map(({ provenTx }) => provenTx.send().wait(waitOpts)));
    return Promise.all(accountsAndDeployments.map(({ account }) => account.getWallet()));
}
