// @aztec/circuit-types/dest/interfaces/prover-client.js
var _a, _BlockProofError_name;
import { __classPrivateFieldGet } from "tslib";
import { Fr } from '@aztec/circuits.js';
import { booleanConfigHelper, numberConfigHelper } from '@aztec/foundation/config';
export const proverConfigMappings = {
    nodeUrl: {
        env: 'AZTEC_NODE_URL',
        description: 'The URL to the Aztec node to take proving jobs from',
    },
    realProofs: {
        env: 'PROVER_REAL_PROOFS',
        description: 'Whether to construct real proofs',
        ...booleanConfigHelper(),
    },
    proverAgentEnabled: {
        env: 'PROVER_AGENT_ENABLED',
        description: 'Whether this prover has a local prover agent',
        ...booleanConfigHelper(true),
    },
    proverAgentPollInterval: {
        env: 'PROVER_AGENT_POLL_INTERVAL_MS',
        description: 'The interval agents poll for jobs at',
        ...numberConfigHelper(100),
    },
    proverAgentConcurrency: {
        env: 'PROVER_AGENT_CONCURRENCY',
        description: 'The maximum number of proving jobs to be run in parallel',
        ...numberConfigHelper(1),
    },
    proverJobTimeoutMs: {
        env: 'PROVER_JOB_TIMEOUT_MS',
        description: 'Jobs are retried if not kept alive for this long',
        ...numberConfigHelper(60000),
    },
    proverJobPollIntervalMs: {
        env: 'PROVER_JOB_POLL_INTERVAL_MS',
        description: 'The interval to check job health status',
        ...numberConfigHelper(1000),
    },
    proverId: {
        env: 'PROVER_ID',
        parseEnv: (val) => parseProverId(val),
        description: 'Identifier of the prover',
    },
    proverTestDelayMs: {
        env: 'PROVER_TEST_DELAY_MS',
        description: 'Artificial delay to introduce to all operations to the test prover.',
        ...numberConfigHelper(0),
    },
};
function parseProverId(str) {
    return Fr.fromString(str.startsWith('0x') ? str : Buffer.from(str, 'utf8').toString('hex'));
}
export class BlockProofError extends Error {
    constructor(message, txHashes) {
        super(message);
        this.txHashes = txHashes;
        this.name = __classPrivateFieldGet(_a, _a, "f", _BlockProofError_name);
    }
    static isBlockProofError(err) {
        return err && typeof err === 'object' && err.name === __classPrivateFieldGet(_a, _a, "f", _BlockProofError_name);
    }
}
_a = BlockProofError;
_BlockProofError_name = { value: 'BlockProofError' };
