// @aztec/circuit-types/dest/interfaces/proving-job.js
export function makeProofAndVerificationKey(proof, verificationKey) {
    return { proof, verificationKey };
}
export function makePublicInputsAndRecursiveProof(inputs, proof, verificationKey) {
    const result = {
        inputs,
        proof,
        verificationKey,
    };
    return result;
}
export var ProvingRequestType;
(function (ProvingRequestType) {
    ProvingRequestType[ProvingRequestType["PRIVATE_KERNEL_EMPTY"] = 0] = "PRIVATE_KERNEL_EMPTY";
    ProvingRequestType[ProvingRequestType["PUBLIC_VM"] = 1] = "PUBLIC_VM";
    ProvingRequestType[ProvingRequestType["PUBLIC_KERNEL_INNER"] = 2] = "PUBLIC_KERNEL_INNER";
    ProvingRequestType[ProvingRequestType["PUBLIC_KERNEL_MERGE"] = 3] = "PUBLIC_KERNEL_MERGE";
    ProvingRequestType[ProvingRequestType["PUBLIC_KERNEL_TAIL"] = 4] = "PUBLIC_KERNEL_TAIL";
    ProvingRequestType[ProvingRequestType["BASE_ROLLUP"] = 5] = "BASE_ROLLUP";
    ProvingRequestType[ProvingRequestType["MERGE_ROLLUP"] = 6] = "MERGE_ROLLUP";
    ProvingRequestType[ProvingRequestType["EMPTY_BLOCK_ROOT_ROLLUP"] = 7] = "EMPTY_BLOCK_ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_ROOT_ROLLUP"] = 8] = "BLOCK_ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_MERGE_ROLLUP"] = 9] = "BLOCK_MERGE_ROLLUP";
    ProvingRequestType[ProvingRequestType["ROOT_ROLLUP"] = 10] = "ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["BASE_PARITY"] = 11] = "BASE_PARITY";
    ProvingRequestType[ProvingRequestType["ROOT_PARITY"] = 12] = "ROOT_PARITY";
    // Recursive Client IVC verification to connect private -> public or rollup
    ProvingRequestType[ProvingRequestType["TUBE_PROOF"] = 13] = "TUBE_PROOF";
})(ProvingRequestType || (ProvingRequestType = {}));
export function mapProvingRequestTypeToCircuitName(type) {
    switch (type) {
        case ProvingRequestType.PRIVATE_KERNEL_EMPTY:
            return 'private-kernel-empty';
        case ProvingRequestType.PUBLIC_VM:
            return 'avm-circuit';
        case ProvingRequestType.PUBLIC_KERNEL_INNER:
            return 'public-kernel-inner';
        case ProvingRequestType.PUBLIC_KERNEL_MERGE:
            return 'public-kernel-merge';
        case ProvingRequestType.PUBLIC_KERNEL_TAIL:
            return 'public-kernel-tail';
        case ProvingRequestType.BASE_ROLLUP:
            return 'base-rollup';
        case ProvingRequestType.MERGE_ROLLUP:
            return 'merge-rollup';
        case ProvingRequestType.EMPTY_BLOCK_ROOT_ROLLUP:
            return 'empty-block-root-rollup';
        case ProvingRequestType.BLOCK_ROOT_ROLLUP:
            return 'block-root-rollup';
        case ProvingRequestType.BLOCK_MERGE_ROLLUP:
            return 'block-merge-rollup';
        case ProvingRequestType.ROOT_ROLLUP:
            return 'root-rollup';
        case ProvingRequestType.BASE_PARITY:
            return 'base-parity';
        case ProvingRequestType.ROOT_PARITY:
            return 'root-parity';
        case ProvingRequestType.TUBE_PROOF:
            return 'tube-circuit';
        default:
            throw new Error(`Cannot find circuit name for proving request type: ${type}`);
    }
}
