// @aztec/circuits.js/dest/structs/read_request_hints/note_hash_read_request_hints.js
import { makeTuple } from '@aztec/foundation/array';
import { Fr } from '@aztec/foundation/fields';
import { MAX_NOTE_HASH_READ_REQUESTS_PER_TX, NOTE_HASH_TREE_HEIGHT } from '../../constants.gen.js';
import { PendingReadHint, ReadRequestResetHints, ReadRequestState, ReadRequestStatus, SettledReadHint, } from './read_request_hints.js';
export function noteHashReadRequestHintsFromBuffer(buffer, numPending, numSettled) {
    return ReadRequestResetHints.fromBuffer(buffer, MAX_NOTE_HASH_READ_REQUESTS_PER_TX, numPending, numSettled, NOTE_HASH_TREE_HEIGHT, Fr);
}
export class NoteHashReadRequestHintsBuilder {
    constructor(maxPending, maxSettled) {
        this.maxPending = maxPending;
        this.maxSettled = maxSettled;
        this.numPendingReadHints = 0;
        this.numSettledReadHints = 0;
        this.hints = new ReadRequestResetHints(makeTuple(MAX_NOTE_HASH_READ_REQUESTS_PER_TX, ReadRequestStatus.nada), makeTuple(maxPending, () => PendingReadHint.nada(MAX_NOTE_HASH_READ_REQUESTS_PER_TX)), makeTuple(maxSettled, () => SettledReadHint.nada(MAX_NOTE_HASH_READ_REQUESTS_PER_TX, NOTE_HASH_TREE_HEIGHT, Fr.zero)));
    }
    static empty(maxPending, maxSettled) {
        return new NoteHashReadRequestHintsBuilder(maxPending, maxSettled).toHints();
    }
    addPendingReadRequest(readRequestIndex, noteHashIndex) {
        if (this.numPendingReadHints === this.maxPending) {
            throw new Error('Cannot add more pending read request.');
        }
        this.hints.readRequestStatuses[readRequestIndex] = new ReadRequestStatus(ReadRequestState.PENDING, this.numPendingReadHints);
        this.hints.pendingReadHints[this.numPendingReadHints] = new PendingReadHint(readRequestIndex, noteHashIndex);
        this.numPendingReadHints++;
    }
    addSettledReadRequest(readRequestIndex, membershipWitness, value) {
        if (this.numSettledReadHints === this.maxSettled) {
            throw new Error('Cannot add more settled read request.');
        }
        this.hints.readRequestStatuses[readRequestIndex] = new ReadRequestStatus(ReadRequestState.SETTLED, this.numSettledReadHints);
        this.hints.settledReadHints[this.numSettledReadHints] = new SettledReadHint(readRequestIndex, membershipWitness, value);
        this.numSettledReadHints++;
    }
    toHints() {
        return this.hints;
    }
}
