// @aztec/aztec.js/dest/fee/private_fee_payment_method.js
import { computeSecretHash } from '@aztec/circuits.js/hash';
import { FunctionSelector, FunctionType } from '@aztec/foundation/abi';
import { Fr } from '@aztec/foundation/fields';
/**
 * Holds information about how the fee for a transaction is to be paid.
 */
export class PrivateFeePaymentMethod {
    constructor(
    /**
     * The asset used to pay the fee.
     */
    asset, 
    /**
     * Address which will hold the fee payment.
     */
    paymentContract, 
    /**
     * An auth witness provider to authorize fee payments
     */
    wallet, 
    /**
     * A secret to shield the rebate amount from the FPC.
     * Use this to claim the shielded amount to private balance
     */
    rebateSecret = Fr.random()) {
        this.asset = asset;
        this.paymentContract = paymentContract;
        this.wallet = wallet;
        this.rebateSecret = rebateSecret;
    }
    /**
     * The asset used to pay the fee.
     * @returns The asset used to pay the fee.
     */
    getAsset() {
        return this.asset;
    }
    getFeePayer() {
        return Promise.resolve(this.paymentContract);
    }
    /**
     * Creates a function call to pay the fee in the given asset.
     * @param gasSettings - The gas settings.
     * @returns The function call to pay the fee.
     */
    async getFunctionCalls(gasSettings) {
        const nonce = Fr.random();
        const maxFee = gasSettings.getFeeLimit();
        await this.wallet.createAuthWit({
            caller: this.paymentContract,
            action: {
                name: 'unshield',
                args: [this.wallet.getCompleteAddress().address, this.paymentContract, maxFee, nonce],
                selector: FunctionSelector.fromSignature('unshield((Field),(Field),Field,Field)'),
                type: FunctionType.PRIVATE,
                isStatic: false,
                to: this.asset,
                returnTypes: [],
            },
        });
        const secretHashForRebate = computeSecretHash(this.rebateSecret);
        return [
            {
                name: 'fee_entrypoint_private',
                to: this.paymentContract,
                selector: FunctionSelector.fromSignature('fee_entrypoint_private(Field,(Field),Field,Field)'),
                type: FunctionType.PRIVATE,
                isStatic: false,
                args: [maxFee, this.asset, secretHashForRebate, nonce],
                returnTypes: [],
            },
        ];
    }
}
