// @aztec/aztec.js/dest/contract/deploy_sent_tx.js
import { createDebugLogger } from '@aztec/foundation/log';
import { SentTx } from './sent_tx.js';
/**
 * A contract deployment transaction sent to the network, extending SentTx with methods to create a contract instance.
 */
export class DeploySentTx extends SentTx {
    constructor(wallet, txHashPromise, postDeployCtor, 
    /** The deployed contract instance */
    instance) {
        super(wallet, txHashPromise);
        this.postDeployCtor = postDeployCtor;
        this.instance = instance;
        this.log = createDebugLogger('aztec:js:deploy_sent_tx');
    }
    /**
     * Awaits for the tx to be mined and returns the contract instance. Throws if tx is not mined.
     * @param opts - Options for configuring the waiting for the tx to be mined.
     * @returns The deployed contract instance.
     */
    async deployed(opts) {
        const receipt = await this.wait(opts);
        this.log.info(`Contract ${this.instance.address.toString()} successfully deployed.`);
        return receipt.contract;
    }
    /**
     * Awaits for the tx to be mined and returns the receipt along with a contract instance. Throws if tx is not mined.
     * @param opts - Options for configuring the waiting for the tx to be mined.
     * @returns The transaction receipt with the deployed contract instance.
     */
    async wait(opts) {
        const receipt = await super.wait(opts);
        const contract = await this.getContractObject(opts?.wallet);
        return { ...receipt, contract };
    }
    getContractObject(wallet) {
        const isWallet = (pxe) => !!pxe.createTxExecutionRequest;
        const contractWallet = wallet ?? (isWallet(this.pxe) && this.pxe);
        if (!contractWallet) {
            throw new Error(`A wallet is required for creating a contract instance`);
        }
        return this.postDeployCtor(this.instance.address, contractWallet);
    }
}
