// @aztec/foundation/dest/promise/running-promise.js
import { InterruptibleSleep } from '../sleep/index.js';
import { promiseWithResolvers } from './utils.js';
/**
 * RunningPromise is a utility class that helps manage the execution of an asynchronous function
 * at a specified polling interval. It allows starting, stopping, and checking the status of the
 * internally managed promise. The class also supports interrupting the polling process when stopped.
 */
export class RunningPromise {
    constructor(fn, pollingIntervalMS = 10000) {
        this.fn = fn;
        this.pollingIntervalMS = pollingIntervalMS;
        this.running = false;
        this.runningPromise = Promise.resolve();
        this.interruptibleSleep = new InterruptibleSleep();
        this.requested = undefined;
    }
    /**
     * Starts the running promise.
     */
    start() {
        this.running = true;
        const poll = async () => {
            while (this.running) {
                const hasRequested = this.requested !== undefined;
                await this.fn();
                // If an immediate run had been requested *before* the function started running, resolve the request.
                if (hasRequested) {
                    this.requested.resolve();
                    this.requested = undefined;
                }
                // If no immediate run was requested, sleep for the polling interval.
                if (this.requested === undefined) {
                    await this.interruptibleSleep.sleep(this.pollingIntervalMS);
                }
            }
        };
        this.runningPromise = poll();
    }
    /**
     * Stops the running promise, resolves any pending interruptible sleep,
     * and waits for the currently executing function to complete.
     */
    async stop() {
        this.running = false;
        this.interruptibleSleep.interrupt();
        await this.runningPromise;
    }
    /**
     * Checks if the running promise is currently active.
     * @returns True if the promise is running.
     */
    isRunning() {
        return this.running;
    }
    /**
     * Triggers an immediate run of the function, bypassing the polling interval.
     * If the function is currently running, it will be allowed to continue and then called again immediately.
     */
    async trigger() {
        if (!this.running) {
            return this.fn();
        }
        let requested = this.requested;
        if (!requested) {
            requested = promiseWithResolvers();
            this.requested = requested;
            this.interruptibleSleep.interrupt();
        }
        await requested.promise;
    }
    /**
     * Updates the polling interval. The new interval will take effect after the next poll.
     * @param pollingIntervalMS The polling interval in milliseconds.
     */
    setPollingIntervalMS(pollingIntervalMS) {
        this.pollingIntervalMS = pollingIntervalMS;
    }
}
