// @aztec/foundation/dest/config/index.js
export function isBooleanConfigValue(obj, key) {
    return typeof obj[key] === 'boolean';
}
export function getConfigFromMappings(configMappings) {
    const config = {};
    for (const key in configMappings) {
        if (configMappings[key]) {
            const { env, parseEnv, defaultValue: def } = configMappings[key];
            // Special case for L1 contract addresses which is an object of config values
            if (key === 'l1Contracts' && def) {
                config[key] = getConfigFromMappings(def);
            }
            else {
                const val = env ? process.env[env] : undefined;
                if (val !== undefined) {
                    config[key] = parseEnv ? parseEnv(val) : val;
                }
                else if (def !== undefined) {
                    config[key] = def;
                }
            }
        }
    }
    return config;
}
/**
 * Filters out a service's config mappings to exclude certain keys.
 * @param configMappings - The service's config mappings
 * @param keysToFilter - The keys to filter out
 * @returns The filtered config mappings
 */
export function filterConfigMappings(configMappings, keysToFilter) {
    return Object.fromEntries(Object.entries(configMappings).filter(([key]) => !keysToFilter.includes(key)));
}
/**
 * Generates parseEnv and default values for a numerical config value.
 * @param defaultVal - The default numerical value to use if the environment variable is not set or is invalid
 * @returns Object with parseEnv and default values for a numerical config value
 */
export function numberConfigHelper(defaultVal) {
    return {
        parseEnv: (val) => safeParseNumber(val, defaultVal),
        defaultValue: defaultVal,
    };
}
/**
 * Generates parseEnv and default values for a numerical config value.
 * @param defaultVal - The default numerical value to use if the environment variable is not set or is invalid
 * @returns Object with parseEnv and default values for a numerical config value
 */
export function bigintConfigHelper(defaultVal) {
    return {
        parseEnv: (val) => BigInt(val),
        defaultValue: defaultVal,
    };
}
/**
 * Generates parseEnv for an optional numerical config value.
 */
export function optionalNumberConfigHelper() {
    return {
        parseEnv: (val) => {
            if (val !== undefined && val.length > 0) {
                const parsedValue = parseInt(val);
                return Number.isSafeInteger(parsedValue) ? parsedValue : undefined;
            }
            return undefined;
        },
    };
}
/**
 * Generates parseEnv and default values for a boolean config value.
 * @param defaultVal - The default value to use if the environment variable is not set or is invalid
 * @returns Object with parseEnv and default values for a boolean config value
 */
export function booleanConfigHelper(defaultVal = false) {
    const parse = (val) => (typeof val === 'boolean' ? val : ['1', 'true', 'TRUE'].includes(val));
    return {
        parseEnv: parse,
        parseVal: parse,
        defaultValue: defaultVal,
        isBoolean: true,
    };
}
/**
 * Safely parses a number from a string.
 * If the value is not a number or is not a safe integer, the default value is returned.
 * @param value - The string value to parse
 * @param defaultValue - The default value to return
 * @returns Either parsed value or default value
 */
function safeParseNumber(value, defaultValue) {
    const parsedValue = parseInt(value, 10);
    return Number.isSafeInteger(parsedValue) ? parsedValue : defaultValue;
}
/**
 * Picks specific keys from the given configuration mappings.
 *
 * @template T - The type of the full configuration object.
 * @template K - The keys to pick from the configuration object.
 * @param {ConfigMappingsType<T>} configMappings - The full configuration mappings object.
 * @param {K[]} keys - The keys to pick from the configuration mappings.
 * @returns {ConfigMappingsType<Pick<T, K>>} - A new configuration mappings object containing only the specified keys.
 */
export function pickConfigMappings(configMappings, keys) {
    return Object.fromEntries(keys.map(key => [key, configMappings[key]]));
}
/**
 * Extracts the default configuration values from the given configuration mappings.
 *
 * @template T - The type of the configuration object.
 * @param {ConfigMappingsType<T>} configMappings - The configuration mappings object.
 * @returns {T} - The configuration object with default values.
 */
export function getDefaultConfig(configMappings) {
    const defaultConfig = {};
    for (const key in configMappings) {
        if (configMappings[key] && configMappings[key].defaultValue !== undefined) {
            defaultConfig[key] = configMappings[key].defaultValue;
        }
    }
    return defaultConfig;
}
