// @aztec/foundation/dest/log/logger.js
import debug from 'debug';
import { inspect } from 'util';
const LogLevels = ['silent', 'error', 'warn', 'info', 'verbose', 'debug'];
function getLogLevel() {
    const envLogLevel = process.env.LOG_LEVEL?.toLowerCase();
    const defaultNonTestLogLevel = process.env.DEBUG === undefined ? 'info' : 'debug';
    const defaultLogLevel = process.env.NODE_ENV === 'test' ? 'silent' : defaultNonTestLogLevel;
    return LogLevels.includes(envLogLevel) ? envLogLevel : defaultLogLevel;
}
export let currentLevel = getLogLevel();
const namespaces = process.env.DEBUG ?? 'aztec:*';
debug.enable(namespaces);
/**
 * Creates a new DebugLogger for the current module, defaulting to the LOG_LEVEL env var.
 * If DEBUG="[module]" env is set, will enable debug logging if the module matches.
 * Uses npm debug for debug level and console.error for other levels.
 * @param name - Name of the module.
 * @param fixedLogData - Additional data to include in the log message.
 * @usage createDebugLogger('aztec:validator', {validatorAddress: '0x1234...'});
 * // will always add the validator address to the log labels
 * @returns A debug logger.
 */
export function createDebugLogger(name, fixedLogData) {
    const debugLogger = debug(name);
    const attatchFixedLogData = (data) => ({ ...fixedLogData, ...data });
    const logger = {
        silent: () => { },
        error: (msg, err, data) => logWithDebug(debugLogger, 'error', fmtErr(msg, err), attatchFixedLogData(data)),
        warn: (msg, data) => logWithDebug(debugLogger, 'warn', msg, attatchFixedLogData(data)),
        info: (msg, data) => logWithDebug(debugLogger, 'info', msg, attatchFixedLogData(data)),
        verbose: (msg, data) => logWithDebug(debugLogger, 'verbose', msg, attatchFixedLogData(data)),
        debug: (msg, data) => logWithDebug(debugLogger, 'debug', msg, attatchFixedLogData(data)),
    };
    return Object.assign((msg, data) => logWithDebug(debugLogger, 'debug', msg, attatchFixedLogData(data)), logger);
}
const logHandlers = [];
/**
 * Registers a callback for all logs, whether they are emitted in the current log level or not.
 * @param handler - Callback to be called on every log.
 */
export function onLog(handler) {
    logHandlers.push(handler);
}
/** Overrides current log level. */
export function setLevel(level) {
    currentLevel = level;
}
/**
 * Logs args to npm debug if enabled or log level is debug, console.error otherwise.
 * @param debug - Instance of npm debug.
 * @param level - Intended log level.
 * @param args - Args to log.
 */
function logWithDebug(debug, level, msg, data) {
    for (const handler of logHandlers) {
        handler(level, debug.namespace, msg, data);
    }
    msg = data ? `${msg} ${fmtLogData(data)}` : msg;
    if (debug.enabled && LogLevels.indexOf(level) <= LogLevels.indexOf(currentLevel)) {
        debug('[%s] %s', level.toUpperCase(), msg);
    }
}
/**
 * Concatenates a log message and an exception.
 * @param msg - Log message
 * @param err - Error to log
 * @returns A string with both the log message and the error message.
 */
function fmtErr(msg, err) {
    return err ? `${msg}: ${inspect(err)}` : msg;
}
/**
 * Formats structured log data as a string for console output.
 * @param data - Optional log data.
 */
export function fmtLogData(data) {
    return Object.entries(data ?? {})
        .map(([key, value]) => `${key}=${typeof value === 'object' && 'toString' in value ? value.toString() : value}`)
        .join(' ');
}
