// @aztec/circuits.js/dest/contract/contract_class.js
import { FunctionSelector, FunctionType } from '@aztec/foundation/abi';
import { Fr } from '@aztec/foundation/fields';
import { PUBLIC_DISPATCH_SELECTOR } from '../constants.gen.js';
import { computeArtifactHash } from './artifact_hash.js';
import { computeContractClassIdWithPreimage } from './contract_class_id.js';
const cmpFunctionArtifacts = (a, b) => a.selector.toField().cmp(b.selector.toField());
/** Creates a ContractClass from a contract compilation artifact. */
export function getContractClassFromArtifact(artifact) {
    const artifactHash = 'artifactHash' in artifact ? artifact.artifactHash : computeArtifactHash(artifact);
    const artifactPublicFunctions = artifact.functions
        .filter(f => f.functionType === FunctionType.PUBLIC)
        .map(f => ({
        selector: FunctionSelector.fromNameAndParameters(f.name, f.parameters),
        bytecode: f.bytecode,
    }))
        .sort(cmpFunctionArtifacts);
    let packedBytecode = Buffer.alloc(0);
    let dispatchFunction = undefined;
    if (artifactPublicFunctions.length > 0) {
        dispatchFunction = artifactPublicFunctions.find(f => f.selector.equals(FunctionSelector.fromField(new Fr(PUBLIC_DISPATCH_SELECTOR))));
        if (!dispatchFunction) {
            throw new Error('A contract with public functions should define a public_dispatch(Field) function as its public entrypoint.');
        }
        packedBytecode = dispatchFunction.bytecode;
    }
    const privateFunctions = artifact.functions
        .filter(f => f.functionType === FunctionType.PRIVATE)
        .map(getContractClassPrivateFunctionFromArtifact)
        .sort(cmpFunctionArtifacts);
    const contractClass = {
        version: 1,
        artifactHash,
        // TODO(https://github.com/AztecProtocol/aztec-packages/issues/8985): Remove public functions.
        publicFunctions: dispatchFunction ? [dispatchFunction] : [],
        packedBytecode,
        privateFunctions,
    };
    return { ...contractClass, ...computeContractClassIdWithPreimage(contractClass) };
}
export function getContractClassPrivateFunctionFromArtifact(f) {
    return {
        selector: FunctionSelector.fromNameAndParameters(f.name, f.parameters),
        vkHash: computeVerificationKeyHash(f.verificationKey),
    };
}
/**
 * Calculates the hash of a verification key.
 * Returns zero for consistency with Noir.
 */
export function computeVerificationKeyHash(_verificationKeyInBase64) {
    // return Fr.fromBuffer(hashVK(Buffer.from(verificationKeyInBase64, 'hex')));
    return Fr.ZERO;
}
