// @aztec/circuits.js/dest/contract/contract_class_id.js
import { poseidon2HashWithSeparator } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { GeneratorIndex } from '../constants.gen.js';
import { computePrivateFunctionsRoot } from './private_function.js';
/**
 * Returns the id of a contract class computed as its hash.
 *
 * ```
 * version = 1
 * private_function_leaves = private_functions.map(fn => pedersen([fn.function_selector as Field, fn.vk_hash], GENERATOR__FUNCTION_LEAF))
 * private_functions_root = merkleize(private_function_leaves)
 * bytecode_commitment = calculate_commitment(packed_bytecode)
 * contract_class_id = pedersen([version, artifact_hash, private_functions_root, bytecode_commitment], GENERATOR__CLASS_IDENTIFIER)
 * ```
 * @param contractClass - Contract class.
 * @returns The identifier.
 */
export function computeContractClassId(contractClass) {
    return computeContractClassIdWithPreimage(contractClass).id;
}
/** Computes a contract class id and returns it along with its preimage. */
export function computeContractClassIdWithPreimage(contractClass) {
    const artifactHash = contractClass.artifactHash;
    const privateFunctionsRoot = 'privateFunctionsRoot' in contractClass
        ? contractClass.privateFunctionsRoot
        : computePrivateFunctionsRoot(contractClass.privateFunctions);
    const publicBytecodeCommitment = 'publicBytecodeCommitment' in contractClass
        ? contractClass.publicBytecodeCommitment
        : computePublicBytecodeCommitment(contractClass.packedBytecode);
    const id = poseidon2HashWithSeparator([artifactHash, privateFunctionsRoot, publicBytecodeCommitment], GeneratorIndex.CONTRACT_LEAF);
    return { id, artifactHash, privateFunctionsRoot, publicBytecodeCommitment };
}
/** Returns the preimage of a contract class id given a contract class. */
export function computeContractClassIdPreimage(contractClass) {
    const privateFunctionsRoot = computePrivateFunctionsRoot(contractClass.privateFunctions);
    const publicBytecodeCommitment = computePublicBytecodeCommitment(contractClass.packedBytecode);
    return { artifactHash: contractClass.artifactHash, privateFunctionsRoot, publicBytecodeCommitment };
}
// TODO(#5860): Replace with actual implementation
// Changed to work with canonical contracts that may have non-deterministic noir compiles and we want to keep the address constant
export function computePublicBytecodeCommitment(_bytecode) {
    return new Fr(5);
}
