// @aztec/foundation/dest/buffer/buffer32.js
import { randomBytes } from '@aztec/foundation/crypto';
import { BufferReader, deserializeBigInt, serializeBigInt } from '@aztec/foundation/serialize';
/**
 * A class representing a 32 byte Buffer.
 */
export class Buffer32 {
    constructor(
    /**
     * The buffer containing the hash.
     */
    buffer) {
        this.buffer = buffer;
        if (buffer.length !== Buffer32.SIZE) {
            throw new Error(`Expected buffer to have length ${Buffer32.SIZE} but was ${buffer.length}`);
        }
    }
    /**
     * Returns the raw buffer of the hash.
     * @returns The buffer containing the hash.
     */
    toBuffer() {
        return this.buffer;
    }
    /**
     * Creates a Buffer32 from a buffer.
     * @param buffer - The buffer to create from.
     * @returns A new Buffer32 object.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new Buffer32(reader.readBytes(Buffer32.SIZE));
    }
    /**
     * Checks if this hash and another hash are equal.
     * @param hash - A hash to compare with.
     * @returns True if the hashes are equal, false otherwise.
     */
    equals(hash) {
        return this.buffer.equals(hash.buffer);
    }
    /**
     * Returns true if this hash is zero.
     * @returns True if this hash is zero.
     */
    isZero() {
        return this.buffer.equals(Buffer.alloc(32, 0));
    }
    /**
     * Convert this hash to a hex string.
     * @returns The hex string.
     */
    toString() {
        return this.buffer.toString('hex');
    }
    to0xString() {
        return `0x${this.buffer.toString('hex')}`;
    }
    /**
     * Convert this hash to a big int.
     * @returns The big int.
     */
    toBigInt() {
        return deserializeBigInt(this.buffer, 0, Buffer32.SIZE).elem;
    }
    /**
     * Creates a Buffer32 from a bigint.
     * @param hash - The tx hash as a big int.
     * @returns The Buffer32.
     */
    static fromBigInt(hash) {
        return new Buffer32(serializeBigInt(hash, Buffer32.SIZE));
    }
    static fromField(hash) {
        return new Buffer32(serializeBigInt(hash.toBigInt()));
    }
    /**
     * Converts this hash from a buffer of 28 bytes.
     * Verifies the input is 28 bytes.
     * @param buffer - The 28 byte buffer to construct from.
     * @returns A Buffer32 created from the input buffer with 4 bytes 0 padding at the front.
     */
    static fromBuffer28(buffer) {
        if (buffer.length != 28) {
            throw new Error(`Expected Buffer32 input buffer to be 28 bytes`);
        }
        const padded = Buffer.concat([Buffer.alloc(this.SIZE - 28), buffer]);
        return new Buffer32(padded);
    }
    /**
     * Converts a string into a Buffer32 object.
     * @param str - The TX hash in string format.
     * @returns A new Buffer32 object.
     */
    static fromStringUnchecked(str) {
        return new Buffer32(Buffer.from(str, 'hex'));
    }
    /**
     * Converts a string into a Buffer32 object.
     * NOTE: this method includes checks for the 0x prefix and the length of the string.
     *       if you dont need this checks, use fromStringUnchecked instead.
     *
     * @param str - The TX hash in string format.
     * @returns A new Buffer32 object.
     */
    static fromString(str) {
        if (str.startsWith('0x')) {
            str = str.slice(2);
        }
        if (str.length !== 64) {
            throw new Error(`Expected string to be 64 characters long, but was ${str.length}`);
        }
        return new Buffer32(Buffer.from(str, 'hex'));
    }
    /**
     * Converts a number into a Buffer32 object.
     * @param num - The number to convert.
     * @returns A new Buffer32 object.
     */
    static fromNumber(num) {
        return new Buffer32(serializeBigInt(BigInt(num), Buffer32.SIZE));
    }
    /**
     * Generates a random Buffer32.
     * @returns A new Buffer32 object.
     */
    static random() {
        return new Buffer32(Buffer.from(randomBytes(Buffer32.SIZE)));
    }
}
/**
 * The size of the hash in bytes.
 */
Buffer32.SIZE = 32;
/**
 * Buffer32 with value zero.
 */
Buffer32.ZERO = new Buffer32(Buffer.alloc(Buffer32.SIZE));
