// @aztec/aztec.js/dest/wallet/signerless_wallet.js
import { DefaultEntrypoint } from '../entrypoint/default_entrypoint.js';
import { BaseWallet } from './base_wallet.js';
/**
 * Wallet implementation which creates a transaction request directly to the requested contract without any signing.
 */
export class SignerlessWallet extends BaseWallet {
    constructor(pxe, entrypoint) {
        super(pxe);
        this.entrypoint = entrypoint;
    }
    async createTxExecutionRequest(execution) {
        let entrypoint = this.entrypoint;
        if (!entrypoint) {
            const { l1ChainId: chainId, protocolVersion } = await this.pxe.getNodeInfo();
            entrypoint = new DefaultEntrypoint(chainId, protocolVersion);
        }
        return entrypoint.createTxExecutionRequest(execution);
    }
    getChainId() {
        throw new Error('SignerlessWallet: Method getChainId not implemented.');
    }
    getVersion() {
        throw new Error('SignerlessWallet: Method getVersion not implemented.');
    }
    getPublicKeysHash() {
        throw new Error('SignerlessWallet: Method getPublicKeysHash not implemented.');
    }
    getCompleteAddress() {
        throw new Error('SignerlessWallet: Method getCompleteAddress not implemented.');
    }
    createAuthWit(_intent) {
        throw new Error('SignerlessWallet: Method createAuthWit not implemented.');
    }
}
