// @aztec/circuit-types/dest/tx/public_simulation_output.js
import { CombinedAccumulatedData, CombinedConstantData, Fr, Gas } from '@aztec/circuits.js';
import { mapValues } from '@aztec/foundation/collection';
import { EncryptedTxL2Logs, UnencryptedTxL2Logs } from '../logs/tx_l2_logs.js';
/** Return values of simulating complete callstack. */
export class NestedProcessReturnValues {
    constructor(values, nested) {
        this.values = values;
        this.nested = nested ?? [];
    }
    toJSON() {
        return {
            values: this.values?.map(fr => fr.toString()),
            nested: this.nested.map(n => n.toJSON()),
        };
    }
    static fromJSON(json) {
        return new NestedProcessReturnValues(json.values?.map(Fr.fromString), json.nested?.map((n) => NestedProcessReturnValues.fromJSON(n)));
    }
    static empty() {
        return new NestedProcessReturnValues([]);
    }
}
/**
 * Outputs of processing the public component of a transaction.
 */
export class PublicSimulationOutput {
    constructor(encryptedLogs, unencryptedLogs, revertReason, constants, end, publicReturnValues, gasUsed) {
        this.encryptedLogs = encryptedLogs;
        this.unencryptedLogs = unencryptedLogs;
        this.revertReason = revertReason;
        this.constants = constants;
        this.end = end;
        this.publicReturnValues = publicReturnValues;
        this.gasUsed = gasUsed;
    }
    toJSON() {
        return {
            encryptedLogs: this.encryptedLogs.toJSON(),
            unencryptedLogs: this.unencryptedLogs.toJSON(),
            revertReason: this.revertReason,
            constants: this.constants.toBuffer().toString('hex'),
            end: this.end.toBuffer().toString('hex'),
            publicReturnValues: this.publicReturnValues.map(returns => returns?.toJSON()),
            gasUsed: mapValues(this.gasUsed, gas => gas?.toJSON()),
        };
    }
    static fromJSON(json) {
        return new PublicSimulationOutput(EncryptedTxL2Logs.fromJSON(json.encryptedLogs), UnencryptedTxL2Logs.fromJSON(json.unencryptedLogs), json.revertReason, CombinedConstantData.fromBuffer(Buffer.from(json.constants, 'hex')), CombinedAccumulatedData.fromBuffer(Buffer.from(json.end, 'hex')), Array.isArray(json.publicReturnValues)
            ? json.publicReturnValues.map((returns) => NestedProcessReturnValues.fromJSON(returns))
            : [], mapValues(json.gasUsed, gas => (gas ? Gas.fromJSON(gas) : undefined)));
    }
}
