// @aztec/circuit-types/dest/tx/simulated_tx.js
import { ClientIvcProof, PrivateKernelTailCircuitPublicInputs } from '@aztec/circuits.js';
import { EncryptedNoteTxL2Logs, EncryptedTxL2Logs, UnencryptedTxL2Logs } from '../index.js';
import { PrivateExecutionResult, collectEnqueuedPublicFunctionCalls, collectPublicTeardownFunctionCall, collectSortedEncryptedLogs, collectSortedNoteEncryptedLogs, collectSortedUnencryptedLogs, } from '../private_execution_result.js';
import { NestedProcessReturnValues, PublicSimulationOutput } from './public_simulation_output.js';
import { Tx } from './tx.js';
export class PrivateSimulationResult {
    constructor(privateExecutionResult, publicInputs) {
        this.privateExecutionResult = privateExecutionResult;
        this.publicInputs = publicInputs;
    }
    getPrivateReturnValues() {
        return accumulatePrivateReturnValues(this.privateExecutionResult);
    }
    toSimulatedTx() {
        const noteEncryptedLogs = new EncryptedNoteTxL2Logs([collectSortedNoteEncryptedLogs(this.privateExecutionResult)]);
        const unencryptedLogs = new UnencryptedTxL2Logs([collectSortedUnencryptedLogs(this.privateExecutionResult)]);
        const encryptedLogs = new EncryptedTxL2Logs([collectSortedEncryptedLogs(this.privateExecutionResult)]);
        const enqueuedPublicFunctions = collectEnqueuedPublicFunctionCalls(this.privateExecutionResult);
        const teardownPublicFunction = collectPublicTeardownFunctionCall(this.privateExecutionResult);
        const tx = new Tx(this.publicInputs, ClientIvcProof.empty(), noteEncryptedLogs, encryptedLogs, unencryptedLogs, enqueuedPublicFunctions, teardownPublicFunction);
        return tx;
    }
    toJSON() {
        return {
            privateExecutionResult: this.privateExecutionResult.toJSON(),
            publicInputs: this.publicInputs.toBuffer().toString('hex'),
        };
    }
    static fromJSON(obj) {
        const privateExecutionResult = PrivateExecutionResult.fromJSON(obj.privateExecutionResult);
        const publicInputs = PrivateKernelTailCircuitPublicInputs.fromBuffer(Buffer.from(obj.publicInputs, 'hex'));
        return new PrivateSimulationResult(privateExecutionResult, publicInputs);
    }
}
export class TxSimulationResult extends PrivateSimulationResult {
    constructor(privateExecutionResult, publicInputs, publicOutput) {
        super(privateExecutionResult, publicInputs);
        this.publicOutput = publicOutput;
    }
    getPublicReturnValues() {
        return this.publicOutput ? this.publicOutput.publicReturnValues : [];
    }
    static fromPrivateSimulationResultAndPublicOutput(privateSimulationResult, publicOutput) {
        return new TxSimulationResult(privateSimulationResult.privateExecutionResult, privateSimulationResult.publicInputs, publicOutput);
    }
    toJSON() {
        return {
            privateExecutionResult: this.privateExecutionResult.toJSON(),
            publicInputs: this.publicInputs.toBuffer().toString('hex'),
            publicOutput: this.publicOutput ? this.publicOutput.toJSON() : undefined,
        };
    }
    static fromJSON(obj) {
        const privateExecutionResult = PrivateExecutionResult.fromJSON(obj.privateExecutionResult);
        const publicInputs = PrivateKernelTailCircuitPublicInputs.fromBuffer(Buffer.from(obj.publicInputs, 'hex'));
        const publicOuput = obj.publicOutput ? PublicSimulationOutput.fromJSON(obj.publicOutput) : undefined;
        return new TxSimulationResult(privateExecutionResult, publicInputs, publicOuput);
    }
}
export class TxProvingResult {
    constructor(privateExecutionResult, publicInputs, clientIvcProof) {
        this.privateExecutionResult = privateExecutionResult;
        this.publicInputs = publicInputs;
        this.clientIvcProof = clientIvcProof;
    }
    toTx() {
        const noteEncryptedLogs = new EncryptedNoteTxL2Logs([collectSortedNoteEncryptedLogs(this.privateExecutionResult)]);
        const unencryptedLogs = new UnencryptedTxL2Logs([collectSortedUnencryptedLogs(this.privateExecutionResult)]);
        const encryptedLogs = new EncryptedTxL2Logs([collectSortedEncryptedLogs(this.privateExecutionResult)]);
        const enqueuedPublicFunctions = collectEnqueuedPublicFunctionCalls(this.privateExecutionResult);
        const teardownPublicFunction = collectPublicTeardownFunctionCall(this.privateExecutionResult);
        const tx = new Tx(this.publicInputs, this.clientIvcProof, noteEncryptedLogs, encryptedLogs, unencryptedLogs, enqueuedPublicFunctions, teardownPublicFunction);
        return tx;
    }
    toJSON() {
        return {
            privateExecutionResult: this.privateExecutionResult.toJSON(),
            publicInputs: this.publicInputs.toBuffer().toString('hex'),
            clientIvcProof: this.clientIvcProof.toBuffer().toString('hex'),
        };
    }
    static fromJSON(obj) {
        const privateExecutionResult = PrivateExecutionResult.fromJSON(obj.privateExecutionResult);
        const publicInputs = PrivateKernelTailCircuitPublicInputs.fromBuffer(Buffer.from(obj.publicInputs, 'hex'));
        const clientIvcProof = ClientIvcProof.fromBuffer(Buffer.from(obj.clientIvcProof, 'hex'));
        return new TxProvingResult(privateExecutionResult, publicInputs, clientIvcProof);
    }
}
/**
 * Recursively accummulate the return values of a call result and its nested executions,
 * so they can be retrieved in order.
 * @param executionResult
 * @returns
 */
export function accumulatePrivateReturnValues(executionResult) {
    const acc = new NestedProcessReturnValues(executionResult.returnValues);
    acc.nested = executionResult.nestedExecutions.map(nestedExecution => accumulatePrivateReturnValues(nestedExecution));
    return acc;
}
