// @aztec/circuit-types/dest/tx/tx_receipt.js
import { RevertCode } from '@aztec/circuits.js';
import { TxHash } from './tx_hash.js';
/**
 * Possible status of a transaction.
 */
export var TxStatus;
(function (TxStatus) {
    TxStatus["DROPPED"] = "dropped";
    TxStatus["PENDING"] = "pending";
    TxStatus["SUCCESS"] = "success";
    TxStatus["APP_LOGIC_REVERTED"] = "app_logic_reverted";
    TxStatus["TEARDOWN_REVERTED"] = "teardown_reverted";
    TxStatus["BOTH_REVERTED"] = "both_reverted";
})(TxStatus || (TxStatus = {}));
/**
 * Represents a transaction receipt in the Aztec network.
 * Contains essential information about the transaction including its status, origin, and associated addresses.
 * REFACTOR: TxReceipt should be returned only once the tx is mined, and all its fields should be required.
 * We should not be using a TxReceipt to answer a query for a pending or dropped tx.
 */
export class TxReceipt {
    constructor(
    /**
     * A unique identifier for a transaction.
     */
    txHash, 
    /**
     * The transaction's status.
     */
    status, 
    /**
     * Description of transaction error, if any.
     */
    error, 
    /**
     * The transaction fee paid for the transaction.
     */
    transactionFee, 
    /**
     * The hash of the block containing the transaction.
     */
    blockHash, 
    /**
     * The block number in which the transaction was included.
     */
    blockNumber, 
    /**
     * Information useful for testing/debugging, set when test flag is set to true in `waitOpts`.
     */
    debugInfo) {
        this.txHash = txHash;
        this.status = status;
        this.error = error;
        this.transactionFee = transactionFee;
        this.blockHash = blockHash;
        this.blockNumber = blockNumber;
        this.debugInfo = debugInfo;
    }
    /**
     * Convert a Tx class object to a plain JSON object.
     * @returns A plain object with Tx properties.
     */
    toJSON() {
        return {
            txHash: this.txHash.toString(),
            status: this.status.toString(),
            error: this.error,
            blockHash: this.blockHash?.toString('hex'),
            blockNumber: this.blockNumber,
            transactionFee: this.transactionFee?.toString(),
        };
    }
    /**
     * Convert a plain JSON object to a Tx class object.
     * @param obj - A plain Tx JSON object.
     * @returns A Tx class object.
     */
    static fromJSON(obj) {
        const txHash = TxHash.fromString(obj.txHash);
        const status = obj.status;
        const error = obj.error;
        const transactionFee = obj.transactionFee ? BigInt(obj.transactionFee) : undefined;
        const blockHash = obj.blockHash ? Buffer.from(obj.blockHash, 'hex') : undefined;
        const blockNumber = obj.blockNumber ? Number(obj.blockNumber) : undefined;
        return new TxReceipt(txHash, status, error, transactionFee, blockHash, blockNumber);
    }
    static statusFromRevertCode(revertCode) {
        if (revertCode.equals(RevertCode.OK)) {
            return TxStatus.SUCCESS;
        }
        else if (revertCode.equals(RevertCode.APP_LOGIC_REVERTED)) {
            return TxStatus.APP_LOGIC_REVERTED;
        }
        else if (revertCode.equals(RevertCode.TEARDOWN_REVERTED)) {
            return TxStatus.TEARDOWN_REVERTED;
        }
        else if (revertCode.equals(RevertCode.BOTH_REVERTED)) {
            return TxStatus.BOTH_REVERTED;
        }
        else {
            throw new Error(`Unknown revert code: ${revertCode}`);
        }
    }
}
