// @aztec/protocol-contracts/dest/protocol_contract_data.js
// GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:data`
import { AztecAddress, Fr } from '@aztec/circuits.js';
import { loadContractArtifact } from '@aztec/types/abi';
import AuthRegistryJson from '../artifacts/AuthRegistry.json' assert { type: 'json' };
import ContractClassRegistererJson from '../artifacts/ContractClassRegisterer.json' assert { type: 'json' };
import ContractInstanceDeployerJson from '../artifacts/ContractInstanceDeployer.json' assert { type: 'json' };
import FeeJuiceJson from '../artifacts/FeeJuice.json' assert { type: 'json' };
import MultiCallEntrypointJson from '../artifacts/MultiCallEntrypoint.json' assert { type: 'json' };
import RouterJson from '../artifacts/Router.json' assert { type: 'json' };
export const protocolContractNames = [
    'AuthRegistry',
    'ContractInstanceDeployer',
    'ContractClassRegisterer',
    'MultiCallEntrypoint',
    'FeeJuice',
    'Router',
];
export const ProtocolContractArtifact = {
    AuthRegistry: loadContractArtifact(AuthRegistryJson),
    ContractInstanceDeployer: loadContractArtifact(ContractInstanceDeployerJson),
    ContractClassRegisterer: loadContractArtifact(ContractClassRegistererJson),
    MultiCallEntrypoint: loadContractArtifact(MultiCallEntrypointJson),
    FeeJuice: loadContractArtifact(FeeJuiceJson),
    Router: loadContractArtifact(RouterJson),
};
export const ProtocolContractSalt = {
    AuthRegistry: new Fr(1),
    ContractInstanceDeployer: new Fr(1),
    ContractClassRegisterer: new Fr(1),
    MultiCallEntrypoint: new Fr(1),
    FeeJuice: new Fr(1),
    Router: new Fr(1),
};
export const ProtocolContractAddress = {
    AuthRegistry: AztecAddress.fromBigInt(1n),
    ContractInstanceDeployer: AztecAddress.fromBigInt(2n),
    ContractClassRegisterer: AztecAddress.fromBigInt(3n),
    MultiCallEntrypoint: AztecAddress.fromBigInt(4n),
    FeeJuice: AztecAddress.fromBigInt(5n),
    Router: AztecAddress.fromBigInt(6n),
};
export const ProtocolContractLeaf = {
    AuthRegistry: Fr.fromString('0x0d7c69bbd50f9da911b5ab26e313dafc7f66b0e03a52d6d98270c54ed751c4e2'),
    ContractInstanceDeployer: Fr.fromString('0x0f91b27c8187748d9710cc579b3fc08b56e5a391b9f7818b71feadd39816bc6d'),
    ContractClassRegisterer: Fr.fromString('0x17f57693577de778fa6f4d83c280597ac659acfbce64175aacc390d289f97953'),
    MultiCallEntrypoint: Fr.fromString('0x054b0c9a77a2fcaf1c3a1519205a21d4627cb950245a462d6a9bfc52892d6b7a'),
    FeeJuice: Fr.fromString('0x03c51587dc12449bd85f33627111224910a934b3c79fcf720fd1c6ccac0f34e1'),
    Router: Fr.fromString('0x250937e9172b417f70fb2d0bdefdf69268e1134f1612418cf58fb80ff4278fe9'),
};
export const protocolContractTreeRoot = Fr.fromString('0x1010ec5ed35322c9adf0ef6573d9f4bdb43bb971ddfe8dfa7776c5d4254eb2ab');
