// @aztec/circuit-types/dest/body.js
import { EncryptedL2BlockL2Logs, EncryptedNoteL2BlockL2Logs, TxEffect, UnencryptedL2BlockL2Logs, } from '@aztec/circuit-types';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { computeUnbalancedMerkleRoot } from '@aztec/foundation/trees';
import { inspect } from 'util';
export class Body {
    constructor(txEffects) {
        this.txEffects = txEffects;
        txEffects.forEach(txEffect => {
            if (txEffect.isEmpty()) {
                throw new Error('Empty tx effect not allowed in Body');
            }
        });
    }
    /**
     * Serializes a block body
     * @returns A serialized L2 block body.
     */
    toBuffer() {
        return serializeToBuffer(this.txEffects.length, this.txEffects);
    }
    /**
     * Deserializes a block from a buffer
     * @returns A deserialized L2 block.
     */
    static fromBuffer(buf) {
        const reader = BufferReader.asReader(buf);
        return new this(reader.readVector(TxEffect));
    }
    [inspect.custom]() {
        return `Body {
  txEffects: ${inspect(this.txEffects)},
  emptyTxEffectsCount: ${this.numberOfTxsIncludingPadded},
  emptyTxEffectHash: ${TxEffect.empty().hash().toString('hex')},
  txsEffectsHash: ${this.getTxsEffectsHash().toString('hex')},
}`;
    }
    /**
     * Computes the transactions effects hash for the L2 block
     * This hash is also computed in the `TxDecoder`.
     * @returns The txs effects hash.
     */
    getTxsEffectsHash() {
        const emptyTxEffectHash = TxEffect.empty().hash();
        const leaves = this.txEffects.map(txEffect => txEffect.hash());
        return computeUnbalancedMerkleRoot(leaves, emptyTxEffectHash);
    }
    get noteEncryptedLogs() {
        const logs = this.txEffects.map(txEffect => txEffect.noteEncryptedLogs);
        return new EncryptedNoteL2BlockL2Logs(logs);
    }
    get encryptedLogs() {
        const logs = this.txEffects.map(txEffect => txEffect.encryptedLogs);
        return new EncryptedL2BlockL2Logs(logs);
    }
    get unencryptedLogs() {
        const logs = this.txEffects.map(txEffect => txEffect.unencryptedLogs);
        return new UnencryptedL2BlockL2Logs(logs);
    }
    /**
     * Computes the number of transactions in the block including padding transactions.
     * @dev Modified code from TxsDecoder.computeNumTxEffectsToPad
     */
    get numberOfTxsIncludingPadded() {
        const numTxEffects = this.txEffects.length;
        // 2 is the minimum number of tx effects
        if (numTxEffects <= 2) {
            return 2;
        }
        return numTxEffects;
    }
    static random(txsPerBlock = 4, numPrivateCallsPerTx = 2, numPublicCallsPerTx = 3, numEncryptedLogsPerCall = 2, numUnencryptedLogsPerCall = 1) {
        const txEffects = [...new Array(txsPerBlock)].map(_ => TxEffect.random(numPrivateCallsPerTx, numPublicCallsPerTx, numEncryptedLogsPerCall, numUnencryptedLogsPerCall));
        return new Body(txEffects);
    }
    static empty() {
        return new Body([]);
    }
}
