// @aztec/circuit-types/dest/public_execution_request.js
import { CallContext, Vector } from '@aztec/circuits.js';
import { computeVarArgsHash } from '@aztec/circuits.js/hash';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
/**
 * The execution request of a public function.
 */
export class PublicExecutionRequest {
    constructor(
    /**
     * Context of the public call.
     */
    callContext, 
    /**
     * Function arguments.
     */
    args) {
        this.callContext = callContext;
        this.args = args;
    }
    getSize() {
        return this.isEmpty() ? 0 : this.toBuffer().length;
    }
    toBuffer() {
        return serializeToBuffer(this.callContext, new Vector(this.args));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicExecutionRequest(CallContext.fromBuffer(reader), reader.readVector(Fr));
    }
    static from(fields) {
        return new PublicExecutionRequest(...PublicExecutionRequest.getFields(fields));
    }
    static getFields(fields) {
        return [fields.callContext, fields.args];
    }
    static empty() {
        return new PublicExecutionRequest(CallContext.empty(), []);
    }
    isEmpty() {
        return this.callContext.isEmpty() && this.args.length === 0;
    }
    isForCallRequest(callRequest) {
        return (this.callContext.equals(callRequest.callContext) && computeVarArgsHash(this.args).equals(callRequest.argsHash));
    }
    [inspect.custom]() {
        return `PublicExecutionRequest {
      callContext: ${this.callContext}
      args: ${this.args}
    }`;
    }
}
