// @aztec/noir-contracts.js/src/AppSubscription.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import AppSubscriptionContractArtifactJson from '../artifacts/app_subscription_contract-AppSubscription.json' assert { type: 'json' };

export const AppSubscriptionContractArtifact = loadContractArtifact(
  AppSubscriptionContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract AppSubscription;
 */
export class AppSubscriptionContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, AppSubscriptionContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, AppSubscriptionContract.artifact, wallet) as Promise<AppSubscriptionContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(
    wallet: Wallet,
    target_address: AztecAddressLike,
    subscription_recipient_address: AztecAddressLike,
    subscription_token_address: AztecAddressLike,
    subscription_price: FieldLike,
    fee_juice_limit_per_tx: FieldLike,
  ) {
    return new DeployMethod<AppSubscriptionContract>(
      PublicKeys.default(),
      wallet,
      AppSubscriptionContractArtifact,
      AppSubscriptionContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(
    publicKeys: PublicKeys,
    wallet: Wallet,
    target_address: AztecAddressLike,
    subscription_recipient_address: AztecAddressLike,
    subscription_token_address: AztecAddressLike,
    subscription_price: FieldLike,
    fee_juice_limit_per_tx: FieldLike,
  ) {
    return new DeployMethod<AppSubscriptionContract>(
      publicKeys,
      wallet,
      AppSubscriptionContractArtifact,
      AppSubscriptionContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AppSubscriptionContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<AppSubscriptionContract['methods'][M]>
  ) {
    return new DeployMethod<AppSubscriptionContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      AppSubscriptionContractArtifact,
      AppSubscriptionContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AppSubscriptionContractArtifact;
  }

  public static get storage(): ContractStorageLayout<
    | 'target_address'
    | 'subscription_token_address'
    | 'subscription_recipient_address'
    | 'subscription_price'
    | 'subscriptions'
    | 'fee_juice_limit_per_tx'
  > {
    return {
      target_address: {
        slot: new Fr(1n),
      },
      subscription_token_address: {
        slot: new Fr(2n),
      },
      subscription_recipient_address: {
        slot: new Fr(3n),
      },
      subscription_price: {
        slot: new Fr(4n),
      },
      subscriptions: {
        slot: new Fr(5n),
      },
      fee_juice_limit_per_tx: {
        slot: new Fr(6n),
      },
    } as ContractStorageLayout<
      | 'target_address'
      | 'subscription_token_address'
      | 'subscription_recipient_address'
      | 'subscription_price'
      | 'subscriptions'
      | 'fee_juice_limit_per_tx'
    >;
  }

  public static get notes(): ContractNotes<'TransparentNote' | 'TokenNote' | 'SubscriptionNote'> {
    return {
      TransparentNote: {
        id: new NoteSelector(3193649735),
      },
      TokenNote: {
        id: new NoteSelector(2350566847),
      },
      SubscriptionNote: {
        id: new NoteSelector(3064460783),
      },
    } as ContractNotes<'TransparentNote' | 'TokenNote' | 'SubscriptionNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(target_address: struct, subscription_recipient_address: struct, subscription_token_address: struct, subscription_price: field, fee_juice_limit_per_tx: field) */
    constructor: ((
      target_address: AztecAddressLike,
      subscription_recipient_address: AztecAddressLike,
      subscription_token_address: AztecAddressLike,
      subscription_price: FieldLike,
      fee_juice_limit_per_tx: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** entrypoint(payload: struct, user_address: struct) */
    entrypoint: ((
      payload: {
        function_calls: {
          args_hash: FieldLike;
          function_selector: FunctionSelectorLike;
          target_address: AztecAddressLike;
          is_public: boolean;
          is_static: boolean;
        }[];
        nonce: FieldLike;
      },
      user_address: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** is_initialized(subscriber: struct) */
    is_initialized: ((subscriber: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** subscribe(subscriber: struct, nonce: field, expiry_block_number: field, tx_count: field) */
    subscribe: ((
      subscriber: AztecAddressLike,
      nonce: FieldLike,
      expiry_block_number: FieldLike,
      tx_count: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
