// @aztec/noir-contracts.js/src/AvmInitializerTest.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import AvmInitializerTestContractArtifactJson from '../artifacts/avm_initializer_test_contract-AvmInitializerTest.json' assert { type: 'json' };

export const AvmInitializerTestContractArtifact = loadContractArtifact(
  AvmInitializerTestContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract AvmInitializerTest;
 */
export class AvmInitializerTestContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, AvmInitializerTestContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, AvmInitializerTestContract.artifact, wallet) as Promise<AvmInitializerTestContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<AvmInitializerTestContract>(
      PublicKeys.default(),
      wallet,
      AvmInitializerTestContractArtifact,
      AvmInitializerTestContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet) {
    return new DeployMethod<AvmInitializerTestContract>(
      publicKeys,
      wallet,
      AvmInitializerTestContractArtifact,
      AvmInitializerTestContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AvmInitializerTestContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<AvmInitializerTestContract['methods'][M]>
  ) {
    return new DeployMethod<AvmInitializerTestContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      AvmInitializerTestContractArtifact,
      AvmInitializerTestContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AvmInitializerTestContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'immutable'> {
    return {
      immutable: {
        slot: new Fr(1n),
      },
    } as ContractStorageLayout<'immutable'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor() */
    constructor: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** read_storage_immutable() */
    read_storage_immutable: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
