// @aztec/noir-contracts.js/src/Claim.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import ClaimContractArtifactJson from '../artifacts/claim_contract-Claim.json' assert { type: 'json' };

export const ClaimContractArtifact = loadContractArtifact(ClaimContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract Claim;
 */
export class ClaimContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, ClaimContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, ClaimContract.artifact, wallet) as Promise<ClaimContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, target_contract: AztecAddressLike, reward_token: AztecAddressLike) {
    return new DeployMethod<ClaimContract>(
      PublicKeys.default(),
      wallet,
      ClaimContractArtifact,
      ClaimContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(
    publicKeys: PublicKeys,
    wallet: Wallet,
    target_contract: AztecAddressLike,
    reward_token: AztecAddressLike,
  ) {
    return new DeployMethod<ClaimContract>(
      publicKeys,
      wallet,
      ClaimContractArtifact,
      ClaimContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof ClaimContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<ClaimContract['methods'][M]>
  ) {
    return new DeployMethod<ClaimContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      ClaimContractArtifact,
      ClaimContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ClaimContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'target_contract' | 'reward_token'> {
    return {
      target_contract: {
        slot: new Fr(1n),
      },
      reward_token: {
        slot: new Fr(2n),
      },
    } as ContractStorageLayout<'target_contract' | 'reward_token'>;
  }

  public static get notes(): ContractNotes<'TransparentNote' | 'TokenNote' | 'ValueNote'> {
    return {
      TransparentNote: {
        id: new NoteSelector(3193649735),
      },
      TokenNote: {
        id: new NoteSelector(2350566847),
      },
      ValueNote: {
        id: new NoteSelector(1038582377),
      },
    } as ContractNotes<'TransparentNote' | 'TokenNote' | 'ValueNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** claim(proof_note: struct, recipient: struct) */
    claim: ((
      proof_note: {
        value: FieldLike;
        npk_m_hash: FieldLike;
        randomness: FieldLike;
        header: {
          contract_address: AztecAddressLike;
          nonce: FieldLike;
          storage_slot: FieldLike;
          note_hash_counter: bigint | number;
        };
      },
      recipient: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(target_contract: struct, reward_token: struct) */
    constructor: ((target_contract: AztecAddressLike, reward_token: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
