// @aztec/noir-contracts.js/src/ContractInstanceDeployer.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import ContractInstanceDeployerContractArtifactJson from '../artifacts/contract_instance_deployer_contract-ContractInstanceDeployer.json' assert { type: 'json' };

export const ContractInstanceDeployerContractArtifact = loadContractArtifact(
  ContractInstanceDeployerContractArtifactJson as NoirCompiledContract,
);

export type ContractInstanceDeployed = {
  DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE: FieldLike;
  address: AztecAddressLike;
  version: bigint | number;
  salt: FieldLike;
  contract_class_id: WrappedFieldLike;
  initialization_hash: FieldLike;
  public_keys: {
    npk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
    ivpk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
    ovpk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
    tpk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
  };
  deployer: AztecAddressLike;
};

/**
 * Type-safe interface for contract ContractInstanceDeployer;
 */
export class ContractInstanceDeployerContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, ContractInstanceDeployerContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(
      address,
      ContractInstanceDeployerContract.artifact,
      wallet,
    ) as Promise<ContractInstanceDeployerContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<ContractInstanceDeployerContract>(
      PublicKeys.default(),
      wallet,
      ContractInstanceDeployerContractArtifact,
      ContractInstanceDeployerContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet) {
    return new DeployMethod<ContractInstanceDeployerContract>(
      publicKeys,
      wallet,
      ContractInstanceDeployerContractArtifact,
      ContractInstanceDeployerContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof ContractInstanceDeployerContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<ContractInstanceDeployerContract['methods'][M]>
  ) {
    return new DeployMethod<ContractInstanceDeployerContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      ContractInstanceDeployerContractArtifact,
      ContractInstanceDeployerContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ContractInstanceDeployerContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** deploy(salt: field, contract_class_id: struct, initialization_hash: field, public_keys: struct, universal_deploy: boolean) */
    deploy: ((
      salt: FieldLike,
      contract_class_id: WrappedFieldLike,
      initialization_hash: FieldLike,
      public_keys: {
        npk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
        ivpk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
        ovpk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
        tpk_m: { inner: { x: FieldLike; y: FieldLike; is_infinite: boolean } };
      },
      universal_deploy: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };

  // Partial application is chosen is to avoid the duplication of so much codegen.
  private static decodeEvent<T>(
    eventSelector: EventSelector,
    eventType: AbiType,
  ): (payload: L1EventPayload | UnencryptedL2Log | undefined) => T | undefined {
    return (payload: L1EventPayload | UnencryptedL2Log | undefined): T | undefined => {
      if (payload === undefined) {
        return undefined;
      }

      if (payload instanceof L1EventPayload) {
        if (!eventSelector.equals(payload.eventTypeId)) {
          return undefined;
        }
        return decodeFromAbi([eventType], payload.event.items) as T;
      } else {
        let items = [];
        for (let i = 0; i < payload.data.length; i += 32) {
          items.push(new Fr(payload.data.subarray(i, i + 32)));
        }

        return decodeFromAbi([eventType], items) as T;
      }
    };
  }

  public static get events(): {
    ContractInstanceDeployed: {
      decode: (payload: L1EventPayload | UnencryptedL2Log | undefined) => ContractInstanceDeployed | undefined;
      eventSelector: EventSelector;
      fieldNames: string[];
    };
  } {
    return {
      ContractInstanceDeployed: {
        decode: this.decodeEvent(
          EventSelector.fromSignature(
            'ContractInstanceDeployed(Field,(Field),u8,Field,(Field),Field,(((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool))),(Field))',
          ),
          {
            fields: [
              {
                name: 'DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE',
                type: {
                  kind: 'field',
                },
              },
              {
                name: 'address',
                type: {
                  fields: [
                    {
                      name: 'inner',
                      type: {
                        kind: 'field',
                      },
                    },
                  ],
                  kind: 'struct',
                  path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
                },
              },
              {
                name: 'version',
                type: {
                  kind: 'integer',
                  sign: 'unsigned',
                  width: 8,
                },
              },
              {
                name: 'salt',
                type: {
                  kind: 'field',
                },
              },
              {
                name: 'contract_class_id',
                type: {
                  fields: [
                    {
                      name: 'inner',
                      type: {
                        kind: 'field',
                      },
                    },
                  ],
                  kind: 'struct',
                  path: 'aztec::protocol_types::contract_class_id::ContractClassId',
                },
              },
              {
                name: 'initialization_hash',
                type: {
                  kind: 'field',
                },
              },
              {
                name: 'public_keys',
                type: {
                  fields: [
                    {
                      name: 'npk_m',
                      type: {
                        fields: [
                          {
                            name: 'inner',
                            type: {
                              fields: [
                                {
                                  name: 'x',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'y',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'is_infinite',
                                  type: {
                                    kind: 'boolean',
                                  },
                                },
                              ],
                              kind: 'struct',
                              path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                            },
                          },
                        ],
                        kind: 'struct',
                        path: 'aztec::protocol_types::public_keys::NpkM',
                      },
                    },
                    {
                      name: 'ivpk_m',
                      type: {
                        fields: [
                          {
                            name: 'inner',
                            type: {
                              fields: [
                                {
                                  name: 'x',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'y',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'is_infinite',
                                  type: {
                                    kind: 'boolean',
                                  },
                                },
                              ],
                              kind: 'struct',
                              path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                            },
                          },
                        ],
                        kind: 'struct',
                        path: 'aztec::protocol_types::public_keys::IvpkM',
                      },
                    },
                    {
                      name: 'ovpk_m',
                      type: {
                        fields: [
                          {
                            name: 'inner',
                            type: {
                              fields: [
                                {
                                  name: 'x',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'y',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'is_infinite',
                                  type: {
                                    kind: 'boolean',
                                  },
                                },
                              ],
                              kind: 'struct',
                              path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                            },
                          },
                        ],
                        kind: 'struct',
                        path: 'aztec::protocol_types::public_keys::OvpkM',
                      },
                    },
                    {
                      name: 'tpk_m',
                      type: {
                        fields: [
                          {
                            name: 'inner',
                            type: {
                              fields: [
                                {
                                  name: 'x',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'y',
                                  type: {
                                    kind: 'field',
                                  },
                                },
                                {
                                  name: 'is_infinite',
                                  type: {
                                    kind: 'boolean',
                                  },
                                },
                              ],
                              kind: 'struct',
                              path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                            },
                          },
                        ],
                        kind: 'struct',
                        path: 'aztec::protocol_types::public_keys::TpkM',
                      },
                    },
                  ],
                  kind: 'struct',
                  path: 'aztec::protocol_types::public_keys::PublicKeys',
                },
              },
              {
                name: 'deployer',
                type: {
                  fields: [
                    {
                      name: 'inner',
                      type: {
                        kind: 'field',
                      },
                    },
                  ],
                  kind: 'struct',
                  path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
                },
              },
            ],
            kind: 'struct',
            path: 'ContractInstanceDeployer::ContractInstanceDeployed',
          },
        ),
        eventSelector: EventSelector.fromSignature(
          'ContractInstanceDeployed(Field,(Field),u8,Field,(Field),Field,(((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool))),(Field))',
        ),
        fieldNames: [
          'DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE',
          'address',
          'version',
          'salt',
          'contract_class_id',
          'initialization_hash',
          'public_keys',
          'deployer',
        ],
      },
    };
  }
}
