// @aztec/noir-contracts.js/src/EcdsaRAccount.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import EcdsaRAccountContractArtifactJson from '../artifacts/ecdsa_r_account_contract-EcdsaRAccount.json' assert { type: 'json' };

export const EcdsaRAccountContractArtifact = loadContractArtifact(
  EcdsaRAccountContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract EcdsaRAccount;
 */
export class EcdsaRAccountContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, EcdsaRAccountContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, EcdsaRAccountContract.artifact, wallet) as Promise<EcdsaRAccountContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, signing_pub_key_x: (bigint | number)[], signing_pub_key_y: (bigint | number)[]) {
    return new DeployMethod<EcdsaRAccountContract>(
      PublicKeys.default(),
      wallet,
      EcdsaRAccountContractArtifact,
      EcdsaRAccountContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(
    publicKeys: PublicKeys,
    wallet: Wallet,
    signing_pub_key_x: (bigint | number)[],
    signing_pub_key_y: (bigint | number)[],
  ) {
    return new DeployMethod<EcdsaRAccountContract>(
      publicKeys,
      wallet,
      EcdsaRAccountContractArtifact,
      EcdsaRAccountContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof EcdsaRAccountContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<EcdsaRAccountContract['methods'][M]>
  ) {
    return new DeployMethod<EcdsaRAccountContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      EcdsaRAccountContractArtifact,
      EcdsaRAccountContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return EcdsaRAccountContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'public_key'> {
    return {
      public_key: {
        slot: new Fr(1n),
      },
    } as ContractStorageLayout<'public_key'>;
  }

  public static get notes(): ContractNotes<'EcdsaPublicKeyNote'> {
    return {
      EcdsaPublicKeyNote: {
        id: new NoteSelector(79774112),
      },
    } as ContractNotes<'EcdsaPublicKeyNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(signing_pub_key_x: array, signing_pub_key_y: array) */
    constructor: ((
      signing_pub_key_x: (bigint | number)[],
      signing_pub_key_y: (bigint | number)[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** entrypoint(app_payload: struct, fee_payload: struct, cancellable: boolean) */
    entrypoint: ((
      app_payload: {
        function_calls: {
          args_hash: FieldLike;
          function_selector: FunctionSelectorLike;
          target_address: AztecAddressLike;
          is_public: boolean;
          is_static: boolean;
        }[];
        nonce: FieldLike;
      },
      fee_payload: {
        function_calls: {
          args_hash: FieldLike;
          function_selector: FunctionSelectorLike;
          target_address: AztecAddressLike;
          is_public: boolean;
          is_static: boolean;
        }[];
        nonce: FieldLike;
        is_fee_payer: boolean;
      },
      cancellable: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** verify_private_authwit(inner_hash: field) */
    verify_private_authwit: ((inner_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
