// @aztec/noir-contracts.js/src/Escrow.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import EscrowContractArtifactJson from '../artifacts/escrow_contract-Escrow.json' assert { type: 'json' };

export const EscrowContractArtifact = loadContractArtifact(EscrowContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract Escrow;
 */
export class EscrowContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, EscrowContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, EscrowContract.artifact, wallet) as Promise<EscrowContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, owner: AztecAddressLike) {
    return new DeployMethod<EscrowContract>(
      PublicKeys.default(),
      wallet,
      EscrowContractArtifact,
      EscrowContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, owner: AztecAddressLike) {
    return new DeployMethod<EscrowContract>(
      publicKeys,
      wallet,
      EscrowContractArtifact,
      EscrowContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof EscrowContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<EscrowContract['methods'][M]>
  ) {
    return new DeployMethod<EscrowContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      EscrowContractArtifact,
      EscrowContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return EscrowContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'owner'> {
    return {
      owner: {
        slot: new Fr(1n),
      },
    } as ContractStorageLayout<'owner'>;
  }

  public static get notes(): ContractNotes<'AddressNote' | 'TransparentNote' | 'TokenNote'> {
    return {
      AddressNote: {
        id: new NoteSelector(2232136525),
      },
      TransparentNote: {
        id: new NoteSelector(3193649735),
      },
      TokenNote: {
        id: new NoteSelector(2350566847),
      },
    } as ContractNotes<'AddressNote' | 'TransparentNote' | 'TokenNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(owner: struct) */
    constructor: ((owner: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** withdraw(token: struct, amount: field, recipient: struct) */
    withdraw: ((
      token: AztecAddressLike,
      amount: FieldLike,
      recipient: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
