// @aztec/noir-contracts.js/src/InclusionProofs.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import InclusionProofsContractArtifactJson from '../artifacts/inclusion_proofs_contract-InclusionProofs.json' assert { type: 'json' };

export const InclusionProofsContractArtifact = loadContractArtifact(
  InclusionProofsContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract InclusionProofs;
 */
export class InclusionProofsContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, InclusionProofsContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, InclusionProofsContract.artifact, wallet) as Promise<InclusionProofsContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, public_value: FieldLike) {
    return new DeployMethod<InclusionProofsContract>(
      PublicKeys.default(),
      wallet,
      InclusionProofsContractArtifact,
      InclusionProofsContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, public_value: FieldLike) {
    return new DeployMethod<InclusionProofsContract>(
      publicKeys,
      wallet,
      InclusionProofsContractArtifact,
      InclusionProofsContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof InclusionProofsContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<InclusionProofsContract['methods'][M]>
  ) {
    return new DeployMethod<InclusionProofsContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      InclusionProofsContractArtifact,
      InclusionProofsContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return InclusionProofsContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'private_values' | 'public_value' | 'public_unused_value'> {
    return {
      private_values: {
        slot: new Fr(1n),
      },
      public_value: {
        slot: new Fr(2n),
      },
      public_unused_value: {
        slot: new Fr(3n),
      },
    } as ContractStorageLayout<'private_values' | 'public_value' | 'public_unused_value'>;
  }

  public static get notes(): ContractNotes<'ValueNote'> {
    return {
      ValueNote: {
        id: new NoteSelector(1038582377),
      },
    } as ContractNotes<'ValueNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(public_value: field) */
    constructor: ((public_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** create_note(owner: struct, value: field) */
    create_note: ((owner: AztecAddressLike, value: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** nullify_note(owner: struct) */
    nullify_note: ((owner: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** push_nullifier_public(nullifier: field) */
    push_nullifier_public: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** test_contract_inclusion(contract_address: struct, block_number: integer, test_deployment: boolean, test_initialization: boolean) */
    test_contract_inclusion: ((
      contract_address: AztecAddressLike,
      block_number: bigint | number,
      test_deployment: boolean,
      test_initialization: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_contract_non_inclusion(contract_address: struct, block_number: integer, test_deployment: boolean, test_initialization: boolean) */
    test_contract_non_inclusion: ((
      contract_address: AztecAddressLike,
      block_number: bigint | number,
      test_deployment: boolean,
      test_initialization: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_note_inclusion(owner: struct, use_block_number: boolean, block_number: integer, nullified: boolean) */
    test_note_inclusion: ((
      owner: AztecAddressLike,
      use_block_number: boolean,
      block_number: bigint | number,
      nullified: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_note_inclusion_fail_case(owner: struct, use_block_number: boolean, block_number: integer) */
    test_note_inclusion_fail_case: ((
      owner: AztecAddressLike,
      use_block_number: boolean,
      block_number: bigint | number,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_note_not_nullified(owner: struct, use_block_number: boolean, block_number: integer, fail_case: boolean) */
    test_note_not_nullified: ((
      owner: AztecAddressLike,
      use_block_number: boolean,
      block_number: bigint | number,
      fail_case: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_note_validity(owner: struct, use_block_number: boolean, block_number: integer, nullified: boolean) */
    test_note_validity: ((
      owner: AztecAddressLike,
      use_block_number: boolean,
      block_number: bigint | number,
      nullified: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_nullifier_inclusion(nullifier: field, use_block_number: boolean, block_number: integer) */
    test_nullifier_inclusion: ((
      nullifier: FieldLike,
      use_block_number: boolean,
      block_number: bigint | number,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_nullifier_inclusion_from_public(nullifier: field) */
    test_nullifier_inclusion_from_public: ((nullifier: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_storage_historical_read(expected: field, use_block_number: boolean, block_number: integer) */
    test_storage_historical_read: ((
      expected: FieldLike,
      use_block_number: boolean,
      block_number: bigint | number,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_storage_historical_read_unset_slot(block_number: integer) */
    test_storage_historical_read_unset_slot: ((block_number: bigint | number) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
