// @aztec/noir-contracts.js/src/PrivateFPC.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import PrivateFPCContractArtifactJson from '../artifacts/private_fpc_contract-PrivateFPC.json' assert { type: 'json' };

export const PrivateFPCContractArtifact = loadContractArtifact(PrivateFPCContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract PrivateFPC;
 */
export class PrivateFPCContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, PrivateFPCContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, PrivateFPCContract.artifact, wallet) as Promise<PrivateFPCContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, other_asset: AztecAddressLike, admin: AztecAddressLike) {
    return new DeployMethod<PrivateFPCContract>(
      PublicKeys.default(),
      wallet,
      PrivateFPCContractArtifact,
      PrivateFPCContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(
    publicKeys: PublicKeys,
    wallet: Wallet,
    other_asset: AztecAddressLike,
    admin: AztecAddressLike,
  ) {
    return new DeployMethod<PrivateFPCContract>(
      publicKeys,
      wallet,
      PrivateFPCContractArtifact,
      PrivateFPCContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof PrivateFPCContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<PrivateFPCContract['methods'][M]>
  ) {
    return new DeployMethod<PrivateFPCContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      PrivateFPCContractArtifact,
      PrivateFPCContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return PrivateFPCContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'settings'> {
    return {
      settings: {
        slot: new Fr(1n),
      },
    } as ContractStorageLayout<'settings'>;
  }

  public static get notes(): ContractNotes<'TransparentNote' | 'TokenNote'> {
    return {
      TransparentNote: {
        id: new NoteSelector(3193649735),
      },
      TokenNote: {
        id: new NoteSelector(2350566847),
      },
    } as ContractNotes<'TransparentNote' | 'TokenNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(other_asset: struct, admin: struct) */
    constructor: ((other_asset: AztecAddressLike, admin: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** fund_transaction_privately(amount: field, asset: struct) */
    fund_transaction_privately: ((amount: FieldLike, asset: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
