// @aztec/noir-contracts.js/src/Spam.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import SpamContractArtifactJson from '../artifacts/spam_contract-Spam.json' assert { type: 'json' };

export const SpamContractArtifact = loadContractArtifact(SpamContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract Spam;
 */
export class SpamContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, SpamContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, SpamContract.artifact, wallet) as Promise<SpamContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<SpamContract>(
      PublicKeys.default(),
      wallet,
      SpamContractArtifact,
      SpamContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet) {
    return new DeployMethod<SpamContract>(
      publicKeys,
      wallet,
      SpamContractArtifact,
      SpamContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof SpamContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<SpamContract['methods'][M]>
  ) {
    return new DeployMethod<SpamContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      SpamContractArtifact,
      SpamContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return SpamContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'balances' | 'public_balances'> {
    return {
      balances: {
        slot: new Fr(1n),
      },
      public_balances: {
        slot: new Fr(2n),
      },
    } as ContractStorageLayout<'balances' | 'public_balances'>;
  }

  public static get notes(): ContractNotes<'TokenNote' | 'ValueNote'> {
    return {
      TokenNote: {
        id: new NoteSelector(2350566847),
      },
      ValueNote: {
        id: new NoteSelector(1038582377),
      },
    } as ContractNotes<'TokenNote' | 'ValueNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** spam(nullifier_seed: field, nullifier_count: integer, call_public: boolean) */
    spam: ((
      nullifier_seed: FieldLike,
      nullifier_count: bigint | number,
      call_public: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
