// @aztec/noir-contracts.js/src/StaticParent.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import StaticParentContractArtifactJson from '../artifacts/static_parent_contract-StaticParent.json' assert { type: 'json' };

export const StaticParentContractArtifact = loadContractArtifact(
  StaticParentContractArtifactJson as NoirCompiledContract,
);

/**
 * Type-safe interface for contract StaticParent;
 */
export class StaticParentContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, StaticParentContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, StaticParentContract.artifact, wallet) as Promise<StaticParentContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<StaticParentContract>(
      PublicKeys.default(),
      wallet,
      StaticParentContractArtifact,
      StaticParentContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet) {
    return new DeployMethod<StaticParentContract>(
      publicKeys,
      wallet,
      StaticParentContractArtifact,
      StaticParentContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof StaticParentContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<StaticParentContract['methods'][M]>
  ) {
    return new DeployMethod<StaticParentContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      StaticParentContractArtifact,
      StaticParentContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return StaticParentContractArtifact;
  }

  public static get notes(): ContractNotes<'ValueNote'> {
    return {
      ValueNote: {
        id: new NoteSelector(1038582377),
      },
    } as ContractNotes<'ValueNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueue_call(target_contract: struct, target_selector: struct, args: array) */
    enqueue_call: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueue_public_get_value_from_child(target_contract: struct, value: field) */
    enqueue_public_get_value_from_child: ((
      target_contract: AztecAddressLike,
      value: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueue_static_call_to_pub_function(target_contract: struct, target_selector: struct, args: array) */
    enqueue_static_call_to_pub_function: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueue_static_nested_call_to_pub_function(target_contract: struct, target_selector: struct, args: array) */
    enqueue_static_nested_call_to_pub_function: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_call(target_contract: struct, target_selector: struct, args: array) */
    private_call: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_call_3_args(target_contract: struct, target_selector: struct, args: array) */
    private_call_3_args: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_get_value_from_child(target_contract: struct, value: field, owner: struct) */
    private_get_value_from_child: ((
      target_contract: AztecAddressLike,
      value: FieldLike,
      owner: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_nested_static_call(target_contract: struct, target_selector: struct, args: array) */
    private_nested_static_call: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_nested_static_call_3_args(target_contract: struct, target_selector: struct, args: array) */
    private_nested_static_call_3_args: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_static_call(target_contract: struct, target_selector: struct, args: array) */
    private_static_call: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** private_static_call_3_args(target_contract: struct, target_selector: struct, args: array) */
    private_static_call_3_args: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_call(target_contract: struct, target_selector: struct, arg: field) */
    public_call: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      arg: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_get_value_from_child(target_contract: struct, value: field) */
    public_get_value_from_child: ((
      target_contract: AztecAddressLike,
      value: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_nested_static_call(target_contract: struct, target_selector: struct, args: array) */
    public_nested_static_call: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_static_call(target_contract: struct, target_selector: struct, args: array) */
    public_static_call: ((
      target_contract: AztecAddressLike,
      target_selector: FunctionSelectorLike,
      args: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
