// @aztec/noir-contracts.js/src/Test.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import TestContractArtifactJson from '../artifacts/test_contract-Test.json' assert { type: 'json' };

export const TestContractArtifact = loadContractArtifact(TestContractArtifactJson as NoirCompiledContract);

export type ExampleEvent = {
  value0: FieldLike;
  value1: FieldLike;
  value2: FieldLike;
  value3: FieldLike;
  value4: FieldLike;
};

/**
 * Type-safe interface for contract Test;
 */
export class TestContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, TestContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, TestContract.artifact, wallet) as Promise<TestContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<TestContract>(
      PublicKeys.default(),
      wallet,
      TestContractArtifact,
      TestContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet) {
    return new DeployMethod<TestContract>(
      publicKeys,
      wallet,
      TestContractArtifact,
      TestContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof TestContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<TestContract['methods'][M]>
  ) {
    return new DeployMethod<TestContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      TestContractArtifact,
      TestContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return TestContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'example_constant' | 'example_set'> {
    return {
      example_constant: {
        slot: new Fr(1n),
      },
      example_set: {
        slot: new Fr(2n),
      },
    } as ContractStorageLayout<'example_constant' | 'example_set'>;
  }

  public static get notes(): ContractNotes<'TestNote' | 'ValueNote'> {
    return {
      TestNote: {
        id: new NoteSelector(1049718859),
      },
      ValueNote: {
        id: new NoteSelector(1038582377),
      },
    } as ContractNotes<'TestNote' | 'ValueNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** assert_header_private(header_hash: field) */
    assert_header_private: ((header_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** assert_private_global_vars(chain_id: field, version: field) */
    assert_private_global_vars: ((chain_id: FieldLike, version: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** assert_public_global_vars(chain_id: field, version: field, block_number: field, timestamp: integer, fee_per_da_gas: field, fee_per_l2_gas: field) */
    assert_public_global_vars: ((
      chain_id: FieldLike,
      version: FieldLike,
      block_number: FieldLike,
      timestamp: bigint | number,
      fee_per_da_gas: FieldLike,
      fee_per_l2_gas: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** call_create_note(value: field, owner: struct, outgoing_viewer: struct, storage_slot: field) */
    call_create_note: ((
      value: FieldLike,
      owner: AztecAddressLike,
      outgoing_viewer: AztecAddressLike,
      storage_slot: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** call_destroy_note(storage_slot: field) */
    call_destroy_note: ((storage_slot: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** call_get_notes(storage_slot: field, active_or_nullified: boolean) */
    call_get_notes: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** call_get_notes_many(storage_slot: field, active_or_nullified: boolean) */
    call_get_notes_many: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** call_view_notes(storage_slot: field, active_or_nullified: boolean) */
    call_view_notes: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** call_view_notes_many(storage_slot: field, active_or_nullified: boolean) */
    call_view_notes_many: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** consume_message_from_arbitrary_sender_private(content: field, secret: field, sender: struct) */
    consume_message_from_arbitrary_sender_private: ((
      content: FieldLike,
      secret: FieldLike,
      sender: EthAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** consume_message_from_arbitrary_sender_public(content: field, secret: field, sender: struct, message_leaf_index: field) */
    consume_message_from_arbitrary_sender_public: ((
      content: FieldLike,
      secret: FieldLike,
      sender: EthAddressLike,
      message_leaf_index: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** consume_mint_private_message(secret_hash_for_redeeming_minted_notes: field, amount: field, secret_for_L1_to_L2_message_consumption: field, portal_address: struct) */
    consume_mint_private_message: ((
      secret_hash_for_redeeming_minted_notes: FieldLike,
      amount: FieldLike,
      secret_for_L1_to_L2_message_consumption: FieldLike,
      portal_address: EthAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** consume_mint_public_message(to: struct, amount: field, secret: field, message_leaf_index: field, portal_address: struct) */
    consume_mint_public_message: ((
      to: AztecAddressLike,
      amount: FieldLike,
      secret: FieldLike,
      message_leaf_index: FieldLike,
      portal_address: EthAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** consume_note_from_secret(secret: field) */
    consume_note_from_secret: ((secret: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** create_l2_to_l1_message_arbitrary_recipient_private(content: field, recipient: struct) */
    create_l2_to_l1_message_arbitrary_recipient_private: ((
      content: FieldLike,
      recipient: EthAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** create_l2_to_l1_message_arbitrary_recipient_public(content: field, recipient: struct) */
    create_l2_to_l1_message_arbitrary_recipient_public: ((
      content: FieldLike,
      recipient: EthAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** create_l2_to_l1_message_public(amount: field, secret_hash: field, portal_address: struct) */
    create_l2_to_l1_message_public: ((
      amount: FieldLike,
      secret_hash: FieldLike,
      portal_address: EthAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** deploy_contract(target: struct) */
    deploy_contract: ((target: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** emit_array_as_encrypted_log(fields: array, owner: struct, outgoing_viewer: struct, nest: boolean) */
    emit_array_as_encrypted_log: ((
      fields: FieldLike[],
      owner: AztecAddressLike,
      outgoing_viewer: AztecAddressLike,
      nest: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** emit_encrypted_logs_nested(value: field, owner: struct, outgoing_viewer: struct) */
    emit_encrypted_logs_nested: ((
      value: FieldLike,
      owner: AztecAddressLike,
      outgoing_viewer: AztecAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** emit_nullifier(nullifier: field) */
    emit_nullifier: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** emit_nullifier_public(nullifier: field) */
    emit_nullifier_public: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** emit_unencrypted(value: field) */
    emit_unencrypted: ((value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_constant() */
    get_constant: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_master_incoming_viewing_public_key(address: struct) */
    get_master_incoming_viewing_public_key: ((address: AztecAddressLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** get_ovsk_app(ovpk_m_hash: field) */
    get_ovsk_app: ((ovpk_m_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_random(kinda_seed: field) */
    get_random: ((kinda_seed: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_this_address() */
    get_this_address: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** is_time_equal(time: integer) */
    is_time_equal: ((time: bigint | number) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_constant(value: field) */
    set_constant: ((value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_tx_max_block_number(max_block_number: integer, enqueue_public_call: boolean) */
    set_tx_max_block_number: ((
      max_block_number: bigint | number,
      enqueue_public_call: boolean,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_code_gen(a_field: field, a_bool: boolean, a_number: integer, an_array: array, a_struct: struct, a_deep_struct: struct) */
    test_code_gen: ((
      a_field: FieldLike,
      a_bool: boolean,
      a_number: bigint | number,
      an_array: FieldLike[],
      a_struct: { amount: FieldLike; secret_hash: FieldLike },
      a_deep_struct: {
        a_field: FieldLike;
        a_bool: boolean;
        a_note: { amount: FieldLike; secret_hash: FieldLike };
        many_notes: { amount: FieldLike; secret_hash: FieldLike }[];
      },
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_nullifier_key_freshness(address: struct, public_nullifying_key: struct) */
    test_nullifier_key_freshness: ((
      address: AztecAddressLike,
      public_nullifying_key: { x: FieldLike; y: FieldLike; is_infinite: boolean },
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_setting_fee_payer() */
    test_setting_fee_payer: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** test_setting_teardown() */
    test_setting_teardown: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  // Partial application is chosen is to avoid the duplication of so much codegen.
  private static decodeEvent<T>(
    eventSelector: EventSelector,
    eventType: AbiType,
  ): (payload: L1EventPayload | UnencryptedL2Log | undefined) => T | undefined {
    return (payload: L1EventPayload | UnencryptedL2Log | undefined): T | undefined => {
      if (payload === undefined) {
        return undefined;
      }

      if (payload instanceof L1EventPayload) {
        if (!eventSelector.equals(payload.eventTypeId)) {
          return undefined;
        }
        return decodeFromAbi([eventType], payload.event.items) as T;
      } else {
        let items = [];
        for (let i = 0; i < payload.data.length; i += 32) {
          items.push(new Fr(payload.data.subarray(i, i + 32)));
        }

        return decodeFromAbi([eventType], items) as T;
      }
    };
  }

  public static get events(): {
    ExampleEvent: {
      decode: (payload: L1EventPayload | UnencryptedL2Log | undefined) => ExampleEvent | undefined;
      eventSelector: EventSelector;
      fieldNames: string[];
    };
  } {
    return {
      ExampleEvent: {
        decode: this.decodeEvent(EventSelector.fromSignature('ExampleEvent(Field,Field,Field,Field,Field)'), {
          fields: [
            {
              name: 'value0',
              type: {
                kind: 'field',
              },
            },
            {
              name: 'value1',
              type: {
                kind: 'field',
              },
            },
            {
              name: 'value2',
              type: {
                kind: 'field',
              },
            },
            {
              name: 'value3',
              type: {
                kind: 'field',
              },
            },
            {
              name: 'value4',
              type: {
                kind: 'field',
              },
            },
          ],
          kind: 'struct',
          path: 'Test::ExampleEvent',
        }),
        eventSelector: EventSelector.fromSignature('ExampleEvent(Field,Field,Field,Field,Field)'),
        fieldNames: ['value0', 'value1', 'value2', 'value3', 'value4'],
      },
    };
  }
}
