// @aztec/noir-contracts.js/src/Uniswap.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractNotes,
  type ContractStorageLayout,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  L1EventPayload,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type UnencryptedL2Log,
  type Wallet,
  type WrappedFieldLike,
  decodeFromAbi,
  loadContractArtifact,
} from '@aztec/aztec.js';

import UniswapContractArtifactJson from '../artifacts/uniswap_contract-Uniswap.json' assert { type: 'json' };

export const UniswapContractArtifact = loadContractArtifact(UniswapContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract Uniswap;
 */
export class UniswapContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, UniswapContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, UniswapContract.artifact, wallet) as Promise<UniswapContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, portal_address: EthAddressLike) {
    return new DeployMethod<UniswapContract>(
      PublicKeys.default(),
      wallet,
      UniswapContractArtifact,
      UniswapContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, portal_address: EthAddressLike) {
    return new DeployMethod<UniswapContract>(
      publicKeys,
      wallet,
      UniswapContractArtifact,
      UniswapContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof UniswapContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<UniswapContract['methods'][M]>
  ) {
    return new DeployMethod<UniswapContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      UniswapContractArtifact,
      UniswapContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return UniswapContractArtifact;
  }

  public static get storage(): ContractStorageLayout<'portal_address'> {
    return {
      portal_address: {
        slot: new Fr(1n),
      },
    } as ContractStorageLayout<'portal_address'>;
  }

  public static get notes(): ContractNotes<'TransparentNote' | 'TokenNote'> {
    return {
      TransparentNote: {
        id: new NoteSelector(3193649735),
      },
      TokenNote: {
        id: new NoteSelector(2350566847),
      },
    } as ContractNotes<'TransparentNote' | 'TokenNote'>;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    /** compute_note_hash_and_optionally_a_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, compute_nullifier: boolean, serialized_note: array) */
    compute_note_hash_and_optionally_a_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      compute_nullifier: boolean,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** constructor(portal_address: struct) */
    constructor: ((portal_address: EthAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** swap_private(input_asset: struct, input_asset_bridge: struct, input_amount: field, output_asset_bridge: struct, nonce_for_unshield_approval: field, uniswap_fee_tier: field, minimum_output_amount: field, secret_hash_for_redeeming_minted_notes: field, secret_hash_for_L1_to_l2_message: field, caller_on_L1: struct) */
    swap_private: ((
      input_asset: AztecAddressLike,
      input_asset_bridge: AztecAddressLike,
      input_amount: FieldLike,
      output_asset_bridge: AztecAddressLike,
      nonce_for_unshield_approval: FieldLike,
      uniswap_fee_tier: FieldLike,
      minimum_output_amount: FieldLike,
      secret_hash_for_redeeming_minted_notes: FieldLike,
      secret_hash_for_L1_to_l2_message: FieldLike,
      caller_on_L1: EthAddressLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** swap_public(sender: struct, input_asset_bridge: struct, input_amount: field, output_asset_bridge: struct, nonce_for_transfer_approval: field, uniswap_fee_tier: field, minimum_output_amount: field, recipient: struct, secret_hash_for_L1_to_l2_message: field, caller_on_L1: struct, nonce_for_swap_approval: field) */
    swap_public: ((
      sender: AztecAddressLike,
      input_asset_bridge: AztecAddressLike,
      input_amount: FieldLike,
      output_asset_bridge: AztecAddressLike,
      nonce_for_transfer_approval: FieldLike,
      uniswap_fee_tier: FieldLike,
      minimum_output_amount: FieldLike,
      recipient: AztecAddressLike,
      secret_hash_for_L1_to_l2_message: FieldLike,
      caller_on_L1: EthAddressLike,
      nonce_for_swap_approval: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
