// @aztec/circuit-types/dest/p2p/block_proposal.js
import { Buffer32 } from '@aztec/foundation/buffer';
import { recoverAddress } from '@aztec/foundation/crypto';
import { Signature } from '@aztec/foundation/eth-signature';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { ConsensusPayload } from './consensus_payload.js';
import { Gossipable } from './gossipable.js';
import { getHashedSignaturePayload, getHashedSignaturePayloadEthSignedMessage } from './signature_utils.js';
import { TopicType, createTopicString } from './topic_type.js';
export class BlockProposalHash extends Buffer32 {
    constructor(hash) {
        super(hash);
    }
}
/**
 * BlockProposal
 *
 * A block proposal is created by the leader of the chain proposing a sequence of transactions to
 * be included in the head of the chain
 */
export class BlockProposal extends Gossipable {
    constructor(
    /** The payload of the message, and what the signature is over */
    payload, 
    /** The signer of the BlockProposal over the header of the new block*/
    signature) {
        super();
        this.payload = payload;
        this.signature = signature;
    }
    p2pMessageIdentifier() {
        return BlockProposalHash.fromField(this.payload.archive);
    }
    get archive() {
        return this.payload.archive;
    }
    static async createProposalFromSigner(payload, payloadSigner) {
        const hashed = getHashedSignaturePayload(payload);
        const sig = await payloadSigner(hashed);
        return new BlockProposal(payload, sig);
    }
    /**Get Sender
     * Lazily evaluate the sender of the proposal; result is cached
     */
    getSender() {
        if (!this.sender) {
            const hashed = getHashedSignaturePayloadEthSignedMessage(this.payload);
            // Cache the sender for later use
            this.sender = recoverAddress(hashed, this.signature);
        }
        return this.sender;
    }
    getPayload() {
        return this.payload.getPayloadToSign();
    }
    toBuffer() {
        return serializeToBuffer([this.payload, this.signature]);
    }
    static fromBuffer(buf) {
        const reader = BufferReader.asReader(buf);
        return new BlockProposal(reader.readObject(ConsensusPayload), reader.readObject(Signature));
    }
    getSize() {
        return this.payload.getSize() + this.signature.getSize();
    }
}
BlockProposal.p2pTopic = createTopicString(TopicType.block_proposal);
