// @aztec/circuit-types/dest/p2p/block_attestation.js
import { Buffer32 } from '@aztec/foundation/buffer';
import { keccak256, recoverAddress } from '@aztec/foundation/crypto';
import { Signature } from '@aztec/foundation/eth-signature';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { ConsensusPayload } from './consensus_payload.js';
import { Gossipable } from './gossipable.js';
import { getHashedSignaturePayloadEthSignedMessage } from './signature_utils.js';
import { TopicType, createTopicString } from './topic_type.js';
export class BlockAttestationHash extends Buffer32 {
    constructor(hash) {
        super(hash);
    }
}
/**
 * BlockAttestation
 *
 * A validator that has attested to seeing the contents of a block
 * will produce a block attestation over the header of the block
 */
export class BlockAttestation extends Gossipable {
    constructor(
    /** The payload of the message, and what the signature is over */
    payload, 
    /** The signature of the block attester */
    signature) {
        super();
        this.payload = payload;
        this.signature = signature;
    }
    p2pMessageIdentifier() {
        return new BlockAttestationHash(keccak256(this.signature.toBuffer()));
    }
    get archive() {
        return this.payload.archive;
    }
    /**Get sender
     *
     * Lazily evaluate and cache the sender of the attestation
     * @returns The sender of the attestation
     */
    getSender() {
        if (!this.sender) {
            // Recover the sender from the attestation
            const hashed = getHashedSignaturePayloadEthSignedMessage(this.payload);
            // Cache the sender for later use
            this.sender = recoverAddress(hashed, this.signature);
        }
        return this.sender;
    }
    getPayload() {
        return this.payload.getPayloadToSign();
    }
    toBuffer() {
        return serializeToBuffer([this.payload, this.signature]);
    }
    static fromBuffer(buf) {
        const reader = BufferReader.asReader(buf);
        return new BlockAttestation(reader.readObject(ConsensusPayload), reader.readObject(Signature));
    }
    static empty() {
        return new BlockAttestation(ConsensusPayload.empty(), Signature.empty());
    }
    getSize() {
        return this.payload.getSize() + this.signature.getSize();
    }
}
BlockAttestation.p2pTopic = createTopicString(TopicType.block_attestation);
