// @aztec/circuit-types/dest/p2p/consensus_payload.js
import { Header } from '@aztec/circuits.js';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { encodeAbiParameters, parseAbiParameters } from 'viem';
import { TxHash } from '../tx/tx_hash.js';
export class ConsensusPayload {
    constructor(
    /** The block header the attestation is made over */
    header, 
    // TODO(https://github.com/AztecProtocol/aztec-packages/pull/7727#discussion_r1713670830): temporary
    archive, 
    /** The sequence of transactions in the block */
    txHashes) {
        this.header = header;
        this.archive = archive;
        this.txHashes = txHashes;
    }
    static getFields(fields) {
        return [fields.header, fields.archive, fields.txHashes];
    }
    getPayloadToSign() {
        const abi = parseAbiParameters('bytes32, bytes32[]');
        const txArray = this.txHashes.map(tx => tx.to0xString());
        const encodedData = encodeAbiParameters(abi, [this.archive.toString(), txArray]);
        return Buffer.from(encodedData.slice(2), 'hex');
    }
    toBuffer() {
        const buffer = serializeToBuffer([this.header, this.archive, this.txHashes.length, this.txHashes]);
        this.size = buffer.length;
        return buffer;
    }
    static fromBuffer(buf) {
        const reader = BufferReader.asReader(buf);
        return new ConsensusPayload(reader.readObject(Header), reader.readObject(Fr), reader.readArray(reader.readNumber(), TxHash));
    }
    static fromFields(fields) {
        return new ConsensusPayload(fields.header, fields.archive, fields.txHashes);
    }
    static empty() {
        return new ConsensusPayload(Header.empty(), Fr.ZERO, []);
    }
    /**
     * Get the size of the consensus payload in bytes.
     * @returns The size of the consensus payload.
     */
    getSize() {
        // We cache size to avoid recalculating it
        if (this.size) {
            return this.size;
        }
        this.size = this.toBuffer().length;
        return this.size;
    }
}
