// @aztec/noir-contracts.js/dest/AppSubscription.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import AppSubscriptionContractArtifactJson from '../artifacts/app_subscription_contract-AppSubscription.json' assert { type: 'json' };
export const AppSubscriptionContractArtifact = loadContractArtifact(AppSubscriptionContractArtifactJson);
/**
 * Type-safe interface for contract AppSubscription;
 */
export class AppSubscriptionContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, AppSubscriptionContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, AppSubscriptionContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, target_address, subscription_recipient_address, subscription_token_address, subscription_price, fee_juice_limit_per_tx) {
        return new DeployMethod(PublicKeys.default(), wallet, AppSubscriptionContractArtifact, AppSubscriptionContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, target_address, subscription_recipient_address, subscription_token_address, subscription_price, fee_juice_limit_per_tx) {
        return new DeployMethod(publicKeys, wallet, AppSubscriptionContractArtifact, AppSubscriptionContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, AppSubscriptionContractArtifact, AppSubscriptionContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return AppSubscriptionContractArtifact;
    }
    static get storage() {
        return {
            target_address: {
                slot: new Fr(1n),
            },
            subscription_token_address: {
                slot: new Fr(2n),
            },
            subscription_recipient_address: {
                slot: new Fr(3n),
            },
            subscription_price: {
                slot: new Fr(4n),
            },
            subscriptions: {
                slot: new Fr(5n),
            },
            fee_juice_limit_per_tx: {
                slot: new Fr(6n),
            },
        };
    }
    static get notes() {
        return {
            TransparentNote: {
                id: new NoteSelector(3193649735),
            },
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
            SubscriptionNote: {
                id: new NoteSelector(3064460783),
            },
        };
    }
}
