// @aztec/noir-contracts.js/dest/AuthWitTest.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import AuthWitTestContractArtifactJson from '../artifacts/auth_wit_test_contract-AuthWitTest.json' assert { type: 'json' };
export const AuthWitTestContractArtifact = loadContractArtifact(AuthWitTestContractArtifactJson);
/**
 * Type-safe interface for contract AuthWitTest;
 */
export class AuthWitTestContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, AuthWitTestContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, AuthWitTestContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, AuthWitTestContractArtifact, AuthWitTestContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, AuthWitTestContractArtifact, AuthWitTestContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, AuthWitTestContractArtifact, AuthWitTestContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return AuthWitTestContractArtifact;
    }
}
