// @aztec/noir-contracts.js/dest/AvmInitializerTest.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import AvmInitializerTestContractArtifactJson from '../artifacts/avm_initializer_test_contract-AvmInitializerTest.json' assert { type: 'json' };
export const AvmInitializerTestContractArtifact = loadContractArtifact(AvmInitializerTestContractArtifactJson);
/**
 * Type-safe interface for contract AvmInitializerTest;
 */
export class AvmInitializerTestContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, AvmInitializerTestContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, AvmInitializerTestContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, AvmInitializerTestContractArtifact, AvmInitializerTestContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, AvmInitializerTestContractArtifact, AvmInitializerTestContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, AvmInitializerTestContractArtifact, AvmInitializerTestContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return AvmInitializerTestContractArtifact;
    }
    static get storage() {
        return {
            immutable: {
                slot: new Fr(1n),
            },
        };
    }
}
