// @aztec/noir-contracts.js/dest/AvmTest.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import AvmTestContractArtifactJson from '../artifacts/avm_test_contract-AvmTest.json' assert { type: 'json' };
export const AvmTestContractArtifact = loadContractArtifact(AvmTestContractArtifactJson);
/**
 * Type-safe interface for contract AvmTest;
 */
export class AvmTestContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, AvmTestContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, AvmTestContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, AvmTestContractArtifact, AvmTestContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, AvmTestContractArtifact, AvmTestContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, AvmTestContractArtifact, AvmTestContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return AvmTestContractArtifact;
    }
    static get storage() {
        return {
            single: {
                slot: new Fr(1n),
            },
            list: {
                slot: new Fr(2n),
            },
            map: {
                slot: new Fr(4n),
            },
        };
    }
}
