// @aztec/noir-contracts.js/dest/Benchmarking.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import BenchmarkingContractArtifactJson from '../artifacts/benchmarking_contract-Benchmarking.json' assert { type: 'json' };
export const BenchmarkingContractArtifact = loadContractArtifact(BenchmarkingContractArtifactJson);
/**
 * Type-safe interface for contract Benchmarking;
 */
export class BenchmarkingContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, BenchmarkingContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, BenchmarkingContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, BenchmarkingContractArtifact, BenchmarkingContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, BenchmarkingContractArtifact, BenchmarkingContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, BenchmarkingContractArtifact, BenchmarkingContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return BenchmarkingContractArtifact;
    }
    static get storage() {
        return {
            notes: {
                slot: new Fr(1n),
            },
            balances: {
                slot: new Fr(2n),
            },
        };
    }
    static get notes() {
        return {
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
}
