// @aztec/noir-contracts.js/dest/CardGame.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import CardGameContractArtifactJson from '../artifacts/card_game_contract-CardGame.json' assert { type: 'json' };
export const CardGameContractArtifact = loadContractArtifact(CardGameContractArtifactJson);
/**
 * Type-safe interface for contract CardGame;
 */
export class CardGameContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, CardGameContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, CardGameContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, CardGameContractArtifact, CardGameContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, CardGameContractArtifact, CardGameContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, CardGameContractArtifact, CardGameContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return CardGameContractArtifact;
    }
    static get storage() {
        return {
            collections: {
                slot: new Fr(1n),
            },
            game_decks: {
                slot: new Fr(2n),
            },
            games: {
                slot: new Fr(3n),
            },
        };
    }
    static get notes() {
        return {
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
}
