// @aztec/noir-contracts.js/dest/Claim.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import ClaimContractArtifactJson from '../artifacts/claim_contract-Claim.json' assert { type: 'json' };
export const ClaimContractArtifact = loadContractArtifact(ClaimContractArtifactJson);
/**
 * Type-safe interface for contract Claim;
 */
export class ClaimContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, ClaimContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, ClaimContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, target_contract, reward_token) {
        return new DeployMethod(PublicKeys.default(), wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, target_contract, reward_token) {
        return new DeployMethod(publicKeys, wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return ClaimContractArtifact;
    }
    static get storage() {
        return {
            target_contract: {
                slot: new Fr(1n),
            },
            reward_token: {
                slot: new Fr(2n),
            },
        };
    }
    static get notes() {
        return {
            TransparentNote: {
                id: new NoteSelector(3193649735),
            },
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
}
