// @aztec/noir-contracts.js/dest/ContractClassRegisterer.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import ContractClassRegistererContractArtifactJson from '../artifacts/contract_class_registerer_contract-ContractClassRegisterer.json' assert { type: 'json' };
export const ContractClassRegistererContractArtifact = loadContractArtifact(ContractClassRegistererContractArtifactJson);
/**
 * Type-safe interface for contract ContractClassRegisterer;
 */
export class ContractClassRegistererContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, ContractClassRegistererContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, ContractClassRegistererContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, ContractClassRegistererContractArtifact, ContractClassRegistererContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, ContractClassRegistererContractArtifact, ContractClassRegistererContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, ContractClassRegistererContractArtifact, ContractClassRegistererContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return ContractClassRegistererContractArtifact;
    }
}
