// @aztec/noir-contracts.js/dest/ContractInstanceDeployer.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EventSelector, Fr, L1EventPayload, PublicKeys, decodeFromAbi, loadContractArtifact, } from '@aztec/aztec.js';
import ContractInstanceDeployerContractArtifactJson from '../artifacts/contract_instance_deployer_contract-ContractInstanceDeployer.json' assert { type: 'json' };
export const ContractInstanceDeployerContractArtifact = loadContractArtifact(ContractInstanceDeployerContractArtifactJson);
/**
 * Type-safe interface for contract ContractInstanceDeployer;
 */
export class ContractInstanceDeployerContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, ContractInstanceDeployerContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, ContractInstanceDeployerContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, ContractInstanceDeployerContractArtifact, ContractInstanceDeployerContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, ContractInstanceDeployerContractArtifact, ContractInstanceDeployerContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, ContractInstanceDeployerContractArtifact, ContractInstanceDeployerContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return ContractInstanceDeployerContractArtifact;
    }
    // Partial application is chosen is to avoid the duplication of so much codegen.
    static decodeEvent(eventSelector, eventType) {
        return (payload) => {
            if (payload === undefined) {
                return undefined;
            }
            if (payload instanceof L1EventPayload) {
                if (!eventSelector.equals(payload.eventTypeId)) {
                    return undefined;
                }
                return decodeFromAbi([eventType], payload.event.items);
            }
            else {
                let items = [];
                for (let i = 0; i < payload.data.length; i += 32) {
                    items.push(new Fr(payload.data.subarray(i, i + 32)));
                }
                return decodeFromAbi([eventType], items);
            }
        };
    }
    static get events() {
        return {
            ContractInstanceDeployed: {
                decode: this.decodeEvent(EventSelector.fromSignature('ContractInstanceDeployed(Field,(Field),u8,Field,(Field),Field,(((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool))),(Field))'), {
                    fields: [
                        {
                            name: 'DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'address',
                            type: {
                                fields: [
                                    {
                                        name: 'inner',
                                        type: {
                                            kind: 'field',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
                            },
                        },
                        {
                            name: 'version',
                            type: {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 8,
                            },
                        },
                        {
                            name: 'salt',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'contract_class_id',
                            type: {
                                fields: [
                                    {
                                        name: 'inner',
                                        type: {
                                            kind: 'field',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'aztec::protocol_types::contract_class_id::ContractClassId',
                            },
                        },
                        {
                            name: 'initialization_hash',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'public_keys',
                            type: {
                                fields: [
                                    {
                                        name: 'npk_m',
                                        type: {
                                            fields: [
                                                {
                                                    name: 'inner',
                                                    type: {
                                                        fields: [
                                                            {
                                                                name: 'x',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'y',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'is_infinite',
                                                                type: {
                                                                    kind: 'boolean',
                                                                },
                                                            },
                                                        ],
                                                        kind: 'struct',
                                                        path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                                                    },
                                                },
                                            ],
                                            kind: 'struct',
                                            path: 'aztec::protocol_types::public_keys::NpkM',
                                        },
                                    },
                                    {
                                        name: 'ivpk_m',
                                        type: {
                                            fields: [
                                                {
                                                    name: 'inner',
                                                    type: {
                                                        fields: [
                                                            {
                                                                name: 'x',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'y',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'is_infinite',
                                                                type: {
                                                                    kind: 'boolean',
                                                                },
                                                            },
                                                        ],
                                                        kind: 'struct',
                                                        path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                                                    },
                                                },
                                            ],
                                            kind: 'struct',
                                            path: 'aztec::protocol_types::public_keys::IvpkM',
                                        },
                                    },
                                    {
                                        name: 'ovpk_m',
                                        type: {
                                            fields: [
                                                {
                                                    name: 'inner',
                                                    type: {
                                                        fields: [
                                                            {
                                                                name: 'x',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'y',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'is_infinite',
                                                                type: {
                                                                    kind: 'boolean',
                                                                },
                                                            },
                                                        ],
                                                        kind: 'struct',
                                                        path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                                                    },
                                                },
                                            ],
                                            kind: 'struct',
                                            path: 'aztec::protocol_types::public_keys::OvpkM',
                                        },
                                    },
                                    {
                                        name: 'tpk_m',
                                        type: {
                                            fields: [
                                                {
                                                    name: 'inner',
                                                    type: {
                                                        fields: [
                                                            {
                                                                name: 'x',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'y',
                                                                type: {
                                                                    kind: 'field',
                                                                },
                                                            },
                                                            {
                                                                name: 'is_infinite',
                                                                type: {
                                                                    kind: 'boolean',
                                                                },
                                                            },
                                                        ],
                                                        kind: 'struct',
                                                        path: 'std::embedded_curve_ops::EmbeddedCurvePoint',
                                                    },
                                                },
                                            ],
                                            kind: 'struct',
                                            path: 'aztec::protocol_types::public_keys::TpkM',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'aztec::protocol_types::public_keys::PublicKeys',
                            },
                        },
                        {
                            name: 'deployer',
                            type: {
                                fields: [
                                    {
                                        name: 'inner',
                                        type: {
                                            kind: 'field',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
                            },
                        },
                    ],
                    kind: 'struct',
                    path: 'ContractInstanceDeployer::ContractInstanceDeployed',
                }),
                eventSelector: EventSelector.fromSignature('ContractInstanceDeployed(Field,(Field),u8,Field,(Field),Field,(((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool)),((Field,Field,bool))),(Field))'),
                fieldNames: [
                    'DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE',
                    'address',
                    'version',
                    'salt',
                    'contract_class_id',
                    'initialization_hash',
                    'public_keys',
                    'deployer',
                ],
            },
        };
    }
}
