// @aztec/noir-contracts.js/dest/Counter.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import CounterContractArtifactJson from '../artifacts/counter_contract-Counter.json' assert { type: 'json' };
export const CounterContractArtifact = loadContractArtifact(CounterContractArtifactJson);
/**
 * Type-safe interface for contract Counter;
 */
export class CounterContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, CounterContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, CounterContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, headstart, owner, outgoing_viewer) {
        return new DeployMethod(PublicKeys.default(), wallet, CounterContractArtifact, CounterContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, headstart, owner, outgoing_viewer) {
        return new DeployMethod(publicKeys, wallet, CounterContractArtifact, CounterContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, CounterContractArtifact, CounterContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return CounterContractArtifact;
    }
    static get storage() {
        return {
            counters: {
                slot: new Fr(1n),
            },
        };
    }
    static get notes() {
        return {
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
}
