// @aztec/noir-contracts.js/dest/Crowdfunding.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EventSelector, Fr, L1EventPayload, NoteSelector, PublicKeys, decodeFromAbi, loadContractArtifact, } from '@aztec/aztec.js';
import CrowdfundingContractArtifactJson from '../artifacts/crowdfunding_contract-Crowdfunding.json' assert { type: 'json' };
export const CrowdfundingContractArtifact = loadContractArtifact(CrowdfundingContractArtifactJson);
/**
 * Type-safe interface for contract Crowdfunding;
 */
export class CrowdfundingContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, CrowdfundingContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, CrowdfundingContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, donation_token, operator, deadline) {
        return new DeployMethod(PublicKeys.default(), wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, donation_token, operator, deadline) {
        return new DeployMethod(publicKeys, wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return CrowdfundingContractArtifact;
    }
    static get storage() {
        return {
            donation_token: {
                slot: new Fr(1n),
            },
            operator: {
                slot: new Fr(2n),
            },
            deadline: {
                slot: new Fr(3n),
            },
            donation_receipts: {
                slot: new Fr(4n),
            },
        };
    }
    static get notes() {
        return {
            TransparentNote: {
                id: new NoteSelector(3193649735),
            },
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
    // Partial application is chosen is to avoid the duplication of so much codegen.
    static decodeEvent(eventSelector, eventType) {
        return (payload) => {
            if (payload === undefined) {
                return undefined;
            }
            if (payload instanceof L1EventPayload) {
                if (!eventSelector.equals(payload.eventTypeId)) {
                    return undefined;
                }
                return decodeFromAbi([eventType], payload.event.items);
            }
            else {
                let items = [];
                for (let i = 0; i < payload.data.length; i += 32) {
                    items.push(new Fr(payload.data.subarray(i, i + 32)));
                }
                return decodeFromAbi([eventType], items);
            }
        };
    }
    static get events() {
        return {
            WithdrawalProcessed: {
                decode: this.decodeEvent(EventSelector.fromSignature('WithdrawalProcessed((Field),u64)'), {
                    fields: [
                        {
                            name: 'who',
                            type: {
                                fields: [
                                    {
                                        name: 'inner',
                                        type: {
                                            kind: 'field',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
                            },
                        },
                        {
                            name: 'amount',
                            type: {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 64,
                            },
                        },
                    ],
                    kind: 'struct',
                    path: 'Crowdfunding::WithdrawalProcessed',
                }),
                eventSelector: EventSelector.fromSignature('WithdrawalProcessed((Field),u64)'),
                fieldNames: ['who', 'amount'],
            },
        };
    }
}
