// @aztec/noir-contracts.js/dest/DocsExample.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import DocsExampleContractArtifactJson from '../artifacts/docs_example_contract-DocsExample.json' assert { type: 'json' };
export const DocsExampleContractArtifact = loadContractArtifact(DocsExampleContractArtifactJson);
/**
 * Type-safe interface for contract DocsExample;
 */
export class DocsExampleContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, DocsExampleContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, DocsExampleContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, DocsExampleContractArtifact, DocsExampleContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, DocsExampleContractArtifact, DocsExampleContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, DocsExampleContractArtifact, DocsExampleContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return DocsExampleContractArtifact;
    }
    static get notes() {
        return {
            CardNote: {
                id: new NoteSelector(3719046069),
            },
        };
    }
}
