// @aztec/noir-contracts.js/dest/EcdsaKAccount.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import EcdsaKAccountContractArtifactJson from '../artifacts/ecdsa_k_account_contract-EcdsaKAccount.json' assert { type: 'json' };
export const EcdsaKAccountContractArtifact = loadContractArtifact(EcdsaKAccountContractArtifactJson);
/**
 * Type-safe interface for contract EcdsaKAccount;
 */
export class EcdsaKAccountContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, EcdsaKAccountContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, EcdsaKAccountContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, signing_pub_key_x, signing_pub_key_y) {
        return new DeployMethod(PublicKeys.default(), wallet, EcdsaKAccountContractArtifact, EcdsaKAccountContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, signing_pub_key_x, signing_pub_key_y) {
        return new DeployMethod(publicKeys, wallet, EcdsaKAccountContractArtifact, EcdsaKAccountContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, EcdsaKAccountContractArtifact, EcdsaKAccountContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return EcdsaKAccountContractArtifact;
    }
    static get storage() {
        return {
            public_key: {
                slot: new Fr(1n),
            },
        };
    }
    static get notes() {
        return {
            EcdsaPublicKeyNote: {
                id: new NoteSelector(79774112),
            },
        };
    }
}
