// @aztec/noir-contracts.js/dest/FPC.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import FPCContractArtifactJson from '../artifacts/fpc_contract-FPC.json' assert { type: 'json' };
export const FPCContractArtifact = loadContractArtifact(FPCContractArtifactJson);
/**
 * Type-safe interface for contract FPC;
 */
export class FPCContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, FPCContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, FPCContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, other_asset) {
        return new DeployMethod(PublicKeys.default(), wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, other_asset) {
        return new DeployMethod(publicKeys, wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return FPCContractArtifact;
    }
    static get storage() {
        return {
            other_asset: {
                slot: new Fr(1n),
            },
        };
    }
    static get notes() {
        return {
            TransparentNote: {
                id: new NoteSelector(3193649735),
            },
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
        };
    }
}
