// @aztec/noir-contracts.js/dest/ImportTest.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import ImportTestContractArtifactJson from '../artifacts/import_test_contract-ImportTest.json' assert { type: 'json' };
export const ImportTestContractArtifact = loadContractArtifact(ImportTestContractArtifactJson);
/**
 * Type-safe interface for contract ImportTest;
 */
export class ImportTestContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, ImportTestContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, ImportTestContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, ImportTestContractArtifact, ImportTestContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, ImportTestContractArtifact, ImportTestContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, ImportTestContractArtifact, ImportTestContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return ImportTestContractArtifact;
    }
    static get notes() {
        return {
            TestNote: {
                id: new NoteSelector(1049718859),
            },
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
}
