// @aztec/noir-contracts.js/dest/Lending.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import LendingContractArtifactJson from '../artifacts/lending_contract-Lending.json' assert { type: 'json' };
export const LendingContractArtifact = loadContractArtifact(LendingContractArtifactJson);
/**
 * Type-safe interface for contract Lending;
 */
export class LendingContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, LendingContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, LendingContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, LendingContractArtifact, LendingContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, LendingContractArtifact, LendingContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, LendingContractArtifact, LendingContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return LendingContractArtifact;
    }
    static get storage() {
        return {
            collateral_asset: {
                slot: new Fr(1n),
            },
            stable_coin: {
                slot: new Fr(2n),
            },
            assets: {
                slot: new Fr(3n),
            },
            collateral: {
                slot: new Fr(4n),
            },
            static_debt: {
                slot: new Fr(5n),
            },
        };
    }
    static get notes() {
        return {
            TransparentNote: {
                id: new NoteSelector(3193649735),
            },
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
        };
    }
}
