// @aztec/noir-contracts.js/dest/MultiCallEntrypoint.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import MultiCallEntrypointContractArtifactJson from '../artifacts/multi_call_entrypoint_contract-MultiCallEntrypoint.json' assert { type: 'json' };
export const MultiCallEntrypointContractArtifact = loadContractArtifact(MultiCallEntrypointContractArtifactJson);
/**
 * Type-safe interface for contract MultiCallEntrypoint;
 */
export class MultiCallEntrypointContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, MultiCallEntrypointContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, MultiCallEntrypointContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, MultiCallEntrypointContractArtifact, MultiCallEntrypointContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, MultiCallEntrypointContractArtifact, MultiCallEntrypointContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, MultiCallEntrypointContractArtifact, MultiCallEntrypointContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return MultiCallEntrypointContractArtifact;
    }
}
