// @aztec/noir-contracts.js/dest/NFT.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EventSelector, Fr, L1EventPayload, NoteSelector, PublicKeys, decodeFromAbi, loadContractArtifact, } from '@aztec/aztec.js';
import NFTContractArtifactJson from '../artifacts/nft_contract-NFT.json' assert { type: 'json' };
export const NFTContractArtifact = loadContractArtifact(NFTContractArtifactJson);
/**
 * Type-safe interface for contract NFT;
 */
export class NFTContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, NFTContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, NFTContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, admin, name, symbol) {
        return new DeployMethod(PublicKeys.default(), wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, admin, name, symbol) {
        return new DeployMethod(publicKeys, wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return NFTContractArtifact;
    }
    static get storage() {
        return {
            symbol: {
                slot: new Fr(1n),
            },
            name: {
                slot: new Fr(2n),
            },
            admin: {
                slot: new Fr(3n),
            },
            minters: {
                slot: new Fr(4n),
            },
            private_nfts: {
                slot: new Fr(5n),
            },
            nft_exists: {
                slot: new Fr(6n),
            },
            public_owners: {
                slot: new Fr(7n),
            },
        };
    }
    static get notes() {
        return {
            NFTNote: {
                id: new NoteSelector(3595710486),
            },
        };
    }
    // Partial application is chosen is to avoid the duplication of so much codegen.
    static decodeEvent(eventSelector, eventType) {
        return (payload) => {
            if (payload === undefined) {
                return undefined;
            }
            if (payload instanceof L1EventPayload) {
                if (!eventSelector.equals(payload.eventTypeId)) {
                    return undefined;
                }
                return decodeFromAbi([eventType], payload.event.items);
            }
            else {
                let items = [];
                for (let i = 0; i < payload.data.length; i += 32) {
                    items.push(new Fr(payload.data.subarray(i, i + 32)));
                }
                return decodeFromAbi([eventType], items);
            }
        };
    }
    static get events() {
        return {
            NFTTransfer: {
                decode: this.decodeEvent(EventSelector.fromSignature('NFTTransfer((Field),(Field),Field)'), {
                    fields: [
                        {
                            name: 'from',
                            type: {
                                fields: [
                                    {
                                        name: 'inner',
                                        type: {
                                            kind: 'field',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'authwit::aztec::protocol_types::address::aztec_address::AztecAddress',
                            },
                        },
                        {
                            name: 'to',
                            type: {
                                fields: [
                                    {
                                        name: 'inner',
                                        type: {
                                            kind: 'field',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'authwit::aztec::protocol_types::address::aztec_address::AztecAddress',
                            },
                        },
                        {
                            name: 'token_id',
                            type: {
                                kind: 'field',
                            },
                        },
                    ],
                    kind: 'struct',
                    path: 'NFT::NFTTransfer',
                }),
                eventSelector: EventSelector.fromSignature('NFTTransfer((Field),(Field),Field)'),
                fieldNames: ['from', 'to', 'token_id'],
            },
        };
    }
}
