// @aztec/noir-contracts.js/dest/Router.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import RouterContractArtifactJson from '../artifacts/router_contract-Router.json' assert { type: 'json' };
export const RouterContractArtifact = loadContractArtifact(RouterContractArtifactJson);
/**
 * Type-safe interface for contract Router;
 */
export class RouterContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, RouterContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, RouterContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, RouterContractArtifact, RouterContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, RouterContractArtifact, RouterContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, RouterContractArtifact, RouterContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return RouterContractArtifact;
    }
}
