// @aztec/noir-contracts.js/dest/Spam.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import SpamContractArtifactJson from '../artifacts/spam_contract-Spam.json' assert { type: 'json' };
export const SpamContractArtifact = loadContractArtifact(SpamContractArtifactJson);
/**
 * Type-safe interface for contract Spam;
 */
export class SpamContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, SpamContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, SpamContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, SpamContractArtifact, SpamContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, SpamContractArtifact, SpamContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, SpamContractArtifact, SpamContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return SpamContractArtifact;
    }
    static get storage() {
        return {
            balances: {
                slot: new Fr(1n),
            },
            public_balances: {
                slot: new Fr(2n),
            },
        };
    }
    static get notes() {
        return {
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
}
