// @aztec/noir-contracts.js/dest/TestLog.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EventSelector, Fr, L1EventPayload, NoteSelector, PublicKeys, decodeFromAbi, loadContractArtifact, } from '@aztec/aztec.js';
import TestLogContractArtifactJson from '../artifacts/test_log_contract-TestLog.json' assert { type: 'json' };
export const TestLogContractArtifact = loadContractArtifact(TestLogContractArtifactJson);
/**
 * Type-safe interface for contract TestLog;
 */
export class TestLogContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, TestLogContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, TestLogContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, TestLogContractArtifact, TestLogContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, TestLogContractArtifact, TestLogContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, TestLogContractArtifact, TestLogContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return TestLogContractArtifact;
    }
    static get storage() {
        return {
            example_set: {
                slot: new Fr(1n),
            },
        };
    }
    static get notes() {
        return {
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
    // Partial application is chosen is to avoid the duplication of so much codegen.
    static decodeEvent(eventSelector, eventType) {
        return (payload) => {
            if (payload === undefined) {
                return undefined;
            }
            if (payload instanceof L1EventPayload) {
                if (!eventSelector.equals(payload.eventTypeId)) {
                    return undefined;
                }
                return decodeFromAbi([eventType], payload.event.items);
            }
            else {
                let items = [];
                for (let i = 0; i < payload.data.length; i += 32) {
                    items.push(new Fr(payload.data.subarray(i, i + 32)));
                }
                return decodeFromAbi([eventType], items);
            }
        };
    }
    static get events() {
        return {
            ExampleEvent1: {
                decode: this.decodeEvent(EventSelector.fromSignature('ExampleEvent1((Field),u8)'), {
                    fields: [
                        {
                            name: 'value2',
                            type: {
                                fields: [
                                    {
                                        name: 'inner',
                                        type: {
                                            kind: 'field',
                                        },
                                    },
                                ],
                                kind: 'struct',
                                path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
                            },
                        },
                        {
                            name: 'value3',
                            type: {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 8,
                            },
                        },
                    ],
                    kind: 'struct',
                    path: 'TestLog::ExampleEvent1',
                }),
                eventSelector: EventSelector.fromSignature('ExampleEvent1((Field),u8)'),
                fieldNames: ['value2', 'value3'],
            },
            ExampleEvent0: {
                decode: this.decodeEvent(EventSelector.fromSignature('ExampleEvent0(Field,Field)'), {
                    fields: [
                        {
                            name: 'value0',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'value1',
                            type: {
                                kind: 'field',
                            },
                        },
                    ],
                    kind: 'struct',
                    path: 'TestLog::ExampleEvent0',
                }),
                eventSelector: EventSelector.fromSignature('ExampleEvent0(Field,Field)'),
                fieldNames: ['value0', 'value1'],
            },
        };
    }
}
