// @aztec/noir-contracts.js/dest/Test.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, EventSelector, Fr, L1EventPayload, NoteSelector, PublicKeys, decodeFromAbi, loadContractArtifact, } from '@aztec/aztec.js';
import TestContractArtifactJson from '../artifacts/test_contract-Test.json' assert { type: 'json' };
export const TestContractArtifact = loadContractArtifact(TestContractArtifactJson);
/**
 * Type-safe interface for contract Test;
 */
export class TestContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, TestContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, TestContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, TestContractArtifact, TestContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, TestContractArtifact, TestContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, TestContractArtifact, TestContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return TestContractArtifact;
    }
    static get storage() {
        return {
            example_constant: {
                slot: new Fr(1n),
            },
            example_set: {
                slot: new Fr(2n),
            },
        };
    }
    static get notes() {
        return {
            TestNote: {
                id: new NoteSelector(1049718859),
            },
            ValueNote: {
                id: new NoteSelector(1038582377),
            },
        };
    }
    // Partial application is chosen is to avoid the duplication of so much codegen.
    static decodeEvent(eventSelector, eventType) {
        return (payload) => {
            if (payload === undefined) {
                return undefined;
            }
            if (payload instanceof L1EventPayload) {
                if (!eventSelector.equals(payload.eventTypeId)) {
                    return undefined;
                }
                return decodeFromAbi([eventType], payload.event.items);
            }
            else {
                let items = [];
                for (let i = 0; i < payload.data.length; i += 32) {
                    items.push(new Fr(payload.data.subarray(i, i + 32)));
                }
                return decodeFromAbi([eventType], items);
            }
        };
    }
    static get events() {
        return {
            ExampleEvent: {
                decode: this.decodeEvent(EventSelector.fromSignature('ExampleEvent(Field,Field,Field,Field,Field)'), {
                    fields: [
                        {
                            name: 'value0',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'value1',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'value2',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'value3',
                            type: {
                                kind: 'field',
                            },
                        },
                        {
                            name: 'value4',
                            type: {
                                kind: 'field',
                            },
                        },
                    ],
                    kind: 'struct',
                    path: 'Test::ExampleEvent',
                }),
                eventSelector: EventSelector.fromSignature('ExampleEvent(Field,Field,Field,Field,Field)'),
                fieldNames: ['value0', 'value1', 'value2', 'value3', 'value4'],
            },
        };
    }
}
