// @aztec/noir-contracts.js/dest/TokenBlacklist.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import TokenBlacklistContractArtifactJson from '../artifacts/token_blacklist_contract-TokenBlacklist.json' assert { type: 'json' };
export const TokenBlacklistContractArtifact = loadContractArtifact(TokenBlacklistContractArtifactJson);
/**
 * Type-safe interface for contract TokenBlacklist;
 */
export class TokenBlacklistContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, TokenBlacklistContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, TokenBlacklistContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, admin) {
        return new DeployMethod(PublicKeys.default(), wallet, TokenBlacklistContractArtifact, TokenBlacklistContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, admin) {
        return new DeployMethod(publicKeys, wallet, TokenBlacklistContractArtifact, TokenBlacklistContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, TokenBlacklistContractArtifact, TokenBlacklistContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return TokenBlacklistContractArtifact;
    }
    static get storage() {
        return {
            balances: {
                slot: new Fr(1n),
            },
            total_supply: {
                slot: new Fr(2n),
            },
            pending_shields: {
                slot: new Fr(3n),
            },
            public_balances: {
                slot: new Fr(4n),
            },
            roles: {
                slot: new Fr(5n),
            },
        };
    }
    static get notes() {
        return {
            TransparentNote: {
                id: new NoteSelector(3193649735),
            },
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
        };
    }
}
