// @aztec/noir-contracts.js/dest/Uniswap.js
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { Contract, ContractBase, DeployMethod, Fr, NoteSelector, PublicKeys, loadContractArtifact, } from '@aztec/aztec.js';
import UniswapContractArtifactJson from '../artifacts/uniswap_contract-Uniswap.json' assert { type: 'json' };
export const UniswapContractArtifact = loadContractArtifact(UniswapContractArtifactJson);
/**
 * Type-safe interface for contract Uniswap;
 */
export class UniswapContract extends ContractBase {
    constructor(instance, wallet) {
        super(instance, UniswapContractArtifact, wallet);
    }
    /**
     * Creates a contract instance.
     * @param address - The deployed contract's address.
     * @param wallet - The wallet to use when interacting with the contract.
     * @returns A promise that resolves to a new Contract instance.
     */
    static async at(address, wallet) {
        return Contract.at(address, UniswapContract.artifact, wallet);
    }
    /**
     * Creates a tx to deploy a new instance of this contract.
     */
    static deploy(wallet, portal_address) {
        return new DeployMethod(PublicKeys.default(), wallet, UniswapContractArtifact, UniswapContract.at, Array.from(arguments).slice(1));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
     */
    static deployWithPublicKeys(publicKeys, wallet, portal_address) {
        return new DeployMethod(publicKeys, wallet, UniswapContractArtifact, UniswapContract.at, Array.from(arguments).slice(2));
    }
    /**
     * Creates a tx to deploy a new instance of this contract using the specified constructor method.
     */
    static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, UniswapContractArtifact, UniswapContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
     * Returns this contract's artifact.
     */
    static get artifact() {
        return UniswapContractArtifact;
    }
    static get storage() {
        return {
            portal_address: {
                slot: new Fr(1n),
            },
        };
    }
    static get notes() {
        return {
            TransparentNote: {
                id: new NoteSelector(3193649735),
            },
            TokenNote: {
                id: new NoteSelector(2350566847),
            },
        };
    }
}
