// @aztec/circuits.js/dest/structs/parity/root_parity_input.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { VK_TREE_HEIGHT } from '../../constants.gen.js';
import { RecursiveProof } from '../recursive_proof.js';
import { VerificationKeyAsFields } from '../verification_key.js';
import { ParityPublicInputs } from './parity_public_inputs.js';
export class RootParityInput {
    constructor(
    /** The proof of the execution of the parity circuit. */
    proof, 
    /** The circuit's verification key */
    verificationKey, 
    /** The vk path in the vk tree*/
    vkPath, 
    /** The public inputs of the parity circuit. */
    publicInputs) {
        this.proof = proof;
        this.verificationKey = verificationKey;
        this.vkPath = vkPath;
        this.publicInputs = publicInputs;
    }
    toBuffer() {
        return serializeToBuffer(...RootParityInput.getFields(this));
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    static from(fields) {
        return new RootParityInput(...RootParityInput.getFields(fields));
    }
    static getFields(fields) {
        return [fields.proof, fields.verificationKey, fields.vkPath, fields.publicInputs];
    }
    static fromBuffer(buffer, expectedSize) {
        const reader = BufferReader.asReader(buffer);
        return new RootParityInput(RecursiveProof.fromBuffer(reader, expectedSize), reader.readObject(VerificationKeyAsFields), reader.readArray(VK_TREE_HEIGHT, Fr), reader.readObject(ParityPublicInputs));
    }
    static fromString(str, expectedSize) {
        return RootParityInput.fromBuffer(Buffer.from(str, 'hex'), expectedSize);
    }
}
