// @aztec/circuits.js/dest/constants.gen.js
/* eslint-disable */
// GENERATED FILE - DO NOT EDIT, RUN yarn remake-constants
export const MAX_FIELD_VALUE = 21888242871839275222246405745257275088548364400416034343698204186575808495616n;
export const ARGS_LENGTH = 16;
export const MAX_NOTE_HASHES_PER_CALL = 16;
export const MAX_NULLIFIERS_PER_CALL = 16;
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL = 4;
export const MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL = 16;
export const MAX_L2_TO_L1_MSGS_PER_CALL = 2;
export const MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL = 32;
export const MAX_PUBLIC_DATA_READS_PER_CALL = 32;
export const MAX_NOTE_HASH_READ_REQUESTS_PER_CALL = 16;
export const MAX_NULLIFIER_READ_REQUESTS_PER_CALL = 16;
export const MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_CALL = 16;
export const MAX_L1_TO_L2_MSG_READ_REQUESTS_PER_CALL = 16;
export const MAX_KEY_VALIDATION_REQUESTS_PER_CALL = 16;
export const MAX_NOTE_ENCRYPTED_LOGS_PER_CALL = 16;
export const MAX_ENCRYPTED_LOGS_PER_CALL = 4;
export const MAX_UNENCRYPTED_LOGS_PER_CALL = 4;
export const ARCHIVE_HEIGHT = 16;
export const VK_TREE_HEIGHT = 6;
export const PROTOCOL_CONTRACT_TREE_HEIGHT = 3;
export const FUNCTION_TREE_HEIGHT = 5;
export const NOTE_HASH_TREE_HEIGHT = 32;
export const PUBLIC_DATA_TREE_HEIGHT = 40;
export const NULLIFIER_TREE_HEIGHT = 20;
export const L1_TO_L2_MSG_TREE_HEIGHT = 16;
export const ARTIFACT_FUNCTION_TREE_MAX_HEIGHT = 5;
export const NULLIFIER_TREE_ID = 0;
export const NOTE_HASH_TREE_ID = 1;
export const PUBLIC_DATA_TREE_ID = 2;
export const L1_TO_L2_MESSAGE_TREE_ID = 3;
export const ARCHIVE_TREE_ID = 4;
export const NOTE_HASH_SUBTREE_HEIGHT = 6;
export const NULLIFIER_SUBTREE_HEIGHT = 6;
export const PUBLIC_DATA_SUBTREE_HEIGHT = 6;
export const L1_TO_L2_MSG_SUBTREE_HEIGHT = 4;
export const NOTE_HASH_SUBTREE_SIBLING_PATH_LENGTH = 26;
export const NULLIFIER_SUBTREE_SIBLING_PATH_LENGTH = 14;
export const PUBLIC_DATA_SUBTREE_SIBLING_PATH_LENGTH = 34;
export const L1_TO_L2_MSG_SUBTREE_SIBLING_PATH_LENGTH = 12;
export const MAX_NOTE_HASHES_PER_TX = 64;
export const MAX_NULLIFIERS_PER_TX = 64;
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX = 8;
export const MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX = 32;
export const PROTOCOL_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 1;
export const MAX_TOTAL_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 64;
export const MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 63;
export const MAX_PUBLIC_DATA_READS_PER_TX = 64;
export const MAX_L2_TO_L1_MSGS_PER_TX = 8;
export const MAX_NOTE_HASH_READ_REQUESTS_PER_TX = 64;
export const MAX_NULLIFIER_READ_REQUESTS_PER_TX = 64;
export const MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_TX = 64;
export const MAX_L1_TO_L2_MSG_READ_REQUESTS_PER_TX = 64;
export const MAX_KEY_VALIDATION_REQUESTS_PER_TX = 64;
export const MAX_NOTE_ENCRYPTED_LOGS_PER_TX = 64;
export const MAX_ENCRYPTED_LOGS_PER_TX = 8;
export const MAX_UNENCRYPTED_LOGS_PER_TX = 8;
export const MAX_PUBLIC_DATA_HINTS = 128;
export const NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP = 16;
export const EMPTY_NESTED_INDEX = 0;
export const PRIVATE_KERNEL_EMPTY_INDEX = 1;
export const PRIVATE_KERNEL_INIT_INDEX = 2;
export const PRIVATE_KERNEL_INNER_INDEX = 3;
export const PRIVATE_KERNEL_TAIL_INDEX = 4;
export const PRIVATE_KERNEL_TAIL_TO_PUBLIC_INDEX = 5;
export const PUBLIC_KERNEL_MERGE_INDEX = 6;
export const PUBLIC_KERNEL_TAIL_INDEX = 7;
export const PUBLIC_KERNEL_INNER_INDEX = 8;
export const BASE_PARITY_INDEX = 10;
export const ROOT_PARITY_INDEX = 11;
export const BASE_ROLLUP_INDEX = 12;
export const MERGE_ROLLUP_INDEX = 13;
export const BLOCK_ROOT_ROLLUP_INDEX = 14;
export const BLOCK_MERGE_ROLLUP_INDEX = 15;
export const ROOT_ROLLUP_INDEX = 16;
export const BLOCK_ROOT_ROLLUP_EMPTY_INDEX = 17;
export const TUBE_INDEX = 18;
export const PRIVATE_KERNEL_RESET_INDEX = 20;
export const FUNCTION_SELECTOR_NUM_BYTES = 4;
export const INITIALIZATION_SLOT_SEPARATOR = 1000000000;
export const INITIAL_L2_BLOCK_NUM = 1;
export const BLOB_SIZE_IN_BYTES = 126976;
export const ETHEREUM_SLOT_DURATION = 12;
export const AZTEC_SLOT_DURATION = 24;
export const AZTEC_EPOCH_DURATION = 16;
export const AZTEC_TARGET_COMMITTEE_SIZE = 48;
export const AZTEC_EPOCH_PROOF_CLAIM_WINDOW_IN_L2_SLOTS = 13;
export const GENESIS_ARCHIVE_ROOT = 8142738430000951296386584486068033372964809139261822027365426310856631083550n;
export const FEE_JUICE_INITIAL_MINT = 20000000000;
export const PUBLIC_DISPATCH_SELECTOR = 3578010381;
export const MAX_PACKED_PUBLIC_BYTECODE_SIZE_IN_FIELDS = 20000;
export const MAX_PACKED_BYTECODE_SIZE_PER_PRIVATE_FUNCTION_IN_FIELDS = 3000;
export const MAX_PACKED_BYTECODE_SIZE_PER_UNCONSTRAINED_FUNCTION_IN_FIELDS = 3000;
export const REGISTERER_PRIVATE_FUNCTION_BROADCASTED_ADDITIONAL_FIELDS = 19;
export const REGISTERER_UNCONSTRAINED_FUNCTION_BROADCASTED_ADDITIONAL_FIELDS = 12;
export const REGISTERER_CONTRACT_CLASS_REGISTERED_MAGIC_VALUE = 11121068431693264234253912047066709627593769337094408533543930778360n;
export const REGISTERER_PRIVATE_FUNCTION_BROADCASTED_MAGIC_VALUE = 2889881020989534926461066592611988634597302675057895885580456197069n;
export const REGISTERER_UNCONSTRAINED_FUNCTION_BROADCASTED_MAGIC_VALUE = 24399338136397901754495080759185489776044879232766421623673792970137n;
export const DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE = 14061769416655647708490531650437236735160113654556896985372298487345n;
export const DEFAULT_GAS_LIMIT = 1000000000;
export const DEFAULT_TEARDOWN_GAS_LIMIT = 100000000;
export const MAX_L2_GAS_PER_ENQUEUED_CALL = 6000000;
export const DEFAULT_MAX_FEE_PER_GAS = 10;
export const DEFAULT_INCLUSION_FEE = 0;
export const DA_BYTES_PER_FIELD = 32;
export const DA_GAS_PER_BYTE = 16;
export const FIXED_DA_GAS = 512;
export const FIXED_L2_GAS = 512;
export const FIXED_AVM_STARTUP_L2_GAS = 1024;
export const L2_GAS_DISTRIBUTED_STORAGE_PREMIUM = 1024;
export const L2_GAS_PER_READ_MERKLE_HASH = 30;
export const L2_GAS_PER_WRITE_MERKLE_HASH = 40;
export const L2_GAS_PER_PUBLIC_DATA_UPDATE = 2624;
export const L2_GAS_PER_NOTE_HASH = 2304;
export const L2_GAS_PER_NULLIFIER = 2624;
export const L2_GAS_PER_PUBLIC_DATA_READ = 1200;
export const L2_GAS_PER_NOTE_HASH_READ_REQUEST = 960;
export const L2_GAS_PER_NULLIFIER_READ_REQUEST = 1200;
export const L2_GAS_PER_L1_TO_L2_MSG_READ_REQUEST = 480;
export const L2_GAS_PER_LOG_BYTE = 4;
export const L2_GAS_PER_L2_TO_L1_MSG = 200;
export const MAX_PROTOCOL_CONTRACTS = 7;
export const CANONICAL_AUTH_REGISTRY_ADDRESS = 1;
export const DEPLOYER_CONTRACT_ADDRESS = 2;
export const REGISTERER_CONTRACT_ADDRESS = 3;
export const MULTI_CALL_ENTRYPOINT_ADDRESS = 4;
export const FEE_JUICE_ADDRESS = 5;
export const ROUTER_ADDRESS = 6;
export const DEFAULT_NPK_M_X = 582240093077765400562621227108555700500271598878376310175765873770292988861n;
export const DEFAULT_NPK_M_Y = 10422444662424639723529825114205836958711284159673861467999592572974769103684n;
export const DEFAULT_IVPK_M_X = 339708709767762472786445938838804872781183545349360029270386718856175781484n;
export const DEFAULT_IVPK_M_Y = 12719619215050539905199178334954929730355853796706924300730604757520758976849n;
export const DEFAULT_OVPK_M_X = 12212787719617305570587928860288475454328008955283046946846066128763901043335n;
export const DEFAULT_OVPK_M_Y = 3646747884782549389807830220601404629716007431341772952958971658285958854707n;
export const DEFAULT_TPK_M_X = 728059161893070741164607238299536939695876538801885465230641192969135857403n;
export const DEFAULT_TPK_M_Y = 14575718736702206050102425029229426215631664471161015518982549597389390371695n;
export const AZTEC_ADDRESS_LENGTH = 1;
export const GAS_FEES_LENGTH = 2;
export const GAS_LENGTH = 2;
export const GAS_SETTINGS_LENGTH = 7;
export const CALL_CONTEXT_LENGTH = 4;
export const CONTENT_COMMITMENT_LENGTH = 4;
export const CONTRACT_INSTANCE_LENGTH = 16;
export const CONTRACT_STORAGE_READ_LENGTH = 3;
export const CONTRACT_STORAGE_UPDATE_REQUEST_LENGTH = 3;
export const ETH_ADDRESS_LENGTH = 1;
export const FUNCTION_DATA_LENGTH = 2;
export const FUNCTION_LEAF_PREIMAGE_LENGTH = 5;
export const GLOBAL_VARIABLES_LENGTH = 9;
export const APPEND_ONLY_TREE_SNAPSHOT_LENGTH = 2;
export const L1_TO_L2_MESSAGE_LENGTH = 6;
export const L2_TO_L1_MESSAGE_LENGTH = 3;
export const SCOPED_L2_TO_L1_MESSAGE_LENGTH = 4;
export const MAX_BLOCK_NUMBER_LENGTH = 2;
export const KEY_VALIDATION_REQUEST_LENGTH = 4;
export const KEY_VALIDATION_REQUEST_AND_GENERATOR_LENGTH = 5;
export const SCOPED_KEY_VALIDATION_REQUEST_AND_GENERATOR_LENGTH = 6;
export const PARTIAL_STATE_REFERENCE_LENGTH = 6;
export const READ_REQUEST_LENGTH = 2;
export const TREE_LEAF_READ_REQUEST_LENGTH = 2;
export const LOG_HASH_LENGTH = 3;
export const SCOPED_LOG_HASH_LENGTH = 4;
export const ENCRYPTED_LOG_HASH_LENGTH = 4;
export const SCOPED_ENCRYPTED_LOG_HASH_LENGTH = 5;
export const NOTE_LOG_HASH_LENGTH = 4;
export const NOTE_HASH_LENGTH = 2;
export const SCOPED_NOTE_HASH_LENGTH = 3;
export const NULLIFIER_LENGTH = 3;
export const SCOPED_NULLIFIER_LENGTH = 4;
export const PUBLIC_CALL_STACK_ITEM_COMPRESSED_LENGTH = 12;
export const PRIVATE_CALL_REQUEST_LENGTH = 8;
export const PUBLIC_CALL_REQUEST_LENGTH = 6;
export const PUBLIC_INNER_CALL_REQUEST_LENGTH = 13;
export const ROLLUP_VALIDATION_REQUESTS_LENGTH = 2;
export const STATE_REFERENCE_LENGTH = 8;
export const TX_CONTEXT_LENGTH = 9;
export const TX_REQUEST_LENGTH = 13;
export const TOTAL_FEES_LENGTH = 1;
export const HEADER_LENGTH = 24;
export const PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH = 501;
export const PUBLIC_CIRCUIT_PUBLIC_INPUTS_LENGTH = 674;
export const PRIVATE_CONTEXT_INPUTS_LENGTH = 38;
export const PUBLIC_CONTEXT_INPUTS_LENGTH = 41;
export const FEE_RECIPIENT_LENGTH = 2;
export const AGGREGATION_OBJECT_LENGTH = 16;
export const SCOPED_READ_REQUEST_LEN = 3;
export const PUBLIC_DATA_READ_LENGTH = 3;
export const PRIVATE_VALIDATION_REQUESTS_LENGTH = 772;
export const NUM_PUBLIC_VALIDATION_REQUEST_ARRAYS = 5;
export const PUBLIC_VALIDATION_REQUESTS_LENGTH = 834;
export const PUBLIC_DATA_UPDATE_REQUEST_LENGTH = 3;
export const COMBINED_ACCUMULATED_DATA_LENGTH = 610;
export const COMBINED_CONSTANT_DATA_LENGTH = 44;
export const PRIVATE_ACCUMULATED_DATA_LENGTH = 1064;
export const PRIVATE_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH = 1888;
export const PUBLIC_ACCUMULATED_DATA_LENGTH = 1055;
export const NUM_PUBLIC_ACCUMULATED_DATA_ARRAYS = 8;
export const PUBLIC_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH = 2997;
export const VM_CIRCUIT_PUBLIC_INPUTS_LENGTH = 2374;
export const KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH = 664;
export const CONSTANT_ROLLUP_DATA_LENGTH = 13;
export const BASE_OR_MERGE_PUBLIC_INPUTS_LENGTH = 30;
export const BLOCK_ROOT_OR_BLOCK_MERGE_PUBLIC_INPUTS_LENGTH = 60;
export const ROOT_ROLLUP_PUBLIC_INPUTS_LENGTH = 44;
export const GET_NOTES_ORACLE_RETURN_LENGTH = 674;
export const NOTE_HASHES_NUM_BYTES_PER_BASE_ROLLUP = 2048;
export const NULLIFIERS_NUM_BYTES_PER_BASE_ROLLUP = 2048;
export const PUBLIC_DATA_WRITES_NUM_BYTES_PER_BASE_ROLLUP = 4096;
export const CONTRACTS_NUM_BYTES_PER_BASE_ROLLUP = 32;
export const CONTRACT_DATA_NUM_BYTES_PER_BASE_ROLLUP = 64;
export const CONTRACT_DATA_NUM_BYTES_PER_BASE_ROLLUP_UNPADDED = 52;
export const L2_TO_L1_MSGS_NUM_BYTES_PER_BASE_ROLLUP = 256;
export const LOGS_HASHES_NUM_BYTES_PER_BASE_ROLLUP = 64;
export const NUM_MSGS_PER_BASE_PARITY = 4;
export const NUM_BASE_PARITY_PER_ROOT_PARITY = 4;
export const RECURSIVE_PROOF_LENGTH = 463;
export const NESTED_RECURSIVE_PROOF_LENGTH = 463;
export const TUBE_PROOF_LENGTH = 463;
export const HONK_VERIFICATION_KEY_LENGTH_IN_FIELDS = 128;
export const CLIENT_IVC_VERIFICATION_KEY_LENGTH_IN_FIELDS = 145;
export const AVM_VERIFICATION_KEY_LENGTH_IN_FIELDS = 86;
export const AVM_PROOF_LENGTH_IN_FIELDS = 3949;
export const AVM_PUBLIC_COLUMN_MAX_SIZE = 1024;
export const AVM_PUBLIC_INPUTS_FLATTENED_SIZE = 2722;
export const MEM_TAG_FF = 0;
export const MEM_TAG_U1 = 1;
export const MEM_TAG_U8 = 2;
export const MEM_TAG_U16 = 3;
export const MEM_TAG_U32 = 4;
export const MEM_TAG_U64 = 5;
export const MEM_TAG_U128 = 6;
export const SENDER_KERNEL_INPUTS_COL_OFFSET = 0;
export const ADDRESS_KERNEL_INPUTS_COL_OFFSET = 1;
export const FUNCTION_SELECTOR_KERNEL_INPUTS_COL_OFFSET = 2;
export const IS_STATIC_CALL_KERNEL_INPUTS_COL_OFFSET = 3;
export const CHAIN_ID_KERNEL_INPUTS_COL_OFFSET = 4;
export const VERSION_KERNEL_INPUTS_COL_OFFSET = 5;
export const BLOCK_NUMBER_KERNEL_INPUTS_COL_OFFSET = 6;
export const TIMESTAMP_KERNEL_INPUTS_COL_OFFSET = 7;
export const FEE_PER_DA_GAS_KERNEL_INPUTS_COL_OFFSET = 8;
export const FEE_PER_L2_GAS_KERNEL_INPUTS_COL_OFFSET = 9;
export const DA_START_GAS_KERNEL_INPUTS_COL_OFFSET = 10;
export const L2_START_GAS_KERNEL_INPUTS_COL_OFFSET = 11;
export const DA_END_GAS_KERNEL_INPUTS_COL_OFFSET = 12;
export const L2_END_GAS_KERNEL_INPUTS_COL_OFFSET = 13;
export const TRANSACTION_FEE_KERNEL_INPUTS_COL_OFFSET = 14;
export const START_NOTE_HASH_EXISTS_WRITE_OFFSET = 0;
export const START_NULLIFIER_EXISTS_OFFSET = 16;
export const START_NULLIFIER_NON_EXISTS_OFFSET = 32;
export const START_L1_TO_L2_MSG_EXISTS_WRITE_OFFSET = 48;
export const START_SSTORE_WRITE_OFFSET = 64;
export const START_SLOAD_WRITE_OFFSET = 96;
export const START_EMIT_NOTE_HASH_WRITE_OFFSET = 128;
export const START_EMIT_NULLIFIER_WRITE_OFFSET = 144;
export const START_EMIT_L2_TO_L1_MSG_WRITE_OFFSET = 160;
export const START_EMIT_UNENCRYPTED_LOG_WRITE_OFFSET = 162;
export const AVM_ADD_BASE_L2_GAS = 32;
export const AVM_SUB_BASE_L2_GAS = 32;
export const AVM_MUL_BASE_L2_GAS = 33;
export const AVM_DIV_BASE_L2_GAS = 43;
export const AVM_FDIV_BASE_L2_GAS = 32;
export const AVM_EQ_BASE_L2_GAS = 32;
export const AVM_LT_BASE_L2_GAS = 64;
export const AVM_LTE_BASE_L2_GAS = 64;
export const AVM_AND_BASE_L2_GAS = 33;
export const AVM_OR_BASE_L2_GAS = 33;
export const AVM_XOR_BASE_L2_GAS = 33;
export const AVM_NOT_BASE_L2_GAS = 27;
export const AVM_SHL_BASE_L2_GAS = 32;
export const AVM_SHR_BASE_L2_GAS = 32;
export const AVM_CAST_BASE_L2_GAS = 30;
export const AVM_GETENVVAR_BASE_L2_GAS = 20;
export const AVM_CALLDATACOPY_BASE_L2_GAS = 29;
export const AVM_JUMP_BASE_L2_GAS = 12;
export const AVM_JUMPI_BASE_L2_GAS = 18;
export const AVM_INTERNALCALL_BASE_L2_GAS = 18;
export const AVM_INTERNALRETURN_BASE_L2_GAS = 18;
export const AVM_SET_BASE_L2_GAS = 18;
export const AVM_MOV_BASE_L2_GAS = 23;
export const AVM_SLOAD_BASE_L2_GAS = 1218;
export const AVM_SSTORE_BASE_L2_GAS = 2642;
export const AVM_NOTEHASHEXISTS_BASE_L2_GAS = 986;
export const AVM_EMITNOTEHASH_BASE_L2_GAS = 2330;
export const AVM_NULLIFIEREXISTS_BASE_L2_GAS = 1226;
export const AVM_EMITNULLIFIER_BASE_L2_GAS = 2650;
export const AVM_L1TOL2MSGEXISTS_BASE_L2_GAS = 506;
export const AVM_GETCONTRACTINSTANCE_BASE_L2_GAS = 1248;
export const AVM_EMITUNENCRYPTEDLOG_BASE_L2_GAS = 18;
export const AVM_SENDL2TOL1MSG_BASE_L2_GAS = 226;
export const AVM_CALL_BASE_L2_GAS = 2445;
export const AVM_STATICCALL_BASE_L2_GAS = 2445;
export const AVM_RETURN_BASE_L2_GAS = 28;
export const AVM_REVERT_BASE_L2_GAS = 28;
export const AVM_DEBUGLOG_BASE_L2_GAS = 12;
export const AVM_POSEIDON2_BASE_L2_GAS = 78;
export const AVM_SHA256COMPRESSION_BASE_L2_GAS = 261;
export const AVM_KECCAKF1600_BASE_L2_GAS = 300;
export const AVM_ECADD_BASE_L2_GAS = 62;
export const AVM_MSM_BASE_L2_GAS = 1000;
export const AVM_TORADIXLE_BASE_L2_GAS = 46;
export const AVM_CALLDATACOPY_DYN_L2_GAS = 6;
export const AVM_EMITUNENCRYPTEDLOG_DYN_L2_GAS = 146;
export const AVM_CALL_DYN_L2_GAS = 4;
export const AVM_STATICCALL_DYN_L2_GAS = 4;
export const AVM_RETURN_DYN_L2_GAS = 6;
export const AVM_REVERT_DYN_L2_GAS = 6;
export const AVM_MSM_DYN_L2_GAS = 100;
export const AVM_TORADIXLE_DYN_L2_GAS = 20;
export const AVM_SSTORE_BASE_DA_GAS = 512;
export const AVM_EMITNOTEHASH_BASE_DA_GAS = 512;
export const AVM_EMITNULLIFIER_BASE_DA_GAS = 512;
export const AVM_SENDL2TOL1MSG_BASE_DA_GAS = 512;
export const AVM_EMITUNENCRYPTEDLOG_DYN_DA_GAS = 512;
export const PROOF_TYPE_PLONK = 0;
export const PROOF_TYPE_HONK = 1;
export const PROOF_TYPE_OINK = 2;
export const PROOF_TYPE_PG = 3;
export const PROOF_TYPE_AVM = 4;
export var GeneratorIndex;
(function (GeneratorIndex) {
    GeneratorIndex[GeneratorIndex["NOTE_HASH"] = 1] = "NOTE_HASH";
    GeneratorIndex[GeneratorIndex["NOTE_HASH_NONCE"] = 2] = "NOTE_HASH_NONCE";
    GeneratorIndex[GeneratorIndex["UNIQUE_NOTE_HASH"] = 3] = "UNIQUE_NOTE_HASH";
    GeneratorIndex[GeneratorIndex["SILOED_NOTE_HASH"] = 4] = "SILOED_NOTE_HASH";
    GeneratorIndex[GeneratorIndex["MESSAGE_NULLIFIER"] = 5] = "MESSAGE_NULLIFIER";
    GeneratorIndex[GeneratorIndex["INITIALIZATION_NULLIFIER"] = 6] = "INITIALIZATION_NULLIFIER";
    GeneratorIndex[GeneratorIndex["OUTER_NULLIFIER"] = 7] = "OUTER_NULLIFIER";
    GeneratorIndex[GeneratorIndex["PUBLIC_DATA_READ"] = 8] = "PUBLIC_DATA_READ";
    GeneratorIndex[GeneratorIndex["PUBLIC_DATA_UPDATE_REQUEST"] = 9] = "PUBLIC_DATA_UPDATE_REQUEST";
    GeneratorIndex[GeneratorIndex["FUNCTION_DATA"] = 10] = "FUNCTION_DATA";
    GeneratorIndex[GeneratorIndex["FUNCTION_LEAF"] = 11] = "FUNCTION_LEAF";
    GeneratorIndex[GeneratorIndex["CONTRACT_DEPLOYMENT_DATA"] = 12] = "CONTRACT_DEPLOYMENT_DATA";
    GeneratorIndex[GeneratorIndex["CONSTRUCTOR"] = 13] = "CONSTRUCTOR";
    GeneratorIndex[GeneratorIndex["CONSTRUCTOR_ARGS"] = 14] = "CONSTRUCTOR_ARGS";
    GeneratorIndex[GeneratorIndex["CONTRACT_ADDRESS_V1"] = 15] = "CONTRACT_ADDRESS_V1";
    GeneratorIndex[GeneratorIndex["CONTRACT_LEAF"] = 16] = "CONTRACT_LEAF";
    GeneratorIndex[GeneratorIndex["CALL_CONTEXT"] = 17] = "CALL_CONTEXT";
    GeneratorIndex[GeneratorIndex["CALL_STACK_ITEM"] = 18] = "CALL_STACK_ITEM";
    GeneratorIndex[GeneratorIndex["CALL_STACK_ITEM_2"] = 19] = "CALL_STACK_ITEM_2";
    GeneratorIndex[GeneratorIndex["SECRET_HASH"] = 20] = "SECRET_HASH";
    GeneratorIndex[GeneratorIndex["L2_TO_L1_MSG"] = 21] = "L2_TO_L1_MSG";
    GeneratorIndex[GeneratorIndex["TX_CONTEXT"] = 22] = "TX_CONTEXT";
    GeneratorIndex[GeneratorIndex["PUBLIC_LEAF_INDEX"] = 23] = "PUBLIC_LEAF_INDEX";
    GeneratorIndex[GeneratorIndex["PUBLIC_DATA_LEAF"] = 24] = "PUBLIC_DATA_LEAF";
    GeneratorIndex[GeneratorIndex["SIGNED_TX_REQUEST"] = 25] = "SIGNED_TX_REQUEST";
    GeneratorIndex[GeneratorIndex["GLOBAL_VARIABLES"] = 26] = "GLOBAL_VARIABLES";
    GeneratorIndex[GeneratorIndex["PARTIAL_ADDRESS"] = 27] = "PARTIAL_ADDRESS";
    GeneratorIndex[GeneratorIndex["BLOCK_HASH"] = 28] = "BLOCK_HASH";
    GeneratorIndex[GeneratorIndex["SIDE_EFFECT"] = 29] = "SIDE_EFFECT";
    GeneratorIndex[GeneratorIndex["FEE_PAYLOAD"] = 30] = "FEE_PAYLOAD";
    GeneratorIndex[GeneratorIndex["COMBINED_PAYLOAD"] = 31] = "COMBINED_PAYLOAD";
    GeneratorIndex[GeneratorIndex["TX_NULLIFIER"] = 32] = "TX_NULLIFIER";
    GeneratorIndex[GeneratorIndex["TX_REQUEST"] = 33] = "TX_REQUEST";
    GeneratorIndex[GeneratorIndex["SIGNATURE_PAYLOAD"] = 34] = "SIGNATURE_PAYLOAD";
    GeneratorIndex[GeneratorIndex["VK"] = 41] = "VK";
    GeneratorIndex[GeneratorIndex["PRIVATE_CIRCUIT_PUBLIC_INPUTS"] = 42] = "PRIVATE_CIRCUIT_PUBLIC_INPUTS";
    GeneratorIndex[GeneratorIndex["PUBLIC_CIRCUIT_PUBLIC_INPUTS"] = 43] = "PUBLIC_CIRCUIT_PUBLIC_INPUTS";
    GeneratorIndex[GeneratorIndex["FUNCTION_ARGS"] = 44] = "FUNCTION_ARGS";
    GeneratorIndex[GeneratorIndex["AUTHWIT_INNER"] = 45] = "AUTHWIT_INNER";
    GeneratorIndex[GeneratorIndex["AUTHWIT_OUTER"] = 46] = "AUTHWIT_OUTER";
    GeneratorIndex[GeneratorIndex["AUTHWIT_NULLIFIER"] = 47] = "AUTHWIT_NULLIFIER";
    GeneratorIndex[GeneratorIndex["NSK_M"] = 48] = "NSK_M";
    GeneratorIndex[GeneratorIndex["IVSK_M"] = 49] = "IVSK_M";
    GeneratorIndex[GeneratorIndex["OVSK_M"] = 50] = "OVSK_M";
    GeneratorIndex[GeneratorIndex["TSK_M"] = 51] = "TSK_M";
    GeneratorIndex[GeneratorIndex["PUBLIC_KEYS_HASH"] = 52] = "PUBLIC_KEYS_HASH";
    GeneratorIndex[GeneratorIndex["NOTE_NULLIFIER"] = 53] = "NOTE_NULLIFIER";
    GeneratorIndex[GeneratorIndex["NOTE_HIDING_POINT"] = 54] = "NOTE_HIDING_POINT";
    GeneratorIndex[GeneratorIndex["SYMMETRIC_KEY"] = 55] = "SYMMETRIC_KEY";
})(GeneratorIndex || (GeneratorIndex = {}));
