// @aztec/foundation/dest/aztec-address/index.js
import { inspect } from 'util';
import { Fr, fromBuffer } from '../fields/index.js';
import { FieldReader } from '../serialize/index.js';
import { TypeRegistry } from '../serialize/type_registry.js';
/**
 * AztecAddress represents a 32-byte address in the Aztec Protocol.
 * It provides methods to create, manipulate, and compare addresses.
 * The maximum value of an address is determined by the field modulus and all instances of AztecAddress.
 * It should have a value less than or equal to this max value.
 * This class also provides helper functions to convert addresses from strings, buffers, and other formats.
 */
export class AztecAddress extends Fr {
    constructor(buffer) {
        if (buffer.length !== 32) {
            throw new Error(`Invalid AztecAddress length ${buffer.length}.`);
        }
        super(buffer);
    }
    [inspect.custom]() {
        return `AztecAddress<${this.toString()}>`;
    }
    static zero() {
        return AztecAddress.ZERO;
    }
    static fromBuffer(buffer) {
        return fromBuffer(buffer, AztecAddress);
    }
    static fromField(fr) {
        return new AztecAddress(fr.toBuffer());
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return AztecAddress.fromField(reader.readField());
    }
    static fromBigInt(value) {
        return AztecAddress.fromField(new Fr(value));
    }
    static fromString(buf) {
        const buffer = Buffer.from(buf.replace(/^0x/i, ''), 'hex');
        return new AztecAddress(buffer);
    }
    static random() {
        return new AztecAddress(super.random().toBuffer());
    }
    toJSON() {
        return {
            type: 'AztecAddress',
            value: this.toString(),
        };
    }
}
AztecAddress.ZERO = new AztecAddress(Buffer.alloc(32));
// For deserializing JSON.
TypeRegistry.register('AztecAddress', AztecAddress);
