// @aztec/circuits.js/dest/tests/fixtures.js
import { loadContractArtifact } from '@aztec/types/abi';
import { readFileSync } from 'fs';
import { dirname, resolve } from 'path';
import { fileURLToPath } from 'url';
// Copied from the build output for the contract `Benchmarking` in noir-contracts
export function getBenchmarkContractArtifact() {
    const path = getPathToFixture('Benchmarking.test.json');
    const content = JSON.parse(readFileSync(path).toString());
    return loadContractArtifact(content);
}
// Copied from the build output for the contract `Benchmarking` in noir-contracts
export function getTestContractArtifact() {
    const path = getPathToFixture('Benchmarking.test.json');
    const content = JSON.parse(readFileSync(path).toString());
    return loadContractArtifact(content);
}
// Copied from the test 'registers a new contract class' in end-to-end/src/e2e_deploy_contract.test.ts
export function getSampleContractClassRegisteredEventPayload() {
    const path = getPathToFixture('ContractClassRegisteredEventData.hex');
    return Buffer.from(readFileSync(path).toString(), 'hex');
}
// This is generated with code like this:
//   const tx = await StatefulTestContract.deploy(wallet, owner, owner, 42).send({ universalDeploy: true }).wait();
//   const logs = await pxe.getUnencryptedLogs({ txHash: tx.txHash });
//   const logData = logs.logs[0].log.data;
//   writeTestData('yarn-project/circuits.js/fixtures/ContractInstanceDeployedEventData.hex', logData);
export function getSampleContractInstanceDeployedEventPayload() {
    const path = getPathToFixture('ContractInstanceDeployedEventData.hex');
    return Buffer.from(readFileSync(path).toString(), 'hex');
}
// Generated from end-to-end/src/e2e_deploy_contract.test.ts with AZTEC_GENERATE_TEST_DATA
export function getSamplePrivateFunctionBroadcastedEventPayload() {
    const path = getPathToFixture('PrivateFunctionBroadcastedEventData.hex');
    return Buffer.from(readFileSync(path).toString(), 'hex');
}
// Generated from end-to-end/src/e2e_deploy_contract.test.ts with AZTEC_GENERATE_TEST_DATA
export function getSampleUnconstrainedFunctionBroadcastedEventPayload() {
    const path = getPathToFixture('UnconstrainedFunctionBroadcastedEventData.hex');
    return Buffer.from(readFileSync(path).toString(), 'hex');
}
function getPathToFixture(name) {
    return resolve(dirname(fileURLToPath(import.meta.url)), `../../fixtures/${name}`);
}
