// @aztec/aztec.js/dest/account_manager/deploy_account_method.js
var _DeployAccountMethod_authWitnessProvider, _DeployAccountMethod_feePaymentArtifact;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { FunctionSelector, encodeArguments, getFunctionArtifact, } from '@aztec/foundation/abi';
import { Contract } from '../contract/contract.js';
import { DeployMethod } from '../contract/deploy_method.js';
import { EntrypointPayload, computeCombinedPayloadHash } from '../entrypoint/payload.js';
/**
 * Contract interaction for deploying an account contract. Handles fee preparation and contract initialization.
 */
export class DeployAccountMethod extends DeployMethod {
    constructor(authWitnessProvider, publicKeys, wallet, artifact, args = [], constructorNameOrArtifact, feePaymentNameOrArtifact) {
        super(publicKeys, wallet, artifact, (address, wallet) => Contract.at(address, artifact, wallet), args, constructorNameOrArtifact);
        _DeployAccountMethod_authWitnessProvider.set(this, void 0);
        _DeployAccountMethod_feePaymentArtifact.set(this, void 0);
        __classPrivateFieldSet(this, _DeployAccountMethod_authWitnessProvider, authWitnessProvider, "f");
        __classPrivateFieldSet(this, _DeployAccountMethod_feePaymentArtifact, typeof feePaymentNameOrArtifact === 'string'
            ? getFunctionArtifact(artifact, feePaymentNameOrArtifact)
            : feePaymentNameOrArtifact, "f");
    }
    async getInitializeFunctionCalls(options) {
        const exec = await super.getInitializeFunctionCalls(options);
        if (options.fee && __classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f")) {
            const { address } = this.getInstance();
            const emptyAppPayload = EntrypointPayload.fromAppExecution([]);
            const feePayload = await EntrypointPayload.fromFeeOptions(address, options?.fee);
            exec.calls.push({
                name: __classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f").name,
                to: address,
                args: encodeArguments(__classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f"), [emptyAppPayload, feePayload, false]),
                selector: FunctionSelector.fromNameAndParameters(__classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f").name, __classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f").parameters),
                type: __classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f").functionType,
                isStatic: __classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f").isStatic,
                returnTypes: __classPrivateFieldGet(this, _DeployAccountMethod_feePaymentArtifact, "f").returnTypes,
            });
            exec.authWitnesses ?? (exec.authWitnesses = []);
            exec.packedArguments ?? (exec.packedArguments = []);
            exec.authWitnesses.push(await __classPrivateFieldGet(this, _DeployAccountMethod_authWitnessProvider, "f").createAuthWit(computeCombinedPayloadHash(emptyAppPayload, feePayload)));
            exec.packedArguments.push(...emptyAppPayload.packedArguments);
            exec.packedArguments.push(...feePayload.packedArguments);
        }
        return exec;
    }
}
_DeployAccountMethod_authWitnessProvider = new WeakMap(), _DeployAccountMethod_feePaymentArtifact = new WeakMap();
