// @aztec/aztec.js/dest/deployment/contract_deployer.js
import { PublicKeys } from '@aztec/circuits.js';
import { Contract } from '../contract/contract.js';
import { DeployMethod } from '../contract/deploy_method.js';
/**
 * A class for deploying contract.
 * @remarks Keeping this around even though we have Aztec.nr contract types because it can be useful for non-TS users.
 */
export class ContractDeployer {
    constructor(artifact, wallet, publicKeys, constructorName) {
        this.artifact = artifact;
        this.wallet = wallet;
        this.publicKeys = publicKeys;
        this.constructorName = constructorName;
    }
    /**
     * Deploy a contract using the provided ABI and constructor arguments.
     * This function creates a new DeployMethod instance that can be used to send deployment transactions
     * and query deployment status. The method accepts any number of constructor arguments, which will
     * be passed to the contract's constructor during deployment.
     *
     * @param args - The constructor arguments for the contract being deployed.
     * @returns A DeployMethod instance configured with the ABI, PXE, and constructor arguments.
     */
    deploy(...args) {
        const postDeployCtor = (address, wallet) => Contract.at(address, this.artifact, wallet);
        return new DeployMethod(this.publicKeys ?? PublicKeys.default(), this.wallet, this.artifact, postDeployCtor, args, this.constructorName);
    }
}
