// @aztec/circuits.js/dest/structs/kernel/private_call_data.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { FUNCTION_TREE_HEIGHT, PROTOCOL_CONTRACT_TREE_HEIGHT } from '../../constants.gen.js';
import { PublicKeys } from '../../types/public_keys.js';
import { MembershipWitness } from '../membership_witness.js';
import { PrivateCircuitPublicInputs } from '../private_circuit_public_inputs.js';
import { VerificationKeyAsFields } from '../verification_key.js';
/**
 * Private call data.
 */
export class PrivateCallData {
    constructor(
    /**
     * Public inputs of the private function circuit.
     */
    publicInputs, 
    /**
     * The verification key for the function being invoked.
     */
    vk, 
    /**
     * Artifact hash of the contract class for this private call.
     */
    contractClassArtifactHash, 
    /**
     * Public bytecode commitment for the contract class for this private call.
     */
    contractClassPublicBytecodeCommitment, 
    /**
     * Public keys hash of the contract instance.
     */
    publicKeys, 
    /**
     * Salted initialization hash of the contract instance.
     */
    saltedInitializationHash, 
    /**
     * The membership witness for the function leaf corresponding to the function being invoked.
     */
    functionLeafMembershipWitness, protocolContractSiblingPath, 
    /**
     * The hash of the ACIR of the function being invoked.
     */
    acirHash) {
        this.publicInputs = publicInputs;
        this.vk = vk;
        this.contractClassArtifactHash = contractClassArtifactHash;
        this.contractClassPublicBytecodeCommitment = contractClassPublicBytecodeCommitment;
        this.publicKeys = publicKeys;
        this.saltedInitializationHash = saltedInitializationHash;
        this.functionLeafMembershipWitness = functionLeafMembershipWitness;
        this.protocolContractSiblingPath = protocolContractSiblingPath;
        this.acirHash = acirHash;
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [
            fields.publicInputs,
            fields.vk,
            fields.contractClassArtifactHash,
            fields.contractClassPublicBytecodeCommitment,
            fields.publicKeys,
            fields.saltedInitializationHash,
            fields.functionLeafMembershipWitness,
            fields.protocolContractSiblingPath,
            fields.acirHash,
        ];
    }
    static from(fields) {
        return new PrivateCallData(...PrivateCallData.getFields(fields));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...PrivateCallData.getFields(this));
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateCallData(reader.readObject(PrivateCircuitPublicInputs), reader.readObject(VerificationKeyAsFields), reader.readObject(Fr), reader.readObject(Fr), reader.readObject(PublicKeys), reader.readObject(Fr), reader.readObject(MembershipWitness.deserializer(FUNCTION_TREE_HEIGHT)), reader.readArray(PROTOCOL_CONTRACT_TREE_HEIGHT, Fr), reader.readObject(Fr));
    }
}
