// @aztec/circuits.js/dest/structs/kernel/private_kernel_reset_circuit_private_inputs.js
import { serializeToBuffer } from '@aztec/foundation/serialize';
export class PrivateKernelResetCircuitPrivateInputsVariants {
    constructor(previousKernel, hints) {
        this.previousKernel = previousKernel;
        this.hints = hints;
    }
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.hints);
    }
}
/**
 * Input to the private kernel circuit - reset call.
 */
export class PrivateKernelResetCircuitPrivateInputs {
    constructor(
    /**
     * The previous kernel data
     */
    previousKernel, hints, dimensions) {
        this.previousKernel = previousKernel;
        this.hints = hints;
        this.dimensions = dimensions;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.hints, this.dimensions);
    }
    trimToSizes() {
        const hints = this.hints.trimToSizes(this.dimensions.NOTE_HASH_PENDING_AMOUNT, this.dimensions.NOTE_HASH_SETTLED_AMOUNT, this.dimensions.NULLIFIER_PENDING_AMOUNT, this.dimensions.NULLIFIER_SETTLED_AMOUNT, this.dimensions.NULLIFIER_KEYS, this.dimensions.TRANSIENT_DATA_AMOUNT);
        return new PrivateKernelResetCircuitPrivateInputsVariants(this.previousKernel, hints);
    }
}
