// @aztec/circuits.js/dest/structs/kernel/private_kernel_empty_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { Header } from '../header.js';
export class PrivateKernelEmptyInputData {
    constructor(header, chainId, version, vkTreeRoot, protocolContractTreeRoot) {
        this.header = header;
        this.chainId = chainId;
        this.version = version;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractTreeRoot = protocolContractTreeRoot;
    }
    toBuffer() {
        return serializeToBuffer(this.header, this.chainId, this.version, this.vkTreeRoot, this.protocolContractTreeRoot);
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromBuffer(buf) {
        const reader = BufferReader.asReader(buf);
        return new PrivateKernelEmptyInputData(reader.readObject(Header), reader.readObject(Fr), reader.readObject(Fr), reader.readObject(Fr), reader.readObject(Fr));
    }
    static fromString(str) {
        return PrivateKernelEmptyInputData.fromBuffer(Buffer.from(str, 'hex'));
    }
    static from(fields) {
        return new PrivateKernelEmptyInputData(fields.header, fields.chainId, fields.version, fields.vkTreeRoot, fields.protocolContractTreeRoot);
    }
}
export class PrivateKernelEmptyInputs {
    constructor(emptyNested, header, chainId, version, vkTreeRoot, protocolContractTreeRoot) {
        this.emptyNested = emptyNested;
        this.header = header;
        this.chainId = chainId;
        this.version = version;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractTreeRoot = protocolContractTreeRoot;
    }
    toBuffer() {
        return serializeToBuffer(this.emptyNested, this.header, this.chainId, this.version, this.vkTreeRoot, this.protocolContractTreeRoot);
    }
    static from(fields) {
        return new PrivateKernelEmptyInputs(fields.emptyNested, fields.header, fields.chainId, fields.version, fields.vkTreeRoot, fields.protocolContractTreeRoot);
    }
}
export class EmptyNestedCircuitInputs {
    toBuffer() {
        return Buffer.alloc(0);
    }
}
export class EmptyNestedData {
    constructor(proof, vk) {
        this.proof = proof;
        this.vk = vk;
    }
    toBuffer() {
        return serializeToBuffer(this.proof, this.vk);
    }
}
