// @aztec/circuits.js/dest/structs/kernel/public_accumulated_data.js
import { makeTuple } from '@aztec/foundation/array';
import { arraySerializedSizeOfNonEmpty } from '@aztec/foundation/collection';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { MAX_ENCRYPTED_LOGS_PER_TX, MAX_L2_TO_L1_MSGS_PER_TX, MAX_NOTE_ENCRYPTED_LOGS_PER_TX, MAX_NOTE_HASHES_PER_TX, MAX_NULLIFIERS_PER_TX, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX, MAX_UNENCRYPTED_LOGS_PER_TX, NUM_PUBLIC_ACCUMULATED_DATA_ARRAYS, } from '../../constants.gen.js';
import { countAccumulatedItems } from '../../utils/index.js';
import { Gas } from '../gas.js';
import { ScopedL2ToL1Message } from '../l2_to_l1_message.js';
import { LogHash, ScopedLogHash } from '../log_hash.js';
import { ScopedNoteHash } from '../note_hash.js';
import { Nullifier } from '../nullifier.js';
import { PublicCallRequest } from '../public_call_request.js';
import { PublicDataUpdateRequest } from '../public_data_update_request.js';
export class PublicAccumulatedData {
    constructor(
    /**
     * The new note hashes made in this transaction.
     */
    noteHashes, 
    /**
     * The new nullifiers made in this transaction.
     */
    nullifiers, 
    /**
     * All the new L2 to L1 messages created in this transaction.
     */
    l2ToL1Msgs, 
    /**
     * Accumulated encrypted note logs hashes from all the previous kernel iterations.
     * Note: Truncated to 31 bytes to fit in Fr.
     */
    noteEncryptedLogsHashes, 
    /**
     * Accumulated encrypted logs hashes from all the previous kernel iterations.
     * Note: Truncated to 31 bytes to fit in Fr.
     */
    encryptedLogsHashes, 
    /**
     * Accumulated unencrypted logs hashes from all the previous kernel iterations.
     * Note: Truncated to 31 bytes to fit in Fr.
     */
    unencryptedLogsHashes, 
    /**
     * All the public data update requests made in this transaction.
     */
    publicDataUpdateRequests, 
    /**
     * Current public call stack.
     */
    publicCallStack, 
    /** Gas used so far by the transaction. */
    gasUsed) {
        this.noteHashes = noteHashes;
        this.nullifiers = nullifiers;
        this.l2ToL1Msgs = l2ToL1Msgs;
        this.noteEncryptedLogsHashes = noteEncryptedLogsHashes;
        this.encryptedLogsHashes = encryptedLogsHashes;
        this.unencryptedLogsHashes = unencryptedLogsHashes;
        this.publicDataUpdateRequests = publicDataUpdateRequests;
        this.publicCallStack = publicCallStack;
        this.gasUsed = gasUsed;
    }
    getSize() {
        return (arraySerializedSizeOfNonEmpty(this.noteHashes) +
            arraySerializedSizeOfNonEmpty(this.nullifiers) +
            arraySerializedSizeOfNonEmpty(this.l2ToL1Msgs) +
            arraySerializedSizeOfNonEmpty(this.noteEncryptedLogsHashes) +
            arraySerializedSizeOfNonEmpty(this.encryptedLogsHashes) +
            arraySerializedSizeOfNonEmpty(this.unencryptedLogsHashes) +
            arraySerializedSizeOfNonEmpty(this.publicDataUpdateRequests) +
            arraySerializedSizeOfNonEmpty(this.publicCallStack) +
            this.gasUsed.toBuffer().length);
    }
    toBuffer() {
        return serializeToBuffer(this.noteHashes, this.nullifiers, this.l2ToL1Msgs, this.noteEncryptedLogsHashes, this.encryptedLogsHashes, this.unencryptedLogsHashes, this.publicDataUpdateRequests, this.publicCallStack, this.gasUsed);
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    isEmpty() {
        return (this.noteHashes.every(x => x.isEmpty()) &&
            this.nullifiers.every(x => x.isEmpty()) &&
            this.l2ToL1Msgs.every(x => x.isEmpty()) &&
            this.noteEncryptedLogsHashes.every(x => x.isEmpty()) &&
            this.encryptedLogsHashes.every(x => x.isEmpty()) &&
            this.unencryptedLogsHashes.every(x => x.isEmpty()) &&
            this.publicDataUpdateRequests.every(x => x.isEmpty()) &&
            this.publicCallStack.every(x => x.isEmpty()) &&
            this.gasUsed.isEmpty());
    }
    [inspect.custom]() {
        // print out the non-empty fields
        return `PublicAccumulatedData {
  noteHashes: [${this.noteHashes
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  nullifiers: [${this.nullifiers
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  l2ToL1Msgs: [${this.l2ToL1Msgs
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  noteEncryptedLogsHashes: [${this.noteEncryptedLogsHashes
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  encryptedLogsHashes: [${this.encryptedLogsHashes
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  unencryptedLogsHashes: [${this.unencryptedLogsHashes
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  publicDataUpdateRequests: [${this.publicDataUpdateRequests
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  publicCallStack: [${this.publicCallStack
            .filter(x => !x.isEmpty())
            .map(h => inspect(h))
            .join(', ')}],
  gasUsed: [${inspect(this.gasUsed)}]
}`;
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns Deserialized object.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new this(reader.readArray(MAX_NOTE_HASHES_PER_TX, ScopedNoteHash), reader.readArray(MAX_NULLIFIERS_PER_TX, Nullifier), reader.readArray(MAX_L2_TO_L1_MSGS_PER_TX, ScopedL2ToL1Message), reader.readArray(MAX_NOTE_ENCRYPTED_LOGS_PER_TX, LogHash), reader.readArray(MAX_ENCRYPTED_LOGS_PER_TX, ScopedLogHash), reader.readArray(MAX_UNENCRYPTED_LOGS_PER_TX, ScopedLogHash), reader.readArray(MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX, PublicDataUpdateRequest), reader.readArray(MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, PublicCallRequest), reader.readObject(Gas));
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new this(reader.readArray(MAX_NOTE_HASHES_PER_TX, ScopedNoteHash), reader.readArray(MAX_NULLIFIERS_PER_TX, Nullifier), reader.readArray(MAX_L2_TO_L1_MSGS_PER_TX, ScopedL2ToL1Message), reader.readArray(MAX_NOTE_ENCRYPTED_LOGS_PER_TX, LogHash), reader.readArray(MAX_ENCRYPTED_LOGS_PER_TX, ScopedLogHash), reader.readArray(MAX_UNENCRYPTED_LOGS_PER_TX, ScopedLogHash), reader.readArray(MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX, PublicDataUpdateRequest), reader.readArray(MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, PublicCallRequest), reader.readObject(Gas));
    }
    /**
     * Deserializes from a string, corresponding to a write in cpp.
     * @param str - String to read from.
     * @returns Deserialized object.
     */
    static fromString(str) {
        return this.fromBuffer(Buffer.from(str, 'hex'));
    }
    static empty() {
        return new this(makeTuple(MAX_NOTE_HASHES_PER_TX, ScopedNoteHash.empty), makeTuple(MAX_NULLIFIERS_PER_TX, Nullifier.empty), makeTuple(MAX_L2_TO_L1_MSGS_PER_TX, ScopedL2ToL1Message.empty), makeTuple(MAX_NOTE_ENCRYPTED_LOGS_PER_TX, LogHash.empty), makeTuple(MAX_ENCRYPTED_LOGS_PER_TX, ScopedLogHash.empty), makeTuple(MAX_UNENCRYPTED_LOGS_PER_TX, ScopedLogHash.empty), makeTuple(MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX, PublicDataUpdateRequest.empty), makeTuple(MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, PublicCallRequest.empty), Gas.empty());
    }
}
export class PublicAccumulatedDataArrayLengths {
    constructor(noteHashes, nullifiers, l2ToL1Msgs, noteEncryptedLogsHashes, encryptedLogsHashes, unencryptedLogsHashes, publicDataUpdateRequests, publicCallStack) {
        this.noteHashes = noteHashes;
        this.nullifiers = nullifiers;
        this.l2ToL1Msgs = l2ToL1Msgs;
        this.noteEncryptedLogsHashes = noteEncryptedLogsHashes;
        this.encryptedLogsHashes = encryptedLogsHashes;
        this.unencryptedLogsHashes = unencryptedLogsHashes;
        this.publicDataUpdateRequests = publicDataUpdateRequests;
        this.publicCallStack = publicCallStack;
    }
    static new(data) {
        return new PublicAccumulatedDataArrayLengths(countAccumulatedItems(data.noteHashes), countAccumulatedItems(data.nullifiers), countAccumulatedItems(data.l2ToL1Msgs), countAccumulatedItems(data.noteEncryptedLogsHashes), countAccumulatedItems(data.encryptedLogsHashes), countAccumulatedItems(data.unencryptedLogsHashes), countAccumulatedItems(data.publicDataUpdateRequests), countAccumulatedItems(data.publicCallStack));
    }
    getSize() {
        return NUM_PUBLIC_ACCUMULATED_DATA_ARRAYS;
    }
    toBuffer() {
        return serializeToBuffer(this.noteHashes, this.nullifiers, this.l2ToL1Msgs, this.noteEncryptedLogsHashes, this.encryptedLogsHashes, this.unencryptedLogsHashes, this.publicDataUpdateRequests, this.publicCallStack);
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    isEmpty() {
        return (this.noteHashes == 0 &&
            this.nullifiers == 0 &&
            this.l2ToL1Msgs == 0 &&
            this.noteEncryptedLogsHashes == 0 &&
            this.encryptedLogsHashes == 0 &&
            this.unencryptedLogsHashes == 0 &&
            this.publicDataUpdateRequests == 0 &&
            this.publicCallStack == 0);
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns Deserialized object.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new this(reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber());
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new this(reader.readU32(), reader.readU32(), reader.readU32(), reader.readU32(), reader.readU32(), reader.readU32(), reader.readU32(), reader.readU32());
    }
    /**
     * Deserializes from a string, corresponding to a write in cpp.
     * @param str - String to read from.
     * @returns Deserialized object.
     */
    static fromString(str) {
        return this.fromBuffer(Buffer.from(str, 'hex'));
    }
    static empty() {
        return new this(0, 0, 0, 0, 0, 0, 0, 0);
    }
}
