// @aztec/circuits.js/dest/structs/kernel/public_call_data.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { Proof } from '../proof.js';
import { PublicCircuitPublicInputs } from '../public_circuit_public_inputs.js';
/**
 * Public calldata assembled from the kernel execution result and proof.
 */
export class PublicCallData {
    constructor(
    /**
     * Public inputs of the public function.
     */
    publicInputs, 
    /**
     * Proof of the call stack item execution.
     */
    proof, 
    /**
     * Hash of the L2 contract bytecode.
     */
    bytecodeHash) {
        this.publicInputs = publicInputs;
        this.proof = proof;
        this.bytecodeHash = bytecodeHash;
    }
    toBuffer() {
        return serializeToBuffer(this.publicInputs, this.proof, this.bytecodeHash);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicCallData(reader.readObject(PublicCircuitPublicInputs), reader.readObject(Proof), reader.readObject(Fr));
    }
}
