// @aztec/circuits.js/dest/structs/kernel/public_kernel_circuit_public_inputs.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { PublicCallRequest } from '../public_call_request.js';
import { PublicValidationRequests } from '../public_validation_requests.js';
import { RevertCode } from '../revert_code.js';
import { CombinedConstantData } from './combined_constant_data.js';
import { PublicAccumulatedData } from './public_accumulated_data.js';
/**
 * Outputs from the public kernel circuits.
 * All Public kernels use this shape for outputs.
 */
export class PublicKernelCircuitPublicInputs {
    constructor(
    /**
     * Data which is not modified by the circuits.
     */
    constants, 
    /**
     * Validation requests accumulated from public functions.
     */
    validationRequests, 
    /**
     * Accumulated side effects and enqueued calls that are not revertible.
     */
    endNonRevertibleData, 
    /**
     * Data accumulated from both public and private circuits.
     */
    end, 
    /**
     * Counter of the last side effect.
     */
    endSideEffectCounter, 
    /**
     * The call request for the public teardown function
     */
    publicTeardownCallRequest, 
    /**
     * The address of the fee payer for the transaction
     */
    feePayer, 
    /**
     * Indicates whether execution of the public circuit reverted.
     */
    revertCode) {
        this.constants = constants;
        this.validationRequests = validationRequests;
        this.endNonRevertibleData = endNonRevertibleData;
        this.end = end;
        this.endSideEffectCounter = endSideEffectCounter;
        this.publicTeardownCallRequest = publicTeardownCallRequest;
        this.feePayer = feePayer;
        this.revertCode = revertCode;
    }
    toBuffer() {
        return serializeToBuffer(this.constants, this.validationRequests, this.endNonRevertibleData, this.end, this.endSideEffectCounter, this.publicTeardownCallRequest, this.feePayer, this.revertCode);
    }
    clone() {
        return PublicKernelCircuitPublicInputs.fromBuffer(this.toBuffer());
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromString(str) {
        return PublicKernelCircuitPublicInputs.fromBuffer(Buffer.from(str, 'hex'));
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of PublicKernelCircuitPublicInputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicKernelCircuitPublicInputs(reader.readObject(CombinedConstantData), reader.readObject(PublicValidationRequests), reader.readObject(PublicAccumulatedData), reader.readObject(PublicAccumulatedData), reader.readNumber(), reader.readObject(PublicCallRequest), reader.readObject(AztecAddress), reader.readObject(RevertCode));
    }
    static empty() {
        return new PublicKernelCircuitPublicInputs(CombinedConstantData.empty(), PublicValidationRequests.empty(), PublicAccumulatedData.empty(), PublicAccumulatedData.empty(), 0, PublicCallRequest.empty(), AztecAddress.ZERO, RevertCode.OK);
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new PublicKernelCircuitPublicInputs(CombinedConstantData.fromFields(reader), PublicValidationRequests.fromFields(reader), PublicAccumulatedData.fromFields(reader), PublicAccumulatedData.fromFields(reader), reader.readU32(), PublicCallRequest.fromFields(reader), AztecAddress.fromFields(reader), RevertCode.fromField(reader.readField()));
    }
    [inspect.custom]() {
        return `PublicKernelCircuitPublicInputs {
      constants: ${inspect(this.constants)},
      validationRequests: ${inspect(this.validationRequests)},
      endNonRevertibleData: ${inspect(this.endNonRevertibleData)},
      end: ${inspect(this.end)},
      endSideEffectCounter: ${this.endSideEffectCounter},
      publicTeardownCallRequest: ${inspect(this.publicTeardownCallRequest)},
      feePayer: ${this.feePayer},
      revertCode: ${this.revertCode},
      }`;
    }
}
