// @aztec/circuits.js/dest/structs/kernel/public_accumulated_data_builder.js
import { padArrayEnd } from '@aztec/foundation/collection';
import { MAX_ENCRYPTED_LOGS_PER_TX, MAX_L2_TO_L1_MSGS_PER_TX, MAX_NOTE_ENCRYPTED_LOGS_PER_TX, MAX_NOTE_HASHES_PER_TX, MAX_NULLIFIERS_PER_TX, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX, MAX_UNENCRYPTED_LOGS_PER_TX, } from '../../constants.gen.js';
import { Gas } from '../gas.js';
import { ScopedL2ToL1Message } from '../l2_to_l1_message.js';
import { LogHash, ScopedLogHash } from '../log_hash.js';
import { ScopedNoteHash } from '../note_hash.js';
import { Nullifier } from '../nullifier.js';
import { PublicCallRequest } from '../public_call_request.js';
import { PublicDataUpdateRequest } from '../public_data_update_request.js';
import { PublicAccumulatedData } from './public_accumulated_data.js';
/**
 * TESTS-ONLY CLASS
 * Builder for PublicAccumulatedData, used to conveniently construct instances for testing,
 * as PublicAccumulatedData is (or will shortly be) immutable.
 *
 */
export class PublicAccumulatedDataBuilder {
    constructor() {
        this.noteHashes = [];
        this.nullifiers = [];
        this.l2ToL1Msgs = [];
        this.noteEncryptedLogsHashes = [];
        this.encryptedLogsHashes = [];
        this.unencryptedLogsHashes = [];
        this.publicDataUpdateRequests = [];
        this.publicCallStack = [];
        this.gasUsed = Gas.empty();
    }
    pushNoteHash(newNoteHash) {
        this.noteHashes.push(newNoteHash);
        return this;
    }
    withNoteHashes(noteHashes) {
        this.noteHashes = noteHashes;
        return this;
    }
    pushNullifier(newNullifier) {
        this.nullifiers.push(newNullifier);
        return this;
    }
    withNullifiers(nullifiers) {
        this.nullifiers = nullifiers;
        return this;
    }
    pushL2ToL1Msg(newL2ToL1Msg) {
        this.l2ToL1Msgs.push(newL2ToL1Msg);
        return this;
    }
    withL2ToL1Msgs(l2ToL1Msgs) {
        this.l2ToL1Msgs = l2ToL1Msgs;
        return this;
    }
    pushNoteEncryptedLogsHash(noteEncryptedLogsHash) {
        this.noteEncryptedLogsHashes.push(noteEncryptedLogsHash);
        return this;
    }
    withNoteEncryptedLogsHashes(noteEncryptedLogsHashes) {
        this.noteEncryptedLogsHashes = noteEncryptedLogsHashes;
        return this;
    }
    pushEncryptedLogsHash(encryptedLogsHash) {
        this.encryptedLogsHashes.push(encryptedLogsHash);
        return this;
    }
    withEncryptedLogsHashes(encryptedLogsHashes) {
        this.encryptedLogsHashes = encryptedLogsHashes;
        return this;
    }
    pushUnencryptedLogsHash(unencryptedLogsHash) {
        this.unencryptedLogsHashes.push(unencryptedLogsHash);
        return this;
    }
    withUnencryptedLogsHashes(unencryptedLogsHashes) {
        this.unencryptedLogsHashes = unencryptedLogsHashes;
        return this;
    }
    pushPublicDataUpdateRequest(publicDataUpdateRequest) {
        this.publicDataUpdateRequests.push(publicDataUpdateRequest);
        return this;
    }
    withPublicDataUpdateRequests(publicDataUpdateRequests) {
        this.publicDataUpdateRequests = publicDataUpdateRequests;
        return this;
    }
    pushPublicCall(publicCall) {
        this.publicCallStack.push(publicCall);
        return this;
    }
    withPublicCallStack(publicCallStack) {
        this.publicCallStack = publicCallStack;
        return this;
    }
    withGasUsed(gasUsed) {
        this.gasUsed = gasUsed;
        return this;
    }
    build() {
        return new PublicAccumulatedData(padArrayEnd(this.noteHashes, ScopedNoteHash.empty(), MAX_NOTE_HASHES_PER_TX), padArrayEnd(this.nullifiers, Nullifier.empty(), MAX_NULLIFIERS_PER_TX), padArrayEnd(this.l2ToL1Msgs, ScopedL2ToL1Message.empty(), MAX_L2_TO_L1_MSGS_PER_TX), padArrayEnd(this.noteEncryptedLogsHashes, LogHash.empty(), MAX_NOTE_ENCRYPTED_LOGS_PER_TX), padArrayEnd(this.encryptedLogsHashes, ScopedLogHash.empty(), MAX_ENCRYPTED_LOGS_PER_TX), padArrayEnd(this.unencryptedLogsHashes, ScopedLogHash.empty(), MAX_UNENCRYPTED_LOGS_PER_TX), padArrayEnd(this.publicDataUpdateRequests, PublicDataUpdateRequest.empty(), MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX), padArrayEnd(this.publicCallStack, PublicCallRequest.empty(), MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX), this.gasUsed);
    }
    static fromPublicAccumulatedData(publicAccumulatedData) {
        return new PublicAccumulatedDataBuilder()
            .withNoteHashes(publicAccumulatedData.noteHashes)
            .withNullifiers(publicAccumulatedData.nullifiers)
            .withL2ToL1Msgs(publicAccumulatedData.l2ToL1Msgs)
            .withNoteEncryptedLogsHashes(publicAccumulatedData.noteEncryptedLogsHashes)
            .withEncryptedLogsHashes(publicAccumulatedData.encryptedLogsHashes)
            .withUnencryptedLogsHashes(publicAccumulatedData.unencryptedLogsHashes)
            .withPublicDataUpdateRequests(publicAccumulatedData.publicDataUpdateRequests)
            .withPublicCallStack(publicAccumulatedData.publicCallStack)
            .withGasUsed(publicAccumulatedData.gasUsed);
    }
}
