// @aztec/circuits.js/dest/structs/kernel/vm_circuit_public_inputs.js
import { makeTuple } from '@aztec/foundation/array';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX } from '../../constants.gen.js';
import { Gas } from '../gas.js';
import { PublicCallRequest } from '../public_call_request.js';
import { PublicInnerCallRequest } from '../public_inner_call_request.js';
import { PublicValidationRequestArrayLengths, PublicValidationRequests } from '../public_validation_requests.js';
import { CombinedConstantData } from './combined_constant_data.js';
import { PublicAccumulatedData, PublicAccumulatedDataArrayLengths } from './public_accumulated_data.js';
/**
 * Call stack item on a public call.
 */
export class VMCircuitPublicInputs {
    constructor(constants, callRequest, publicCallStack, previousValidationRequestArrayLengths, validationRequests, previousAccumulatedDataArrayLengths, accumulatedData, startSideEffectCounter, endSideEffectCounter, startGasLeft, transactionFee, reverted) {
        this.constants = constants;
        this.callRequest = callRequest;
        this.publicCallStack = publicCallStack;
        this.previousValidationRequestArrayLengths = previousValidationRequestArrayLengths;
        this.validationRequests = validationRequests;
        this.previousAccumulatedDataArrayLengths = previousAccumulatedDataArrayLengths;
        this.accumulatedData = accumulatedData;
        this.startSideEffectCounter = startSideEffectCounter;
        this.endSideEffectCounter = endSideEffectCounter;
        this.startGasLeft = startGasLeft;
        this.transactionFee = transactionFee;
        this.reverted = reverted;
    }
    toBuffer() {
        return serializeToBuffer(this.constants, this.callRequest, this.publicCallStack, this.previousValidationRequestArrayLengths, this.validationRequests, this.previousAccumulatedDataArrayLengths, this.accumulatedData, this.startSideEffectCounter, this.endSideEffectCounter, this.startGasLeft, this.transactionFee, this.reverted);
    }
    clone() {
        return VMCircuitPublicInputs.fromBuffer(this.toBuffer());
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromString(str) {
        return VMCircuitPublicInputs.fromBuffer(Buffer.from(str, 'hex'));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new VMCircuitPublicInputs(reader.readObject(CombinedConstantData), reader.readObject(PublicCallRequest), reader.readArray(MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, PublicInnerCallRequest), reader.readObject(PublicValidationRequestArrayLengths), reader.readObject(PublicValidationRequests), reader.readObject(PublicAccumulatedDataArrayLengths), reader.readObject(PublicAccumulatedData), reader.readNumber(), reader.readNumber(), reader.readObject(Gas), reader.readObject(Fr), reader.readBoolean());
    }
    static empty() {
        return new VMCircuitPublicInputs(CombinedConstantData.empty(), PublicCallRequest.empty(), makeTuple(MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, PublicInnerCallRequest.empty), PublicValidationRequestArrayLengths.empty(), PublicValidationRequests.empty(), PublicAccumulatedDataArrayLengths.empty(), PublicAccumulatedData.empty(), 0, 0, Gas.empty(), Fr.ZERO, false);
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new VMCircuitPublicInputs(CombinedConstantData.fromFields(reader), PublicCallRequest.fromFields(reader), reader.readArray(MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, PublicInnerCallRequest), PublicValidationRequestArrayLengths.fromFields(reader), PublicValidationRequests.fromFields(reader), PublicAccumulatedDataArrayLengths.fromFields(reader), PublicAccumulatedData.fromFields(reader), reader.readU32(), reader.readU32(), Gas.fromFields(reader), reader.readField(), reader.readBoolean());
    }
    [inspect.custom]() {
        return `VMCircuitPublicInputs {
      constants: ${inspect(this.constants)},
      callRequest: ${inspect(this.callRequest)}
      previousValidationRequestArrayLengths: ${inspect(this.previousValidationRequestArrayLengths)},
      validationRequests: ${inspect(this.validationRequests)},
      previousAccumulatedDataArrayLengths: ${inspect(this.previousAccumulatedDataArrayLengths)},
      accumulatedData: ${inspect(this.accumulatedData)},
      startSideEffectCounter: ${this.startSideEffectCounter},
      endSideEffectCounter: ${this.endSideEffectCounter},
      startGasLeft: ${this.startGasLeft},
      transactionFee: ${this.transactionFee},
      reverted: ${this.reverted},
      }`;
    }
}
