// @aztec/circuits.js/dest/contract/events/contract_class_registered_event.js
import { FunctionSelector, bufferFromFields } from '@aztec/foundation/abi';
import { toBigIntBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import chunk from 'lodash.chunk';
import { PUBLIC_DISPATCH_SELECTOR, REGISTERER_CONTRACT_CLASS_REGISTERED_MAGIC_VALUE } from '../../constants.gen.js';
import { computeContractClassId, computePublicBytecodeCommitment } from '../contract_class_id.js';
/** Event emitted from the ContractClassRegisterer. */
export class ContractClassRegisteredEvent {
    constructor(contractClassId, version, artifactHash, privateFunctionsRoot, packedPublicBytecode) {
        this.contractClassId = contractClassId;
        this.version = version;
        this.artifactHash = artifactHash;
        this.privateFunctionsRoot = privateFunctionsRoot;
        this.packedPublicBytecode = packedPublicBytecode;
    }
    static isContractClassRegisteredEvent(log) {
        return toBigIntBE(log.subarray(0, 32)) == REGISTERER_CONTRACT_CLASS_REGISTERED_MAGIC_VALUE;
    }
    static fromLogs(logs, registererContractAddress) {
        return logs
            .filter(log => ContractClassRegisteredEvent.isContractClassRegisteredEvent(log.data))
            .filter(log => log.contractAddress.equals(registererContractAddress))
            .map(log => this.fromLogData(log.data));
    }
    static fromLogData(log) {
        if (!this.isContractClassRegisteredEvent(log)) {
            throw new Error(`Log data for ContractClassRegisteredEvent is not prefixed with magic value 0x${REGISTERER_CONTRACT_CLASS_REGISTERED_MAGIC_VALUE}`);
        }
        const reader = new BufferReader(log.subarray(32));
        const contractClassId = reader.readObject(Fr);
        const version = reader.readObject(Fr).toNumber();
        const artifactHash = reader.readObject(Fr);
        const privateFunctionsRoot = reader.readObject(Fr);
        const packedPublicBytecode = bufferFromFields(chunk(reader.readToEnd(), Fr.SIZE_IN_BYTES).map(Buffer.from).map(Fr.fromBuffer));
        return new ContractClassRegisteredEvent(contractClassId, version, artifactHash, privateFunctionsRoot, packedPublicBytecode);
    }
    toContractClassPublic() {
        const computedClassId = computeContractClassId({
            artifactHash: this.artifactHash,
            privateFunctionsRoot: this.privateFunctionsRoot,
            publicBytecodeCommitment: computePublicBytecodeCommitment(this.packedPublicBytecode),
        });
        if (!computedClassId.equals(this.contractClassId)) {
            throw new Error(`Invalid contract class id: computed ${computedClassId.toString()} but event broadcasted ${this.contractClassId.toString()}`);
        }
        if (this.version !== 1) {
            throw new Error(`Unexpected contract class version ${this.version}`);
        }
        // TODO(https://github.com/AztecProtocol/aztec-packages/issues/8985): Remove public functions.
        const publicFunctions = [];
        if (this.packedPublicBytecode.length > 0) {
            publicFunctions.push({
                selector: FunctionSelector.fromField(new Fr(PUBLIC_DISPATCH_SELECTOR)),
                bytecode: this.packedPublicBytecode,
            });
        }
        return {
            id: this.contractClassId,
            artifactHash: this.artifactHash,
            packedBytecode: this.packedPublicBytecode,
            privateFunctionsRoot: this.privateFunctionsRoot,
            publicFunctions: publicFunctions,
            version: this.version,
            privateFunctions: [],
            unconstrainedFunctions: [],
        };
    }
}
