// @aztec/circuits.js/dest/structs/rollup/append_only_tree_snapshot.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { STRING_ENCODING } from '../shared.js';
/**
 * Snapshot of an append only tree.
 *
 * Used in circuits to verify that tree insertions are performed correctly.
 */
export class AppendOnlyTreeSnapshot {
    constructor(
    /**
     * Root of the append only tree when taking the snapshot.
     */
    root, 
    /**
     * Index of the next available leaf in the append only tree.
     *
     * Note: We include the next available leaf index in the snapshot so that the snapshot can be used to verify that
     *       the insertion was performed at the correct place. If we only verified tree root then it could happen that
     *       some leaves would get overwritten and the tree root check would still pass.
     *       TLDR: We need to store the next available leaf index to ensure that the "append only" property was
     *             preserved when verifying state transitions.
     */
    nextAvailableLeafIndex) {
        this.root = root;
        this.nextAvailableLeafIndex = nextAvailableLeafIndex;
    }
    getSize() {
        return this.root.size + 4;
    }
    toBuffer() {
        return serializeToBuffer(this.root, this.nextAvailableLeafIndex);
    }
    toFields() {
        return [this.root, new Fr(this.nextAvailableLeafIndex)];
    }
    toString() {
        return this.toBuffer().toString(STRING_ENCODING);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new AppendOnlyTreeSnapshot(Fr.fromBuffer(reader), reader.readNumber());
    }
    static fromString(str) {
        return AppendOnlyTreeSnapshot.fromBuffer(Buffer.from(str, STRING_ENCODING));
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new AppendOnlyTreeSnapshot(reader.readField(), Number(reader.readField().toBigInt()));
    }
    static zero() {
        return new AppendOnlyTreeSnapshot(Fr.ZERO, 0);
    }
    isZero() {
        return this.root.isZero() && this.nextAvailableLeafIndex === 0;
    }
    [inspect.custom]() {
        return `AppendOnlyTreeSnapshot { root: ${this.root.toString()}, nextAvailableLeafIndex: ${this.nextAvailableLeafIndex} }`;
    }
    equals(other) {
        return this.root.equals(other.root) && this.nextAvailableLeafIndex === other.nextAvailableLeafIndex;
    }
}
