// @aztec/circuits.js/dest/structs/rollup/base_or_merge_rollup_public_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { PartialStateReference } from '../partial_state_reference.js';
import { ConstantRollupData } from './base_rollup.js';
/**
 * Output of the base and merge rollup circuits.
 */
export class BaseOrMergeRollupPublicInputs {
    constructor(
    /**
     * Specifies from which type of rollup circuit these inputs are from.
     */
    rollupType, 
    /**
     * Number of txs in this rollup.
     */
    numTxs, 
    /**
     * Data which is forwarded through the rollup circuits unchanged.
     */
    constants, 
    /**
     * Partial state reference at the start of the rollup circuit.
     */
    start, 
    /**
     * Partial state reference at the end of the rollup circuit.
     */
    end, 
    /**
     * SHA256 hash of transactions effects. Used to make public inputs constant-sized (to then be unpacked on-chain).
     * Note: Truncated to 31 bytes to fit in Fr.
     */
    txsEffectsHash, 
    /**
     * SHA256 hash of outhash. Used to make public inputs constant-sized (to then be unpacked on-chain).
     * Note: Truncated to 31 bytes to fit in Fr.
     */
    outHash, 
    /**
     * The summed `transaction_fee` of the constituent transactions.
     */
    accumulatedFees) {
        this.rollupType = rollupType;
        this.numTxs = numTxs;
        this.constants = constants;
        this.start = start;
        this.end = end;
        this.txsEffectsHash = txsEffectsHash;
        this.outHash = outHash;
        this.accumulatedFees = accumulatedFees;
    }
    /**
     * Deserializes from a buffer or reader.
     * Note: Corresponds to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized public inputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new BaseOrMergeRollupPublicInputs(reader.readNumber(), reader.readNumber(), reader.readObject(ConstantRollupData), reader.readObject(PartialStateReference), reader.readObject(PartialStateReference), 
        //TODO check
        Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.rollupType, this.numTxs, this.constants, this.start, this.end, this.txsEffectsHash, this.outHash, this.accumulatedFees);
    }
    /**
     * Serialize this as a hex string.
     * @returns - The hex string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /**
     * Deserializes from a hex string.
     * @param str - A hex string to deserialize from.
     * @returns A new BaseOrMergeRollupPublicInputs instance.
     */
    static fromString(str) {
        return BaseOrMergeRollupPublicInputs.fromBuffer(Buffer.from(str, 'hex'));
    }
}
