// @aztec/circuits.js/dest/structs/rollup/block_root_rollup.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { ARCHIVE_HEIGHT, L1_TO_L2_MSG_SUBTREE_SIBLING_PATH_LENGTH, NESTED_RECURSIVE_PROOF_LENGTH, NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP, } from '../../constants.gen.js';
import { RootParityInput } from '../parity/root_parity_input.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
import { PreviousRollupData } from './previous_rollup_data.js';
/**
 * Represents inputs of the block root rollup circuit.
 */
export class BlockRootRollupInputs {
    constructor(
    /**
     * The previous rollup data from 2 merge or base rollup circuits.
     */
    previousRollupData, 
    /**
     * The original and converted roots of the L1 to L2 messages subtrees.
     */
    l1ToL2Roots, 
    /**
     * New L1 to L2 messages.
     */
    newL1ToL2Messages, 
    /**
     * Sibling path of the new L1 to L2 message tree root.
     */
    newL1ToL2MessageTreeRootSiblingPath, 
    /**
     * Snapshot of the L1 to L2 message tree at the start of the rollup.
     */
    startL1ToL2MessageTreeSnapshot, 
    /**
     * Snapshot of the historical block roots tree at the start of the rollup.
     */
    startArchiveSnapshot, 
    /**
     * Sibling path of the new block tree root.
     */
    newArchiveSiblingPath, 
    /**
     * The hash of the block preceding this one.
     */
    previousBlockHash, 
    /**
     * TODO(#7346): Temporarily added prover_id while we verify block-root proofs on L1
     */
    proverId) {
        this.previousRollupData = previousRollupData;
        this.l1ToL2Roots = l1ToL2Roots;
        this.newL1ToL2Messages = newL1ToL2Messages;
        this.newL1ToL2MessageTreeRootSiblingPath = newL1ToL2MessageTreeRootSiblingPath;
        this.startL1ToL2MessageTreeSnapshot = startL1ToL2MessageTreeSnapshot;
        this.startArchiveSnapshot = startArchiveSnapshot;
        this.newArchiveSiblingPath = newArchiveSiblingPath;
        this.previousBlockHash = previousBlockHash;
        this.proverId = proverId;
    }
    /**
     * Serializes the inputs to a buffer.
     * @returns - The inputs serialized to a buffer.
     */
    toBuffer() {
        return serializeToBuffer(...BlockRootRollupInputs.getFields(this));
    }
    /**
     * Serializes the inputs to a hex string.
     * @returns The instance serialized to a hex string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /**
     * Creates a new instance from fields.
     * @param fields - Fields to create the instance from.
     * @returns A new RootRollupInputs instance.
     */
    static from(fields) {
        return new BlockRootRollupInputs(...BlockRootRollupInputs.getFields(fields));
    }
    /**
     * Extracts fields from an instance.
     * @param fields - Fields to create the instance from.
     * @returns An array of fields.
     */
    static getFields(fields) {
        return [
            fields.previousRollupData,
            fields.l1ToL2Roots,
            fields.newL1ToL2Messages,
            fields.newL1ToL2MessageTreeRootSiblingPath,
            fields.startL1ToL2MessageTreeSnapshot,
            fields.startArchiveSnapshot,
            fields.newArchiveSiblingPath,
            fields.previousBlockHash,
            fields.proverId,
        ];
    }
    /**
     * Deserializes the inputs from a buffer.
     * @param buffer - A buffer to deserialize from.
     * @returns A new RootRollupInputs instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new BlockRootRollupInputs([reader.readObject(PreviousRollupData), reader.readObject(PreviousRollupData)], RootParityInput.fromBuffer(reader, NESTED_RECURSIVE_PROOF_LENGTH), reader.readArray(NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP, Fr), reader.readArray(L1_TO_L2_MSG_SUBTREE_SIBLING_PATH_LENGTH, Fr), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readArray(ARCHIVE_HEIGHT, Fr), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    /**
     * Deserializes the inputs from a hex string.
     * @param str - A hex string to deserialize from.
     * @returns A new RootRollupInputs instance.
     */
    static fromString(str) {
        return BlockRootRollupInputs.fromBuffer(Buffer.from(str, 'hex'));
    }
}
