// @aztec/circuits.js/dest/structs/rollup/block_root_or_block_merge_public_inputs.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { AZTEC_EPOCH_DURATION } from '../../constants.gen.js';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
/**
 * Output of the block root and block merge rollup circuits.
 */
export class BlockRootOrBlockMergePublicInputs {
    constructor(
    /**
     * Archive tree immediately before this block range.
     */
    previousArchive, 
    /**
     * Archive tree after adding this block range.
     */
    newArchive, 
    /**
     * Identifier of the previous block before the range.
     */
    previousBlockHash, 
    /**
     * Identifier of the last block in the range.
     */
    endBlockHash, 
    /**
     * Global variables for the first block in the range.
     */
    startGlobalVariables, 
    /**
     * Global variables for the last block in the range.
     */
    endGlobalVariables, 
    /**
     * SHA256 hash of outhash. Used to make public inputs constant-sized (to then be unpacked on-chain).
     * Note: Truncated to 31 bytes to fit in Fr.
     */
    outHash, 
    /**
     * The summed `transaction_fee`s and recipients of the constituent blocks.
     */
    fees, 
    /**
     * Root of the verification key tree.
     */
    vkTreeRoot, 
    /**
     * Root of the protocol contract tree.
     */
    protocolContractTreeRoot, 
    /**
     * TODO(#7346): Temporarily added prover_id while we verify block-root proofs on L1
     */
    proverId) {
        this.previousArchive = previousArchive;
        this.newArchive = newArchive;
        this.previousBlockHash = previousBlockHash;
        this.endBlockHash = endBlockHash;
        this.startGlobalVariables = startGlobalVariables;
        this.endGlobalVariables = endGlobalVariables;
        this.outHash = outHash;
        this.fees = fees;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractTreeRoot = protocolContractTreeRoot;
        this.proverId = proverId;
    }
    /**
     * Deserializes from a buffer or reader.
     * @param buffer - Buffer or reader to read from.
     * @returns The deserialized public inputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new BlockRootOrBlockMergePublicInputs(reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(GlobalVariables), reader.readObject(GlobalVariables), Fr.fromBuffer(reader), reader.readArray(AZTEC_EPOCH_DURATION, FeeRecipient), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.previousArchive, this.newArchive, this.previousBlockHash, this.endBlockHash, this.startGlobalVariables, this.endGlobalVariables, this.outHash, this.fees, this.vkTreeRoot, this.protocolContractTreeRoot, this.proverId);
    }
    /**
     * Serialize this as a hex string.
     * @returns - The hex string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /**
     * Deserializes from a hex string.
     * @param str - A hex string to deserialize from.
     * @returns A new BaseOrMergeRollupPublicInputs instance.
     */
    static fromString(str) {
        return BlockRootOrBlockMergePublicInputs.fromBuffer(Buffer.from(str, 'hex'));
    }
}
export class FeeRecipient {
    constructor(recipient, value) {
        this.recipient = recipient;
        this.value = value;
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new FeeRecipient(reader.readObject(EthAddress), Fr.fromBuffer(reader));
    }
    toBuffer() {
        return serializeToBuffer(this.recipient, this.value);
    }
    static getFields(fields) {
        return [fields.recipient, fields.value];
    }
    toFields() {
        return serializeToFields(...FeeRecipient.getFields(this));
    }
    isEmpty() {
        return this.value.isZero() && this.recipient.isZero();
    }
    toFriendlyJSON() {
        if (this.isEmpty()) {
            return {};
        }
        return { recipient: this.recipient.toString(), value: this.value.toString() };
    }
}
