// @aztec/circuits.js/dest/structs/rollup/empty_block_root_rollup_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
/**
 * Represents inputs of the empty block root rollup circuit.
 */
export class EmptyBlockRootRollupInputs {
    constructor(archive, blockHash, globalVariables, vkTreeRoot, protocolContractTreeRoot, 
    // // TODO(#7346): Temporarily added prover_id while we verify block-root proofs on L1
    proverId) {
        this.archive = archive;
        this.blockHash = blockHash;
        this.globalVariables = globalVariables;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractTreeRoot = protocolContractTreeRoot;
        this.proverId = proverId;
    }
    /**
     * Serializes the inputs to a buffer.
     * @returns - The inputs serialized to a buffer.
     */
    toBuffer() {
        return serializeToBuffer(...EmptyBlockRootRollupInputs.getFields(this));
    }
    /**
     * Serializes the inputs to a hex string.
     * @returns The instance serialized to a hex string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /**
     * Creates a new instance from fields.
     * @param fields - Fields to create the instance from.
     * @returns A new instance.
     */
    static from(fields) {
        return new EmptyBlockRootRollupInputs(...EmptyBlockRootRollupInputs.getFields(fields));
    }
    /**
     * Extracts fields from an instance.
     * @param fields - Fields to create the instance from.
     * @returns An array of fields.
     */
    static getFields(fields) {
        return [
            fields.archive,
            fields.blockHash,
            fields.globalVariables,
            fields.vkTreeRoot,
            fields.protocolContractTreeRoot,
            fields.proverId,
        ];
    }
    /**
     * Deserializes the inputs from a buffer.
     * @param buffer - A buffer to deserialize from.
     * @returns A new RootRollupInputs instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new EmptyBlockRootRollupInputs(reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), GlobalVariables.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    /**
     * Deserializes the inputs from a hex string.
     * @param str - A hex string to deserialize from.
     * @returns A new RootRollupInputs instance.
     */
    static fromString(str) {
        return EmptyBlockRootRollupInputs.fromBuffer(Buffer.from(str, 'hex'));
    }
}
