// @aztec/circuits.js/dest/structs/rollup/previous_rollup_data.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { NESTED_RECURSIVE_PROOF_LENGTH, VK_TREE_HEIGHT } from '../../constants.gen.js';
import { MembershipWitness } from '../membership_witness.js';
import { RecursiveProof } from '../recursive_proof.js';
import { VerificationKeyAsFields } from '../verification_key.js';
import { BaseOrMergeRollupPublicInputs } from './base_or_merge_rollup_public_inputs.js';
/**
 * Represents the data of a previous merge or base rollup circuit.
 */
export class PreviousRollupData {
    constructor(
    /**
     * Public inputs to the base or merge rollup circuit.
     */
    baseOrMergeRollupPublicInputs, 
    /**
     * The proof of the base or merge rollup circuit.
     */
    proof, 
    /**
     * The verification key of the base or merge rollup circuit.
     */
    vk, 
    /**
     * Sibling path of the rollup circuit's vk in a big tree of rollup circuit vks.
     */
    vkWitness) {
        this.baseOrMergeRollupPublicInputs = baseOrMergeRollupPublicInputs;
        this.proof = proof;
        this.vk = vk;
        this.vkWitness = vkWitness;
    }
    /**
     * Serializes previous rollup data to a buffer.
     * @returns The buffer of the serialized previous rollup data.
     */
    toBuffer() {
        return serializeToBuffer(this.baseOrMergeRollupPublicInputs, this.proof, this.vk, this.vkWitness);
    }
    /**
     * Deserializes previous rollup data from a buffer.
     * @param buffer - A buffer to deserialize from.
     * @returns A new PreviousRollupData instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PreviousRollupData(reader.readObject(BaseOrMergeRollupPublicInputs), RecursiveProof.fromBuffer(reader, NESTED_RECURSIVE_PROOF_LENGTH), reader.readObject(VerificationKeyAsFields), MembershipWitness.fromBuffer(reader, VK_TREE_HEIGHT));
    }
}
