// @aztec/circuit-types/dest/logs/encrypted_l2_log.js
import { Fr, Point } from '@aztec/circuits.js';
import { randomBytes, sha256Trunc } from '@aztec/foundation/crypto';
/**
 * Represents an individual encrypted event log entry.
 */
export class EncryptedL2Log {
    constructor(data, maskedContractAddress) {
        this.data = data;
        this.maskedContractAddress = maskedContractAddress;
    }
    // We do not 'count' the maskedContractAddress in .length, as this method is called to calculate ciphertext length
    get length() {
        return this.data.length;
    }
    /**
     * Serializes log to a buffer.
     * @returns A buffer containing the serialized log.
     */
    toBuffer() {
        return Buffer.concat([this.maskedContractAddress.toBuffer(), this.data]);
    }
    /** Returns a JSON-friendly representation of the log. */
    toJSON() {
        return {
            data: this.data.toString('hex'),
            maskedContractAddress: this.maskedContractAddress.toString(),
        };
    }
    /** Converts a plain JSON object into an instance. */
    static fromJSON(obj) {
        return new EncryptedL2Log(Buffer.from(obj.data, 'hex'), Fr.fromString(obj.maskedContractAddress));
    }
    /**
     * Deserializes log from a buffer.
     * @param buffer - The buffer containing the log.
     * @returns Deserialized instance of `Log`.
     */
    static fromBuffer(data) {
        return new EncryptedL2Log(data.subarray(32), new Fr(data.subarray(0, 32)));
    }
    /**
     * Calculates hash of serialized logs.
     * @returns Buffer containing 248 bits of information of sha256 hash.
     */
    hash() {
        return sha256Trunc(this.data);
    }
    /**
     * Calculates siloed hash of serialized encryptedlogs.
     * @returns Buffer containing 248 bits of information of sha256 hash.
     */
    getSiloedHash() {
        const hash = this.hash();
        return sha256Trunc(Buffer.concat([this.maskedContractAddress.toBuffer(), hash]));
    }
    /**
     * Crates a random log.
     * @returns A random log.
     */
    static random() {
        const randomEphPubKey = Point.random();
        const randomLogContent = randomBytes(144 - Point.COMPRESSED_SIZE_IN_BYTES);
        const data = Buffer.concat([Fr.random().toBuffer(), randomLogContent, randomEphPubKey.toCompressedBuffer()]);
        return new EncryptedL2Log(data, Fr.random());
    }
}
