// @aztec/circuit-types/dest/logs/unencrypted_l2_log.js
import { AztecAddress } from '@aztec/circuits.js';
import { randomBytes, sha256Trunc } from '@aztec/foundation/crypto';
import { BufferReader, prefixBufferWithLength, toHumanReadable } from '@aztec/foundation/serialize';
/**
 * Represents an individual unencrypted log entry.
 */
export class UnencryptedL2Log {
    constructor(
    /**
     * Address of the contract that emitted the event
     * NOTE: It would make sense to have the address only in `FunctionL2Logs` because contract address is shared for all
     * function logs. I didn't do this because it would require us to have 2 FunctionL2Logs classes (one with contract
     * address and one without) for unencrypted and encrypted because encrypted logs can't expose the address in an
     * unencrypted form. For this reason separating the classes seems like a premature optimization.
     * TODO: Optimize this once it makes sense.
     */
    contractAddress, 
    /** The data contents of the log. */
    data) {
        this.contractAddress = contractAddress;
        this.data = data;
    }
    get length() {
        // This +4 is because we prefix the log length - see toBuffer below
        return this.data.length + AztecAddress.SIZE_IN_BYTES + 4;
    }
    /**
     * Serializes log to a buffer.
     * @returns A buffer containing the serialized log.
     */
    toBuffer() {
        return Buffer.concat([this.contractAddress.toBuffer(), prefixBufferWithLength(this.data)]);
    }
    /**
     * Serializes log to a human readable string.
     * Outputs the log data as ascii if all bytes are valid ascii characters between 32 and 126, or as hex otherwise.
     * @returns A human readable representation of the log.
     */
    toHumanReadable() {
        const payload = toHumanReadable(this.data);
        return `UnencryptedL2Log(contractAddress: ${this.contractAddress.toString()}, data: ${payload})`;
    }
    /** Returns a JSON-friendly representation of the log. */
    toJSON() {
        return {
            contractAddress: this.contractAddress.toString(),
            data: this.data.toString('hex'),
        };
    }
    /** Converts a plain JSON object into an instance. */
    static fromJSON(obj) {
        return new UnencryptedL2Log(AztecAddress.fromString(obj.contractAddress), Buffer.from(obj.data, 'hex'));
    }
    /**
     * Deserializes log from a buffer.
     * @param buffer - The buffer or buffer reader containing the log.
     * @returns Deserialized instance of `Log`.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const contractAddress = AztecAddress.fromBuffer(reader);
        const data = reader.readBuffer();
        return new UnencryptedL2Log(contractAddress, data);
    }
    /**
     * Calculates hash of serialized logs.
     * @returns Buffer containing 248 bits of information of sha256 hash.
     */
    hash() {
        const preimage = this.toBuffer();
        return sha256Trunc(preimage);
    }
    /**
     * Calculates siloed hash of serialized logs.
     * In the kernels, we use the storage contract address and not the one encoded here.
     * They should match, so it seems fine to use the existing info here.
     * @returns Buffer containing 248 bits of information of sha256 hash.
     */
    getSiloedHash() {
        const hash = this.hash();
        return sha256Trunc(Buffer.concat([this.contractAddress.toBuffer(), hash]));
    }
    /**
     * Crates a random log.
     * @returns A random log.
     */
    static random() {
        const contractAddress = AztecAddress.random();
        const dataLength = randomBytes(1)[0];
        const data = randomBytes(dataLength);
        return new UnencryptedL2Log(contractAddress, data);
    }
}
